// @empathyco/x-components/js/x-modules/search/components/partial-query-button.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { searchXModule } from '../x-module.js';

/**
 * A button that when pressed emits the {@link XEventsTypes.UserAcceptedAQuery}
 * and {@link SearchXEvents.UserClickedPartialQuery} events, expressing the user
 * intention to set the partial query.
 *
 * @public
 */
let PartialQueryButton = class PartialQueryButton extends Vue {
    /**
     * Generates the {@link WireMetadata | event metadata} object omitting the moduleName.
     *
     * @returns The {@link WireMetadata} object omitting the moduleName.
     * @internal
     */
    createEventMetadata() {
        return {
            target: this.$el,
            feature: 'partial_result'
        };
    }
    /**
     * Emits events when the button is clicked.
     *
     * @public
     */
    emitEvents() {
        this.$x.emit('UserAcceptedAQuery', this.query, this.createEventMetadata());
        this.$x.emit('UserClickedPartialQuery', this.query, this.createEventMetadata());
    }
};
__decorate([
    Prop({ required: true })
], PartialQueryButton.prototype, "query", void 0);
PartialQueryButton = __decorate([
    Component({
        mixins: [xComponentMixin(searchXModule)]
    })
], PartialQueryButton);
var script = PartialQueryButton;

export { script as default };
