// @empathyco/x-components/js/x-modules/search/components/redirection.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Watch, Component } from 'vue-property-decorator';
import { XOn } from '../../../components/decorators/bus.decorators.js';
import { State } from '../../../components/decorators/store.decorators.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { searchXModule } from '../x-module.js';

/**
 * A redirection is a component that sends the user to a link in the website. It is helpful when
 * there are queries like `help`, `shipping costs`, `my account`, where a link to a section in the
 * website will be more helpful than the set of results returned.
 *
 * @public
 */
let Redirection = class Redirection extends Vue {
    constructor() {
        super(...arguments);
        /**
         * Boolean flag which indicates if the redirection is running.
         *
         * @public
         */
        this.isRedirecting = true;
    }
    /**
     * Computed property which returns the first recommendation of the state, if any returns null.
     *
     * @returns The first redirection of the state.
     *
     * @internal
     */
    get redirection() {
        return this.redirections?.[0] ?? null;
    }
    /**
     * Watcher function which adds a setTimeout to the redirect method is the component
     * is in auto mode and there are redirections.
     *
     * @internal
     */
    redirectWithDelay() {
        this.isRedirecting = true;
        if (this.mode === 'auto' && this.redirection) {
            // eslint-disable-next-line @typescript-eslint/unbound-method
            this.timeoutId = setTimeout(this.redirect, this.delayInSeconds * 1000);
        }
    }
    /**
     * Dispatches the redirection, updating the url.
     *
     * @public
     */
    redirect() {
        clearTimeout(this.timeoutId);
        this.$x.emit('UserClickedARedirection', this.redirection);
        window.location.replace(this.redirection.url);
    }
    /**
     * Stops the redirection, emitting `UserClickedAbortARedirection` event.
     *
     * @public
     */
    abortRedirect() {
        this.cancelRedirect();
        this.$x.emit('UserClickedAbortARedirection');
    }
    /**
     * Stops the animation if the user search another query.
     *
     * @internal
     */
    cancelRedirect() {
        clearTimeout(this.timeoutId);
        this.isRedirecting = false;
    }
};
__decorate([
    State('search', 'redirections')
], Redirection.prototype, "redirections", void 0);
__decorate([
    Prop({ default: 'auto' })
], Redirection.prototype, "mode", void 0);
__decorate([
    Prop({ default: 0 })
], Redirection.prototype, "delayInSeconds", void 0);
__decorate([
    Watch('redirections', { immediate: true })
], Redirection.prototype, "redirectWithDelay", null);
__decorate([
    XOn(['UserAcceptedAQuery', 'UserClearedQuery'])
], Redirection.prototype, "cancelRedirect", null);
Redirection = __decorate([
    Component({
        mixins: [xComponentMixin(searchXModule)]
    })
], Redirection);
var script = Redirection;

export { script as default };
