// @empathyco/x-components/js/x-modules/search/components/sort-dropdown.vue.js
import script from './sort-dropdown.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c("BaseDropdown", {
    staticClass: "x-sort-dropdown",
    attrs: {
      animation: _vm.animation,
      items: _vm.items,
      value: _vm.selectedSort,
      "data-test": "sort-dropdown",
    },
    on: { change: _vm.emitUserClickedASort },
    scopedSlots: _vm._u(
      [
        {
          key: "toggle",
          fn: function (ref) {
            var isOpen = ref.isOpen;
            var item = ref.item;
            return [
              _vm._t(
                "toggle",
                function () {
                  return [_vm._v(_vm._s(item))]
                },
                null,
                { isOpen: isOpen, item: item }
              ),
            ]
          },
        },
        {
          key: "item",
          fn: function (ref) {
            var item = ref.item;
            var isHighlighted = ref.isHighlighted;
            var isSelected = ref.isSelected;
            return [
              _vm._t(
                "item",
                function () {
                  return [_vm._v(_vm._s(item))]
                },
                null,
                {
                  item: item,
                  isHighlighted: isHighlighted,
                  isSelected: isSelected,
                }
              ),
            ]
          },
        },
      ],
      null,
      true
    ),
  })
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = undefined;
  /* scoped */
  const __vue_scope_id__ = undefined;
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    undefined,
    undefined,
    undefined
  );

export { __vue_component__ as default };
