// @empathyco/x-components/js/x-modules/search/components/sort-dropdown.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import { mixins } from 'vue-class-component';
import { Prop, Component } from 'vue-property-decorator';
import __vue_component__ from '../../../components/base-dropdown.vue.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { searchXModule } from '../x-module.js';
import SortMixin from './sort.mixin.js';

/**
 * The `SortDropdown` component allows user to select the search results order. This component
 * also allows to change the selected sort programmatically.
 *
 * @remarks It extends {@link SortMixin}.
 *
 * @public
 */
let SortDropdown = class SortDropdown extends mixins(SortMixin) {
    /**
     * Emits the events related to the selection of a sort value.
     *
     * @param sort - The selected sort.
     * @internal
     */
    emitUserClickedASort(sort) {
        this.$x.emit('UserClickedASort', sort, { target: this.$el });
        this.$emit('change', sort);
    }
};
__decorate([
    Prop()
], SortDropdown.prototype, "animation", void 0);
SortDropdown = __decorate([
    Component({
        mixins: [xComponentMixin(searchXModule)],
        components: { BaseDropdown: __vue_component__ },
        model: {
            event: 'change'
        }
    })
], SortDropdown);
var script = SortDropdown;

export { script as default };
