// @empathyco/x-components/js/x-modules/search/components/spellcheck-button.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Component } from 'vue-property-decorator';
import { State } from '../../../components/decorators/store.decorators.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { searchXModule } from '../x-module.js';

/**
 * A button that when pressed emits the {@link XEventsTypes.UserAcceptedAQuery}
 * and {@link XEventsTypes.UserAcceptedSpellcheckQuery} events, expressing the user
 * intention to set the spellchecked query.
 *
 * @public
 */
let SpellcheckButton = class SpellcheckButton extends Vue {
    /**
     * Generates the {@link WireMetadata | event metadata} object omitting the moduleName.
     *
     * @returns The {@link WireMetadata} object omitting the moduleName.
     * @internal
     */
    createEventMetadata() {
        return {
            target: this.$el,
            feature: 'spellcheck'
        };
    }
    /**
     * Emits events when the button is clicked.
     *
     * @public
     */
    emitEvents() {
        this.$x.emit('UserAcceptedAQuery', this.spellcheckedQuery, this.createEventMetadata());
        this.$x.emit('UserAcceptedSpellcheckQuery', this.spellcheckedQuery, this.createEventMetadata());
    }
};
__decorate([
    State('search', 'spellcheckedQuery')
], SpellcheckButton.prototype, "spellcheckedQuery", void 0);
SpellcheckButton = __decorate([
    Component({
        mixins: [xComponentMixin(searchXModule)]
    })
], SpellcheckButton);
var script = SpellcheckButton;

export { script as default };
