// @empathyco/x-adapter/dist/cjs/empathy/empathy.adapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyAdapter = void 0;
var tslib_1 = require("tslib");
var x_deep_merge_1 = require("@empathyco/x-deep-merge");
var inversify_1 = require("inversify");
var container_const_1 = require("./container/container.const");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyAdapter = /** @class */ (function () {
    function EmpathyAdapter(config, requestors, cache) {
        this.config = config;
        this.requestors = requestors;
        this.cache = cache;
        this.configChangedListeners = new Set();
    }
    EmpathyAdapter.prototype.getNextQueries = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.nextQueries.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getTopRecommendations = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.topRecommendations.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getClicksRecommendations = function (request, requestOptions) {
        return this.requestors.clicksRecommendations.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getQueriesRecommendations = function (request, requestOptions) {
        return this.requestors.queriesRecommendations.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getSectionRecommendations = function (request, requestOptions) {
        return this.requestors.sectionRecommendations.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getUserRecommendations = function (request, requestOptions) {
        return this.requestors.userRecommendations.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getRelatedTags = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.relatedTags.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getSuggestions = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.suggestions.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.search = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.search.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.searchById = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.searchById.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.track = function (trackingRequest) {
        return this.requestors.track.request(trackingRequest);
    };
    EmpathyAdapter.prototype.invalidateCache = function () {
        if (this.cache) {
            this.cache.clear();
        }
    };
    EmpathyAdapter.prototype.setConfig = function (newConfig) {
        (0, x_deep_merge_1.deepMerge)(this.config, newConfig);
        this.notifyConfigChangedListeners();
    };
    EmpathyAdapter.prototype.addConfigChangedListener = function (listener) {
        this.configChangedListeners.add(listener);
    };
    EmpathyAdapter.prototype.removeConfigChangedListener = function (listener) {
        this.configChangedListeners.delete(listener);
    };
    EmpathyAdapter.prototype.notifyConfigChangedListeners = function () {
        var _this = this;
        this.configChangedListeners.forEach(function (listener) {
            listener(_this.config);
        });
    };
    EmpathyAdapter = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.config)),
        tslib_1.__param(1, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.requestors)),
        tslib_1.__param(2, (0, inversify_1.optional)()),
        tslib_1.__param(2, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.cacheService)),
        tslib_1.__metadata("design:paramtypes", [Object, Object, Object])
    ], EmpathyAdapter);
    return EmpathyAdapter;
}());
exports.EmpathyAdapter = EmpathyAdapter;
