// @empathyco/x-adapter/dist/esm/empathy/requestors/beacon-tracking.requestor.js
import { __decorate, __metadata } from "tslib";
import { injectable } from 'inversify';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var BeaconTrackingRequestor = /** @class */ (function () {
    function BeaconTrackingRequestor() {
        var _this = this;
        this.isAdblockPresent = false;
        this.detectAdblock()
            .then(function () {
            _this.isAdblockPresent = false;
        })
            .catch(function () {
            _this.isAdblockPresent = true;
        });
    }
    BeaconTrackingRequestor.prototype.request = function (trackingRequest) {
        var url = this.buildUrl(trackingRequest);
        if (!this.isAdblockPresent && navigator.sendBeacon(url)) {
            return Promise.resolve();
        }
        else {
            if (this.isAdblockPresent) {
                return fetch(url);
            }
            return Promise.reject('Beacon not queued');
        }
    };
    BeaconTrackingRequestor.prototype.buildUrl = function (trackingRequest) {
        var url = new URL(trackingRequest.url);
        Object.entries(trackingRequest.params).forEach(function (_a) {
            var key = _a[0], value = _a[1];
            return url.searchParams.append(key, value);
        });
        return url.href;
    };
    BeaconTrackingRequestor.prototype.detectAdblock = function () {
        var flaggedURL = 'https://google.com/pagead/js/adsbygoogle.js';
        if (!('fetch' in window)) {
            return Promise.resolve();
        }
        return fetch(new Request(flaggedURL, {
            method: 'HEAD',
            mode: 'no-cors'
        }));
    };
    BeaconTrackingRequestor = __decorate([
        injectable(),
        __metadata("design:paramtypes", [])
    ], BeaconTrackingRequestor);
    return BeaconTrackingRequestor;
}());
export { BeaconTrackingRequestor };
