// @empathyco/x-adapter/dist/esm/empathy/requestors/feature.requestor.js
import { __assign, __decorate, __metadata, __param } from "tslib";
import { getSafePropertyChain } from '@empathyco/x-get-safe-property-chain';
import { inject, injectable, multiInject, optional } from 'inversify';
import { DEPENDENCIES } from '../container/container.const';
import { pipeMappers } from '../mappers/pipe-mappers';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var FeatureRequestor = /** @class */ (function () {
    function FeatureRequestor(config, feature, httpClient, endpointsService, requestMappers, mappers, beforeRequest, beforeResponseTransformed, responseTransformed) {
        if (beforeRequest === void 0) { beforeRequest = []; }
        if (beforeResponseTransformed === void 0) { beforeResponseTransformed = []; }
        if (responseTransformed === void 0) { responseTransformed = []; }
        var _this = this;
        this.config = config;
        this.feature = feature;
        this.httpClient = httpClient;
        this.endpointsService = endpointsService;
        this.beforeRequest = beforeRequest;
        this.beforeResponseTransformed = beforeResponseTransformed;
        this.responseTransformed = responseTransformed;
        this.mapRequest = pipeMappers.apply(void 0, requestMappers);
        this.featureConfig = config.features[feature];
        this.responseMappers = {};
        Object.keys(this.featureConfig.responsePaths).forEach(function (entity) {
            _this.responseMappers[entity] = pipeMappers.apply(void 0, mappers[entity]);
        }, {});
    }
    FeatureRequestor.prototype.request = function (rawRequest, requestOptions) {
        var _this = this;
        if (requestOptions === void 0) { requestOptions = {}; }
        var feature = this.feature;
        var url = this.endpointsService.buildUrl(this.featureConfig.endpoint);
        if (requestOptions.ttlInMinutes === undefined) {
            requestOptions.ttlInMinutes = this.featureConfig.cacheTTLInMinutes;
        }
        var requestContext = { requestOptions: requestOptions, feature: feature, url: url };
        var request = this.mapRequest(rawRequest, {}, requestContext);
        var beforeRequestContext = __assign(__assign({}, requestContext), { rawRequest: rawRequest, request: request });
        this.runHooks(this.beforeRequest, beforeRequestContext);
        return this.httpClient.get(url, request, requestOptions)
            .then(function (rawResponse) {
            var responseContext = __assign(__assign({}, beforeRequestContext), { rawResponse: rawResponse });
            _this.runHooks(_this.beforeResponseTransformed, responseContext);
            var response = _this.transformResponse(rawResponse, responseContext);
            var responseTransformedContext = __assign(__assign({}, responseContext), { response: response });
            _this.runHooks(_this.responseTransformed, responseTransformedContext);
            return response;
        });
    };
    FeatureRequestor.prototype.runHooks = function (hooks, context) {
        hooks.forEach(function (hook) { return hook(context); });
    };
    FeatureRequestor.prototype.transformResponse = function (rawResponse, context) {
        var _this = this;
        return Object.entries(this.featureConfig.responsePaths)
            .reduce(function (transformedResponse, _a) {
            var entityName = _a[0], rawResponsePath = _a[1];
            var rawResponseSelection = getSafePropertyChain(rawResponse, rawResponsePath);
            if (rawResponseSelection !== undefined) {
                // @ts-ignore The error does not make sense: Record<string, any> cannot be indexed with type string???
                transformedResponse[entityName] = _this.transformResponseSelection(rawResponseSelection, entityName, context);
            }
            return transformedResponse;
        }, {});
    };
    FeatureRequestor.prototype.transformResponseSelection = function (rawResponseSelection, entityName, context) {
        var mapFn = this.responseMappers[entityName];
        return Array.isArray(rawResponseSelection)
            ? rawResponseSelection.map(function (rawEntity) { return mapFn(rawEntity, {}, context); })
            : mapFn(rawResponseSelection, {}, context);
    };
    FeatureRequestor = __decorate([
        injectable(),
        __param(0, inject(DEPENDENCIES.config)),
        __param(1, inject(DEPENDENCIES.featureName)),
        __param(2, inject(DEPENDENCIES.httpClient)),
        __param(3, inject(DEPENDENCIES.endpointsService)),
        __param(4, multiInject(DEPENDENCIES.requestMappers)),
        __param(5, inject(DEPENDENCIES.entityMappers)),
        __param(6, optional()),
        __param(6, multiInject(DEPENDENCIES.Hooks.beforeRequest)),
        __param(7, optional()),
        __param(7, multiInject(DEPENDENCIES.Hooks.beforeResponseTransformed)),
        __param(8, optional()),
        __param(8, multiInject(DEPENDENCIES.Hooks.responseTransformed)),
        __metadata("design:paramtypes", [Object, String, Object, Object, Array, Object, Array, Array, Array])
    ], FeatureRequestor);
    return FeatureRequestor;
}());
export { FeatureRequestor };
