// @empathyco/x-adapter/dist/esm/empathy/requestors/requestors.js
import { __decorate, __metadata, __param } from "tslib";
import { inject, injectable } from 'inversify';
import { DEPENDENCIES } from '../container/container.const';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var Requestors = /** @class */ (function () {
    function Requestors(nextQueries, topRecommendations, sectionRecommendations, clicksRecommendations, queriesRecommendations, userRecommendations, relatedTags, searchById, search, suggestions, track) {
        this.nextQueries = nextQueries;
        this.topRecommendations = topRecommendations;
        this.sectionRecommendations = sectionRecommendations;
        this.clicksRecommendations = clicksRecommendations;
        this.queriesRecommendations = queriesRecommendations;
        this.userRecommendations = userRecommendations;
        this.relatedTags = relatedTags;
        this.searchById = searchById;
        this.search = search;
        this.suggestions = suggestions;
        this.track = track;
    }
    Requestors = __decorate([
        injectable(),
        __param(0, inject(DEPENDENCIES.Requestors.nextQueries)),
        __param(1, inject(DEPENDENCIES.Requestors.topRecommendations)),
        __param(2, inject(DEPENDENCIES.Requestors.sectionRecommendations)),
        __param(3, inject(DEPENDENCIES.Requestors.clicksRecommendations)),
        __param(4, inject(DEPENDENCIES.Requestors.queriesRecommendations)),
        __param(5, inject(DEPENDENCIES.Requestors.userRecommendations)),
        __param(6, inject(DEPENDENCIES.Requestors.relatedTags)),
        __param(7, inject(DEPENDENCIES.Requestors.searchById)),
        __param(8, inject(DEPENDENCIES.Requestors.search)),
        __param(9, inject(DEPENDENCIES.Requestors.suggestions)),
        __param(10, inject(DEPENDENCIES.Requestors.track)),
        __metadata("design:paramtypes", [Object, Object, Object, Object, Object, Object, Object, Object, Object, Object, Object])
    ], Requestors);
    return Requestors;
}());
export { Requestors };
