// @empathyco/x-components/js/x-modules/tagging/wiring.js
import { DefaultSessionService } from '@empathyco/x-utils';
import { namespacedWireCommit, namespacedWireDispatch } from '../../wiring/namespaced-wires.factory.js';
import { namespacedDebounce } from '../../wiring/namespaced-wires.operators.js';
import { wireServiceWithoutPayload, wireService } from '../../wiring/wires.factory.js';
import { filter, mapWire } from '../../wiring/wires.operators.js';
import { createWiring } from '../../wiring/wiring.utils.js';
import { DefaultPDPAddToCartService } from './service/pdp-add-to-cart.service.js';

/**
 * `tagging` {@link XModuleName | XModule name}.
 *
 * @internal
 */
const moduleName = 'tagging';
/**
 * Debounce function for the module.
 */
const moduleDebounce = namespacedDebounce(moduleName);
/**
 * WireCommit for {@link TaggingXModule}.
 *
 * @internal
 */
const wireCommit = namespacedWireCommit(moduleName);
/**
 * WireDispatch for {@link TaggingXModule}.
 *
 * @internal
 */
const wireDispatch = namespacedWireDispatch(moduleName);
/**
 * Wires without payload factory for {@link DefaultSessionService}.
 */
const wireSessionServiceWithoutPayload = wireServiceWithoutPayload(DefaultSessionService.instance);
/**
 * Wires factory for {@link DefaultPDPAddToCartService}.
 */
const wirePDPAddToCartService = wireService(DefaultPDPAddToCartService.instance);
/**
 * Stores the given result on the local storage.
 *
 * @public
 */
const storeClickedResultWire = wirePDPAddToCartService('storeResultClicked');
/**
 * Moves the result information from the local storage to session storage.
 *
 * @public
 */
const moveClickedResultToSessionWire = mapWire(wirePDPAddToCartService('moveToSessionStorage'), (payload) => {
    return payload === 'url' ? undefined : payload;
});
/**
 * Triggers the add to cart tracking.
 *
 * @public
 */
const trackAddToCartFromSessionStorage = wirePDPAddToCartService('trackAddToCart');
/**
 * Clears the session id.
 *
 * @public
 */
const clearSessionWire = filter(wireSessionServiceWithoutPayload('clearSessionId'), ({ eventPayload: consent }) => !consent);
/**
 * Sets the tagging state `consent`.
 *
 * @public
 */
const setConsent = wireCommit('setConsent');
/**
 * Sets the tagging config state.
 *
 * @public
 */
const setTaggingConfig = wireCommit('setTaggingConfig');
/**
 * Tracks the tagging of the query.
 *
 * @public
 */
const trackQueryWire = wireDispatch('track');
/**
 * Sets the tagging state of the query tagging info using a debounce which ends if the user
 * accepts a query.
 *
 * @public
 */
const setQueryTaggingInfo = moduleDebounce(wireCommit('setQueryTaggingInfo'), ({ state }) => state.config.queryTaggingDebounceMs, {
    cancelOn: 'UserClearedQuery',
    forceOn: [
        'UserClickedAResult',
        'UserClickedAPromoted',
        'UserClickedABanner',
        'UserClickedARedirection',
        'UserReachedResultsListEnd'
    ]
});
/**
 * Tracks the tagging of the result.
 *
 * @public
 */
const trackResultClickedWire = createTrackResultWire('click');
/**
 * Performs a track of a result added to the cart.
 *
 * @public
 */
const trackAddToCartWire = createTrackResultWire('add2cart');
/**
 * Factory helper to create a wire for the track of a result.
 *
 * @param property - Key of the tagging object to track.
 * @returns A new wire for the given property of the result tagging.
 *
 * @internal
 */
function createTrackResultWire(property) {
    return filter(wireDispatch('track', ({ eventPayload: { tagging }, metadata: { location } }) => {
        const taggingInfo = tagging[property];
        taggingInfo.params.location = location;
        return taggingInfo;
    }), ({ eventPayload: { tagging } }) => !!tagging?.[property]);
}
/**
 * Wiring configuration for the {@link TaggingXModule | tagging module}.
 *
 * @internal
 */
const taggingWiring = createWiring({
    ConsentProvided: {
        setConsent
    },
    ConsentChanged: {
        clearSessionWire
    },
    PDPIsLoaded: {
        moveClickedResultToSessionWire
    },
    ResultURLTrackingEnabled: {
        moveClickedResultToSessionWire
    },
    SearchTaggingChanged: {
        setQueryTaggingInfo
    },
    SearchTaggingReceived: {
        trackQueryWire
    },
    TaggingConfigProvided: {
        setTaggingConfig
    },
    UserClickedAResult: {
        trackResultClickedWire,
        storeClickedResultWire
    },
    UserClickedResultAddToCart: {
        trackAddToCartWire,
        trackResultClickedWire
    },
    UserClickedPDPAddToCart: {
        trackAddToCartFromSessionStorage
    }
});

export { setConsent, setQueryTaggingInfo, setTaggingConfig, taggingWiring, trackAddToCartWire, trackQueryWire, trackResultClickedWire };
