// @empathyco/x-components/js/x-modules/related-tags/components/related-tag.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { State } from '../../../components/decorators/store.decorators.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { relatedTagsXModule } from '../x-module.js';

/**
 * This component renders a related tag for a query. A related tag is a descriptive keyword
 * related to the current query to fine-tune the search. For example, if you are searching
 * for *lego*, a related tag could be *city*, refining the search with *lego city*.
 *
 * @public
 */
let RelatedTag = class RelatedTag extends Vue {
    /**
     * Generates the {@link WireMetadata | event metadata} object omitting the moduleName.
     *
     * @returns The {@link WireMetadata} object omitting the moduleName.
     * @internal
     */
    createEventMetadata() {
        return {
            target: this.$el,
            feature: 'related_tag'
        };
    }
    /**
     * Emits events when the button is clicked.
     *
     * @public
     */
    emitEvents() {
        // We have to emit this events first to avoid the UserPickedARelatedTag wires to change the
        // isSelected value before emitting this selection events.
        this.$x.emit(this.isSelected ? 'UserDeselectedARelatedTag' : 'UserSelectedARelatedTag', this.relatedTag, this.createEventMetadata());
        this.$x.emit('UserPickedARelatedTag', this.relatedTag, this.createEventMetadata());
    }
    /**
     * Check if the related tag is selected or not.
     *
     * @returns If the related tag is selected.
     *
     * @internal
     */
    get isSelected() {
        return this.selectedRelatedTags.includes(this.relatedTag);
    }
    /**
     * Check if the related tag is curated and should be highlighted.
     *
     * @returns True if the related tag is curated and should be highlighted.
     *
     * @internal
     */
    get shouldHighlightCurated() {
        return this.highlightCurated && (this.relatedTag.isCurated ?? false);
    }
    /**
     * Adds the dynamic css classes to the component.
     *
     * @returns The class to be added to the component.
     *
     * @internal
     */
    get dynamicClasses() {
        return {
            'x-tag--is-curated': this.shouldHighlightCurated,
            'x-related-tag--is-curated': this.shouldHighlightCurated,
            'x-tag--is-selected': this.isSelected,
            'x-related-tag--is-selected': this.isSelected
        };
    }
};
__decorate([
    Prop({ default: false, type: Boolean })
], RelatedTag.prototype, "highlightCurated", void 0);
__decorate([
    Prop({ required: true })
], RelatedTag.prototype, "relatedTag", void 0);
__decorate([
    State('relatedTags', 'selectedRelatedTags')
], RelatedTag.prototype, "selectedRelatedTags", void 0);
RelatedTag = __decorate([
    Component({
        mixins: [xComponentMixin(relatedTagsXModule)]
    })
], RelatedTag);
var script = RelatedTag;

export { script as default };
