// @empathyco/x-components/js/x-modules/related-tags/components/related-tag.vue.js
import script from './related-tag.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(
    "button",
    {
      staticClass: "x-tag x-related-tag",
      class: _vm.dynamicClasses,
      attrs: { "data-test": "related-tag" },
      on: { click: _vm.emitEvents },
    },
    [
      _vm._t(
        "default",
        function () {
          return [_vm._v(_vm._s(_vm.relatedTag.tag))]
        },
        null,
        {
          relatedTag: _vm.relatedTag,
          isSelected: _vm.isSelected,
          shouldHighlightCurated: _vm.shouldHighlightCurated,
        }
      ),
    ],
    2
  )
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-07b615f0_0", { source: ".x-related-tag[data-v-07b615f0] {\n  white-space: nowrap;\n}\n.x-related-tag--is-selected[data-v-07b615f0] {\n  background: lightgrey;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-07b615f0";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
