// @empathyco/x-components/js/x-modules/related-tags/components/related-tags.vue.js
import script from './related-tags.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _vm.relatedTags.length
    ? _c(
        _vm.animation,
        {
          tag: "component",
          staticClass: "x-list x-related-tags",
          attrs: { tag: "ul", "data-test": "related-tags" },
        },
        _vm._l(_vm.relatedTags, function (relatedTag) {
          return _c(
            "li",
            {
              key: relatedTag.tag,
              staticClass: "x-related-tags__item",
              attrs: { "data-test": "related-tag-item" },
            },
            [
              _vm._t(
                "related-tag",
                function () {
                  return [
                    _c("RelatedTag", {
                      attrs: {
                        highlightCurated: _vm.highlightCurated,
                        relatedTag: relatedTag,
                      },
                      scopedSlots: _vm._u(
                        [
                          {
                            key: "default",
                            fn: function (ref) {
                              var relatedTag = ref.relatedTag;
                              var isSelected = ref.isSelected;
                              var shouldHighlightCurated =
                                ref.shouldHighlightCurated;
                              return [
                                _vm._t("related-tag-content", null, null, {
                                  relatedTag: relatedTag,
                                  isSelected: isSelected,
                                  shouldHighlightCurated:
                                    shouldHighlightCurated,
                                }),
                              ]
                            },
                          },
                        ],
                        null,
                        true
                      ),
                    }),
                  ]
                },
                null,
                {
                  relatedTag: relatedTag,
                  highlightCurated: _vm.highlightCurated,
                }
              ),
            ],
            2
          )
        }),
        0
      )
    : _vm._e()
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-5759b0b8_0", { source: ".x-related-tags[data-v-5759b0b8] {\n  display: flex;\n  list-style: none;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-5759b0b8";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
