// @empathyco/x-components/js/x-modules/url/components/url-handler.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import { objectFilter } from '@empathyco/x-utils';
import Vue from 'vue';
import GlobalEvents from 'vue-global-events';
import { Inject, Component } from 'vue-property-decorator';
import '../../../components/animations/animate-width.vue.js';
import '../../../components/animations/collapse-height.vue.js';
import '../../../components/animations/collapse-width.vue.js';
import '../../../components/animations/cross-fade.vue.js';
import '../../../components/animations/fade-and-slide.vue.js';
import '../../../components/animations/staggered-fade-and-slide.vue.js';
import '../../../components/animations/staggering-transition-group.vue_rollup-plugin-vue_script.vue.js';
import 'vue-runtime-helpers/dist/normalize-component.mjs';
import 'vue-runtime-helpers/dist/inject-style/browser.mjs';
import '../../../components/animations/animate-clip-path/animate-clip-path.style.scss.js';
import '../../../components/animations/animate-scale/animate-scale.style.scss.js';
import '../../../components/animations/animate-translate/animate-translate.style.scss.js';
import '../../../components/column-picker/base-column-picker-dropdown.vue.js';
import '../../../components/column-picker/base-column-picker-list.vue.js';
import '../../../components/column-picker/column-picker.mixin.js';
import '../../../components/currency/base-currency.vue.js';
import '../../../components/filters/labels/base-price-filter-label.vue_rollup-plugin-vue_script.vue.js';
import '../../../components/filters/labels/base-rating-filter-label.vue.js';
import '../../../components/icons/arrow-down.vue.js';
import '../../../components/icons/arrow-left.vue.js';
import '../../../components/icons/arrow-right.vue.js';
import '../../../components/icons/arrow-up.vue.js';
import '../../../components/icons/bag.vue.js';
import '../../../components/icons/bar-code.vue.js';
import '../../../components/icons/barcode-tiny.vue.js';
import '../../../components/icons/cart.vue.js';
import '../../../components/icons/check-tiny.vue.js';
import '../../../components/icons/check.vue.js';
import '../../../components/icons/checkbox-card-selected.vue.js';
import '../../../components/icons/checkbox-card-unselected.vue.js';
import '../../../components/icons/checkbox-selected.vue.js';
import '../../../components/icons/checkbox-unselected.vue.js';
import '../../../components/icons/chevron-down.vue.js';
import '../../../components/icons/chevron-left.vue.js';
import '../../../components/icons/chevron-right.vue.js';
import '../../../components/icons/chevron-tiny-down.vue.js';
import '../../../components/icons/chevron-tiny-left.vue.js';
import '../../../components/icons/chevron-tiny-right.vue.js';
import '../../../components/icons/chevron-tiny-up.vue.js';
import '../../../components/icons/chevron-up.vue.js';
import '../../../components/icons/cross-tiny.vue.js';
import '../../../components/icons/cross.vue.js';
import '../../../components/icons/curated-check.vue.js';
import '../../../components/icons/filters.vue.js';
import '../../../components/icons/grid-1-col.vue.js';
import '../../../components/icons/grid-2-col.vue.js';
import '../../../components/icons/grid-2-rows.vue.js';
import '../../../components/icons/hide.vue.js';
import '../../../components/icons/history-tiny.vue.js';
import '../../../components/icons/history.vue.js';
import '../../../components/icons/light-bulb-off.vue.js';
import '../../../components/icons/light-bulb-on.vue.js';
import '../../../components/icons/menu.vue.js';
import '../../../components/icons/minus-tiny.vue.js';
import '../../../components/icons/minus.vue.js';
import '../../../components/icons/nq-1.vue.js';
import '../../../components/icons/nq-2.vue.js';
import '../../../components/icons/nq-3.vue.js';
import '../../../components/icons/nq-4.vue.js';
import '../../../components/icons/plus.vue.js';
import '../../../components/icons/search.vue.js';
import '../../../components/icons/show.vue.js';
import '../../../components/icons/star.vue.js';
import '../../../components/icons/search-tiny.vue.js';
import '../../../components/icons/tag.vue.js';
import '../../../components/icons/trash-open.vue.js';
import '../../../components/icons/trash.vue.js';
import '../../../components/icons/trending-tiny.vue.js';
import '../../../components/icons/trending.vue.js';
import '../../../components/icons/user.vue.js';
import '../../../components/layouts/fixed-header-and-asides-layout.vue.js';
import '../../../components/layouts/multi-column-max-width-layout.vue.js';
import '../../../components/layouts/single-column-layout.vue.js';
import '../../../components/modals/base-events-modal-close.vue.js';
import '../../../components/modals/base-events-modal.vue.js';
import '../../../components/modals/base-events-modal-open.vue.js';
import '../../../components/modals/base-modal.vue.js';
import '../../../components/modals/base-id-modal.vue.js';
import '../../../components/modals/base-id-modal-close.vue.js';
import '../../../components/modals/base-id-modal-open.vue.js';
import '../../../components/panels/base-header-toggle-panel.vue.js';
import '../../../components/panels/base-toggle-panel.vue.js';
import '../../../components/panels/base-id-toggle-panel.vue.js';
import '../../../components/panels/base-id-toggle-panel-button.vue.js';
import '../../../components/result/base-result-add-to-cart.vue.js';
import '../../../components/result/base-result-current-price.vue.js';
import '../../../components/result/base-result-image.vue.js';
import '../../../components/result/base-result-link.vue.js';
import '../../../components/result/base-result-previous-price.vue.js';
import '../../../components/result/base-result-fallback-image.vue.js';
import '../../../components/result/base-result-placeholder-image.vue.js';
import '../../../components/result/base-result-rating.vue.js';
import '../../../components/scroll/base-scroll.vue.js';
import '../../../components/scroll/scroll.mixin.js';
import '../../../components/suggestions/base-suggestion.vue.js';
import '../../../components/suggestions/base-suggestions.vue.js';
import '../../../components/auto-progress-bar.vue.js';
import '../../../components/base-dropdown.vue.js';
import '../../../components/base-event-button.vue.js';
import '../../../components/base-grid.vue.js';
import '../../../components/base-keyboard-navigation.vue.js';
import '../../../components/base-rating.vue.js';
import '../../../components/base-variable-column-grid.vue.js';
import '../../../components/global-x-bus.vue_rollup-plugin-vue_script.vue.js';
import '../../../components/items-list.vue.js';
import '../../../components/location-provider.vue_rollup-plugin-vue_script.vue.js';
import '../../../components/no-element.js';
import '../../../components/sliding-panel.vue.js';
import '../../../components/snippet-callbacks.vue.js';
import { XOn } from '../../../components/decorators/bus.decorators.js';
import 'vue-class-component';
import { isArrayEmpty } from '../../../utils/array.js';
import { State } from '../../../components/decorators/store.decorators.js';
import '../../../components/items-list-injection.mixin.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { initialUrlState } from '../store/initial-state.js';
import { urlXModule } from '../x-module.js';

/**
 * This component manage the browser URL parameters to perserve them through reloads and browser
 * history navigation. It allow to configure the default url parameter names using its attributes.
 * This component doesn't render elements to the DOM.
 *
 * @public
 */
let UrlHandler = class UrlHandler extends Vue {
    constructor() {
        super(...arguments);
        /**
         * Flag to know if the params were already loaded from the URL.
         *
         * @internal
         */
        this.urlLoaded = false;
        /**
         * Flag to know if the page has been persisted by the browser's back-forward cache.
         *
         * @internal
         */
        this.isPagePersisted = false;
    }
    /**
     * Computed to know which params we must get from URL. It gets the params names from the initial
     * state, to get all default params names, and also from the `$attrs` to get the extra params
     * names to take into account.
     *
     * @returns An array with the name of the params.
     *
     * @internal
     */
    get managedParamsNames() {
        return Object.keys({ ...initialUrlState, ...this.$attrs });
    }
    /**
     * Returns the mapping of the param keys used in the URL is configured through $attrs. This way
     * we can support any param and extra param, no matters its name.
     *
     * @param paramName - The param name to get the Url key.
     * @returns The key used in the URL for the `paramName` passed.
     *
     * @internal
     */
    getUrlKey(paramName) {
        return this.$attrs[paramName] ?? paramName;
    }
    /**
     * To emit the Url events just when the URL is load, and before the components mounted events
     * and state changes, we do it in the created of this component.
     */
    created() {
        this.emitEvents();
    }
    /**
     * Updates the browser URL with the new {@link UrlParams} using the history `pushState` method.
     *
     * @param newUrlParams - The new params to update browser URL.
     */
    updateUrlWithPush(newUrlParams) {
        this.updateUrl(newUrlParams, window.history.pushState.bind(window.history));
    }
    /**
     * Updates the browser URL with the new {@link UrlParams} using the history `replaceState`
     * method.
     *
     * @param newUrlParams - The new params to update browser URL.
     */
    updateUrlWithReplace(newUrlParams) {
        this.updateUrl(newUrlParams, window.history.replaceState.bind(window.history));
    }
    /**
     * Handler of the
     * {@link https://developer.mozilla.org/en-US/docs/Web/API/Window/pageshow_event | pageshow }
     * event.
     *
     * @remarks The pageshow event is listened to check if the browser has performed a navigation
     * using the back-forward cache. This information is available in the
     * PageTransitionEvent.persisted property.
     *
     * @param event - The page transition event.
     * @internal
     */
    onPageShow(event) {
        this.isPagePersisted = event.persisted;
        if (event.persisted) {
            // The internal url is reset due to the back-forward cache storing the previous value which
            // is no longer valid.
            this.url = undefined;
        }
    }
    /**
     * Emits the {@link UrlXEvents.ParamsLoadedFromUrl} XEvent,
     * the {@link UrlXEvents.ExtraParamsLoadedFromUrl} XEvent and, if there is query, also emits
     * the {@link XEventsTypes.UserOpenXProgrammatically}.
     *
     * @internal
     */
    emitEvents() {
        const { all, extra } = this.parseUrlParams();
        const metadata = this.createWireMetadata();
        this.$x.emit('ParamsLoadedFromUrl', all, metadata);
        this.$x.emit('ExtraParamsLoadedFromUrl', extra, metadata);
        // TODO: Move this logic from here.
        if (all.query) {
            this.$x.emit('UserOpenXProgrammatically', undefined, metadata);
        }
        this.urlLoaded = true;
    }
    /**
     * Creates the wire metadata to include in every emitted {@link XEvent | XEvents}.
     *
     * @returns The {@link WireMetadata | metadata}.
     * @internal
     */
    createWireMetadata() {
        return {
            feature: 'url',
            location: this.detectLocation()
        };
    }
    /**
     * Detects the {@link FeatureLocation | location} used to build the
     * {@link QueryOriginInit | events metadata origin}.
     *
     * @returns The {@link FeatureLocation | location}.
     * @internal
     */
    detectLocation() {
        const currentUrl = new URL(window.location.href);
        const previousUrl = this.url;
        this.url = currentUrl;
        const isInternalNavigation = previousUrl?.search !== currentUrl.search && previousUrl?.pathname === currentUrl.pathname;
        if (isInternalNavigation) {
            return 'url_history';
        }
        if (this.isNavigatingFromPdp()) {
            return 'url_history_pdp';
        }
        return 'external';
    }
    /**
     * Check if the navigation is from a product page.
     *
     * @remarks Due to Safari 14 not supporting the new and standard PerformanceNavigationTiming
     * API, we are falling back to the deprecated one, PerformanceNavigation. We also fallback to
     * this API whenever we get a navigationType equal to reload, because Safari has a bug that the
     * navigationType is permanently set to reload after you have reload the page and it never
     * resets. As some browsers have a back-forward cache implemented, we also take into account if
     * the page is persisted.
     *
     * @returns True if the navigation is from a product page, false otherwise.
     * @internal
     */
    isNavigatingFromPdp() {
        const isPagePersisted = this.isPagePersisted;
        const navigationEntries = window.performance.getEntriesByType('navigation');
        const navigationType = navigationEntries[0]?.type;
        const useFallbackStrategy = window.performance.navigation &&
            (isArrayEmpty(navigationEntries) || navigationType === 'reload');
        // Reset internal isPagePersisted property value
        this.isPagePersisted = false;
        if (useFallbackStrategy) {
            const { type: fallbackNavigationType, TYPE_BACK_FORWARD, TYPE_NAVIGATE } = window.performance.navigation;
            const isNavigatingInSpa = !!this.snippetConfig?.isSpa && fallbackNavigationType === TYPE_NAVIGATE;
            return fallbackNavigationType === TYPE_BACK_FORWARD || isNavigatingInSpa || isPagePersisted;
        }
        else {
            const isNavigatingInSpa = !!this.snippetConfig?.isSpa && navigationType === 'navigate';
            return navigationType === 'back_forward' || isNavigatingInSpa || isPagePersisted;
        }
    }
    /**
     * Gets the {@link UrlParams} from the URL, including only the params defined by `paramsNames`.
     *
     * @returns ParsedUrlParams obtained from URL.
     * @internal
     */
    parseUrlParams() {
        const urlSearchParams = new URL(window.location.href).searchParams;
        return this.managedParamsNames.reduce((params, name) => {
            const urlKey = this.getUrlKey(name);
            if (urlSearchParams.has(urlKey)) {
                if (name in initialUrlState) {
                    const urlValue = urlSearchParams.getAll(urlKey);
                    params.all[name] = this.parseUrlParam(name, urlValue);
                }
                else {
                    params.all[name] = params.extra[name] = urlSearchParams.get(urlKey);
                }
            }
            return params;
        }, { all: { ...initialUrlState }, extra: { ...this.initialExtraParams } });
    }
    /**
     * Updates the browser URL with the passed `newUrlParams` and using the browser history method
     * passed as `historyMethod`. It only updates the browser history if the new URL is different
     * from the current.
     *
     * @param newUrlParams - The new params to add to the browser URL.
     * @param historyMethod - The browser history method used to add the new URL.
     *
     * @internal
     */
    updateUrl(newUrlParams, historyMethod) {
        if (this.urlLoaded) {
            const url = new URL(window.location.href);
            this.deleteUrlParameters(url);
            this.setUrlParameters(url, newUrlParams);
            url.href = url.href.replace(/\+/g, '%20');
            if (url.href !== window.location.href) {
                historyMethod({ ...window.history.state }, document.title, url.href);
            }
            this.url = url;
        }
    }
    /**
     * Deletes all the parameters in the passed URL.
     *
     * @param url - The URL to remove parameters from.
     * @internal
     */
    deleteUrlParameters(url) {
        this.managedParamsNames.forEach(paramName => url.searchParams.delete(this.getUrlKey(paramName)));
    }
    /**
     * Set all the provided parameters to the url with the mapped key.
     *
     * @param url - The current URL.
     * @param urlParams - The list of parameters to add.
     * @remarks The params are filtered because there maybe received extra params which will not be
     * managed by URL. This is defined by the `managedParamsNames` computed. Also, the parameters
     * are sorted Alphabetically to produce always the same URL with the same parameters.This is
     * important for SEO purposes.
     *
     * @internal
     */
    setUrlParameters(url, urlParams) {
        // Only when there is a query the rest of the parameters are valid.
        if (!urlParams.query) {
            return;
        }
        const filteredParams = objectFilter(urlParams, paramName => this.managedParamsNames.includes(paramName));
        const sortedParameters = this.sortParams(filteredParams);
        sortedParameters.forEach(([paramName, paramValue]) => {
            const urlParamKey = this.getUrlKey(paramName);
            if (Array.isArray(paramValue)) {
                paramValue.forEach(value => {
                    url.searchParams.append(urlParamKey, String(value));
                });
            }
            else {
                url.searchParams.set(urlParamKey, String(paramValue));
            }
        });
    }
    /**
     * Sorts the params in a tuple array [key,value] to generate always the same URL with the params
     * in the same order.
     *
     * @param urlParams - The {@link UrlParams} to sort.
     * @returns An array of tuples with the key-value of each paramter, sorted by key.
     * @internal
     */
    sortParams(urlParams) {
        return Object.entries(urlParams).sort(([param1], [param2]) => {
            return param1 < param2 ? -1 : 1;
        });
    }
    /**
     * Returns the URL param value parsed depending on its type in the initial store state. As we
     * can not know what type can have an extra param, all extra params are parsed as strings. We
     * know if it is an extra param because it is not in the initial state.
     *
     * @param name - The name of the param in {@link UrlParams}.
     * @param value - The `URLSearchParams` value as an arry of strings.
     * @returns The parsed value.
     *
     * @internal
     */
    parseUrlParam(name, value) {
        switch (typeof initialUrlState[name]) {
            case 'number':
                return Number(value[0]);
            case 'boolean':
                return value[0].toLowerCase() === 'true';
            case 'string':
                return value[0];
            default:
                // array
                return value;
        }
    }
};
__decorate([
    Inject({ default: undefined })
], UrlHandler.prototype, "snippetConfig", void 0);
__decorate([
    State('url', 'initialExtraParams')
], UrlHandler.prototype, "initialExtraParams", void 0);
__decorate([
    XOn('PushableUrlStateChanged')
], UrlHandler.prototype, "updateUrlWithPush", null);
__decorate([
    XOn('ReplaceableUrlStateChanged')
], UrlHandler.prototype, "updateUrlWithReplace", null);
UrlHandler = __decorate([
    Component({
        components: {
            GlobalEvents
        },
        mixins: [xComponentMixin(urlXModule)]
    })
], UrlHandler);
var script = UrlHandler;

export { script as default };
