// @empathyco/x-adapter/dist/esm/empathy/mappers/request/params/empathy-request-filters-solr-syntax.mapper.js
import { __decorate, __metadata, __param } from "tslib";
import { isEditableNumberRangeFilter } from '@empathyco/x-types';
import { inject, injectable } from 'inversify';
import { DEPENDENCIES } from '../../../container/container.const';
import { Logger } from '../../../logger';
// TODO When the backend updates their services to be agnostic of the solr db syntax, we can remove this class
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyRequestFiltersSolrSyntaxMapper = /** @class */ (function () {
    function EmpathyRequestFiltersSolrSyntaxMapper(config) {
        this.config = config;
        this.logger = Logger.child('EmpathyRequestFiltersSolrSyntaxMapper');
    }
    EmpathyRequestFiltersSolrSyntaxMapper.prototype.map = function (filters, filterValues) {
        filterValues.push.apply(filterValues, this.composeFilters(filters));
        return filterValues;
    };
    EmpathyRequestFiltersSolrSyntaxMapper.prototype.composeFilters = function (filters) {
        return this.isArrayOfEditableNumberRangeFilters(filters)
            ? this.mapEditableNumberRangeFiltersList(filters)
            : filters.map(function (filter) { return filter.id.toString(); });
    };
    EmpathyRequestFiltersSolrSyntaxMapper.prototype.mapEditableNumberRangeFiltersList = function (filters) {
        var _this = this;
        var facetId = filters[0].facetId;
        var facetsConfig = this.config.mappings.facets;
        var template = (facetsConfig.named[facetId] || facetsConfig.default).template;
        if (template) {
            return filters.map(function (filter) { return _this.mapEditableNumberRangeFilter(template, filter); });
        }
        this.logger.warn("The facet with facetId ".concat(facetId, " doesn't have a template configured."));
        return [];
    };
    EmpathyRequestFiltersSolrSyntaxMapper.prototype.mapEditableNumberRangeFilter = function (template, _a) {
        var _b = _a.range, min = _b.min, max = _b.max;
        function parseNullValues(value) {
            return value === null ? '*' : String(value);
        }
        return template
            .replace(/<min>/g, parseNullValues(min))
            .replace(/<max>/g, parseNullValues(max));
    };
    /**
     * Check if the filters passed are of type {@link @empathyco/x-types#EditableNumberRangeFilter | EditableNumberRangeFilter}.
     *
     * @param filters - The array of filters to check.
     *
     * @internal
     */
    EmpathyRequestFiltersSolrSyntaxMapper.prototype.isArrayOfEditableNumberRangeFilters = function (filters) {
        return isEditableNumberRangeFilter(filters[0]);
    };
    EmpathyRequestFiltersSolrSyntaxMapper = __decorate([
        injectable(),
        __param(0, inject(DEPENDENCIES.config)),
        __metadata("design:paramtypes", [Object])
    ], EmpathyRequestFiltersSolrSyntaxMapper);
    return EmpathyRequestFiltersSolrSyntaxMapper;
}());
export { EmpathyRequestFiltersSolrSyntaxMapper };
