// @empathyco/x-adapter/dist/esm/empathy/builder/empathy-adapter.builder.js
import { BrowserStorageService } from '@empathyco/x-storage-service';
import { deepMerge } from '@empathyco/x-deep-merge';
import { Container, injectable } from 'inversify';
import { ContainerConfigParser } from '../container/container-config-parser';
import { BINDINGS } from '../container/container.bindings';
import { DEPENDENCIES } from '../container/container.const';
import { EmpathyAdapter } from '../empathy.adapter';
import { Logger } from '../logger';
import { EmpathyCacheService } from '../services/empathy-cache.service';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyAdapterBuilder = /** @class */ (function () {
    function EmpathyAdapterBuilder(container, defaultBindingsConfig, adapterClass) {
        if (container === void 0) { container = new Container(); }
        if (defaultBindingsConfig === void 0) { defaultBindingsConfig = BINDINGS; }
        if (adapterClass === void 0) { adapterClass = EmpathyAdapter; }
        this.container = container;
        this.defaultBindingsConfig = defaultBindingsConfig;
        this.adapterClass = adapterClass;
        this.initContainerBindings();
        // There is only one config object. As it has been binded with inversify toConstant() method, we are just getting a reference to it here
        this.config = container.get(DEPENDENCIES.config);
    }
    EmpathyAdapterBuilder.prototype.enableCache = function (_a) {
        var _b = _a === void 0 ? {} : _a, _c = _b.storageService, storageService = _c === void 0 ? typeof localStorage !== 'undefined'
            ? new BrowserStorageService(localStorage)
            : undefined : _c, _d = _b.cacheService, cacheService = _d === void 0 ? EmpathyCacheService : _d;
        if (storageService && cacheService) {
            this.container.bind(DEPENDENCIES.storageService).toConstantValue(storageService);
            this.container.bind(DEPENDENCIES.cacheService).to(cacheService);
        }
        else {
            Logger.warn('Tried to enable cache with invalid options');
        }
        return this;
    };
    EmpathyAdapterBuilder.prototype.addClassMapper = function (mapper, entity, feature) {
        var mapperKey = DEPENDENCIES.ResponseMappers[entity];
        if (feature) {
            var requestorKey = DEPENDENCIES.Requestors[feature];
            this.container.bind(mapperKey).to(mapper).whenAnyAncestorIs(requestorKey);
        }
        else {
            this.container.bind(mapperKey).to(mapper);
        }
        return this;
    };
    EmpathyAdapterBuilder.prototype.addMapper = function (mapper, entity, feature) {
        var mapperClass = this.createMapperClass(mapper);
        this.addClassMapper(mapperClass, entity, feature);
        return this;
    };
    EmpathyAdapterBuilder.prototype.replaceMapper = function (mapper, entity) {
        var mapperClass = this.createMapperClass(mapper);
        this.replaceClassMapper(mapperClass, entity);
        return this;
    };
    EmpathyAdapterBuilder.prototype.replaceClassMapper = function (mapper, entity) {
        var mapperKey = DEPENDENCIES.ResponseMappers[entity];
        this.container.rebind(mapperKey).to(mapper);
        return this;
    };
    EmpathyAdapterBuilder.prototype.addClassRequestMapper = function (mapper, feature) {
        if (feature) {
            var requestorKey = DEPENDENCIES.Requestors[feature];
            this.container.bind(DEPENDENCIES.requestMappers).to(mapper).whenAnyAncestorIs(requestorKey);
        }
        else {
            this.container.bind(DEPENDENCIES.requestMappers).to(mapper);
        }
        return this;
    };
    EmpathyAdapterBuilder.prototype.addRequestMapper = function (mapper, feature) {
        var mapperClass = this.createRequestMapperClass(mapper);
        this.addClassRequestMapper(mapperClass, feature);
        return this;
    };
    EmpathyAdapterBuilder.prototype.replaceRequestMapper = function (mapper) {
        var mapperClass = this.createRequestMapperClass(mapper);
        this.replaceClassRequestMapper(mapperClass);
        return this;
    };
    EmpathyAdapterBuilder.prototype.replaceClassRequestMapper = function (mapper) {
        this.container.rebind(DEPENDENCIES.requestMappers).to(mapper);
        return this;
    };
    EmpathyAdapterBuilder.prototype.onResponseTransformed = function (hook, feature) {
        return this.addHook(hook, 'responseTransformed', feature);
    };
    EmpathyAdapterBuilder.prototype.onBeforeRequest = function (hook, feature) {
        return this.addHook(hook, 'beforeRequest', feature);
    };
    EmpathyAdapterBuilder.prototype.onBeforeResponseTransformed = function (hook, feature) {
        return this.addHook(hook, 'beforeResponseTransformed', feature);
    };
    EmpathyAdapterBuilder.prototype.configureContainer = function (configurator) {
        this.configurator = configurator;
        return this;
    };
    EmpathyAdapterBuilder.prototype.setRequestParams = function (requestParams) {
        Object.assign(this.config.requestParams, requestParams);
        return this;
    };
    EmpathyAdapterBuilder.prototype.setLang = function (lang) {
        this.config.requestParams.lang = lang;
        return this;
    };
    EmpathyAdapterBuilder.prototype.setScope = function (scope) {
        this.config.requestParams.scope = scope;
        return this;
    };
    EmpathyAdapterBuilder.prototype.setInstance = function (instance) {
        this.config.instance = instance;
        return this;
    };
    EmpathyAdapterBuilder.prototype.setUserInfo = function (user) {
        Object.assign(this.config.requestParams, user);
        return this;
    };
    EmpathyAdapterBuilder.prototype.setEnvironment = function (environment) {
        this.config.env = environment;
        return this;
    };
    EmpathyAdapterBuilder.prototype.setFeatureConfig = function (featureName, featureConfig) {
        deepMerge(this.config.features[featureName], featureConfig);
        return this;
    };
    EmpathyAdapterBuilder.prototype.setResultTrackingConfig = function (resultTrackingConfig) {
        deepMerge(this.config.mappings.tracking.result, resultTrackingConfig);
        return this;
    };
    EmpathyAdapterBuilder.prototype.setFacetConfig = function (config, facetId) {
        if (facetId) {
            var namedFacetsConfig = this.config.mappings.facets.named;
            if (!namedFacetsConfig[facetId]) {
                // We will complete the partial facet config in the build method, so we can cast this safely here.
                namedFacetsConfig[facetId] = config;
            }
            else {
                deepMerge(namedFacetsConfig[facetId], config);
            }
        }
        else {
            deepMerge(this.config.mappings.facets.default, config);
        }
        return this;
    };
    EmpathyAdapterBuilder.prototype.setQueryConfig = function (queryConfig) {
        deepMerge(this.config.mappings.query, queryConfig);
        return this;
    };
    EmpathyAdapterBuilder.prototype.withConfiguration = function (config) {
        deepMerge(this.config, config);
        return this;
    };
    EmpathyAdapterBuilder.prototype.build = function () {
        if (this.configurator) {
            this.configurator(this.container);
        }
        this.completePartialFacetsConfig();
        return this.container.resolve(this.adapterClass);
    };
    EmpathyAdapterBuilder.prototype.initContainerBindings = function () {
        new ContainerConfigParser(this.defaultBindingsConfig, this.container).parse();
    };
    EmpathyAdapterBuilder.prototype.createMapperClass = function (mapFn) {
        return injectable()(/** @class */ (function () {
            function MockedMapper() {
                this.map = mapFn;
            }
            return MockedMapper;
        }()));
    };
    EmpathyAdapterBuilder.prototype.createRequestMapperClass = function (mapFn) {
        return injectable()(/** @class */ (function () {
            function MockedMapper() {
                this.map = mapFn;
            }
            return MockedMapper;
        }()));
    };
    EmpathyAdapterBuilder.prototype.addHook = function (hook, hookName, feature) {
        if (feature) {
            this.container
                .bind(DEPENDENCIES.Hooks[hookName])
                .toFunction(hook)
                .whenInjectedInto(DEPENDENCIES.Requestors[feature]);
        }
        else {
            this.container.bind(DEPENDENCIES.Hooks[hookName]).toFunction(hook);
        }
        return this;
    };
    EmpathyAdapterBuilder.prototype.completePartialFacetsConfig = function () {
        var config = this.container.get(DEPENDENCIES.config);
        var facetsConfig = config.mappings.facets;
        Object.entries(facetsConfig.named).forEach(function (_a) {
            var facetId = _a[0], namedFacetConfig = _a[1];
            facetsConfig.named[facetId] = deepMerge({}, facetsConfig.default, namedFacetConfig);
        });
    };
    return EmpathyAdapterBuilder;
}());
export { EmpathyAdapterBuilder };
