// @empathyco/x-adapter/dist/cjs/empathy/mappers/response/results/empathy-result.mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyResultMapper = void 0;
var tslib_1 = require("tslib");
var x_get_safe_property_chain_1 = require("@empathyco/x-get-safe-property-chain");
var inversify_1 = require("inversify");
var container_const_1 = require("../../../container/container.const");
var logger_1 = require("../../../logger");
var pipe_mappers_1 = require("../../pipe-mappers");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyResultMapper = /** @class */ (function () {
    function EmpathyResultMapper(config, taggingMappers) {
        this.logger = logger_1.Logger.child('EmpathyResultMapper');
        this.mapTagging = pipe_mappers_1.pipeMappers.apply(void 0, taggingMappers);
        this.trackingResultConfig = config.mappings.tracking.result;
    }
    EmpathyResultMapper.prototype.map = function (rawResult, result, context) {
        var value = Number.parseFloat(rawResult.price);
        var originalValue = Number.parseFloat((rawResult.originalPrice || rawResult.price));
        var sku = rawResult.eb_sku || '';
        return Object.assign(result, {
            modelName: 'Result',
            type: 'default',
            id: rawResult.id,
            name: rawResult.name || '',
            url: rawResult.url,
            images: rawResult.image ? [rawResult.image] : [],
            price: {
                value: value,
                originalValue: originalValue,
                hasDiscount: value < originalValue
            },
            rating: {
                value: rawResult.rating !== undefined ? Number.parseFloat(rawResult.rating) : null
            },
            identifier: {
                value: sku
            },
            tagging: this.createResultTagging(rawResult, result.tagging || {}, context),
            isWishlisted: rawResult.isWishlisted || false
        });
    };
    EmpathyResultMapper.prototype.createResultTagging = function (rawResult, initialResultTagging, context) {
        var _this = this;
        return Object.entries(this.trackingResultConfig).reduce(function (resultTagging, _a) {
            var key = _a[0], path = _a[1];
            var rawTagging = (0, x_get_safe_property_chain_1.getSafePropertyChain)(rawResult, path);
            if (rawTagging) {
                resultTagging[key] = _this.mapTagging(rawTagging, {}, context);
            }
            else {
                if (process.env.NODE_ENV !== 'production') {
                    _this.logger.warn("Tagging for event \"".concat(key, "\" at path \"").concat(path, "\" not found when mapping for feature \"").concat(context.feature, "\""));
                }
            }
            return resultTagging;
        }, initialResultTagging);
    };
    EmpathyResultMapper = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.config)),
        tslib_1.__param(1, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.Helpers.tagging)),
        tslib_1.__metadata("design:paramtypes", [Object, Array])
    ], EmpathyResultMapper);
    return EmpathyResultMapper;
}());
exports.EmpathyResultMapper = EmpathyResultMapper;
