// @empathyco/x-adapter/dist/cjs/empathy/mappers/response/results/empathy-result-query-tagging.mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyResultQueryTaggingMapper = void 0;
var tslib_1 = require("tslib");
var x_get_safe_property_chain_1 = require("@empathyco/x-get-safe-property-chain");
var inversify_1 = require("inversify");
var container_const_1 = require("../../../container/container.const");
var logger_1 = require("../../../logger");
var pipe_mappers_1 = require("../../pipe-mappers");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyResultQueryTaggingMapper = /** @class */ (function () {
    function EmpathyResultQueryTaggingMapper(config, queryTaggingMappers) {
        this.logger = logger_1.Logger.child('EmpathyResultQueryTaggingMapper');
        this.mapQueryTagging = pipe_mappers_1.pipeMappers.apply(void 0, queryTaggingMappers);
        this.queryTrackingPath = config.features.search.responsePaths.queryTagging;
    }
    EmpathyResultQueryTaggingMapper.prototype.map = function (_rawResult, result, context) {
        Object.assign(result.tagging, { query: context.resultsQueryTagging || this.mapNewQueryTagging(context) });
        return result;
    };
    EmpathyResultQueryTaggingMapper.prototype.mapNewQueryTagging = function (context) {
        var queryTagging = (0, x_get_safe_property_chain_1.getSafePropertyChain)(context.rawResponse, this.queryTrackingPath);
        if (queryTagging) {
            return context.resultsQueryTagging = this.mapQueryTagging(queryTagging, {}, context);
        }
        this.logger.warn("Query tagging at path \"".concat(this.queryTrackingPath, "\" not found when mapping for feature \"").concat(context.feature, "\""));
        return null;
    };
    EmpathyResultQueryTaggingMapper = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.config)),
        tslib_1.__param(1, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.queryTagging)),
        tslib_1.__metadata("design:paramtypes", [Object, Array])
    ], EmpathyResultQueryTaggingMapper);
    return EmpathyResultQueryTaggingMapper;
}());
exports.EmpathyResultQueryTaggingMapper = EmpathyResultQueryTaggingMapper;
