// @empathyco/x-components/js/x-modules/tagging/components/tagging.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Inject, Component } from 'vue-property-decorator';
import { XEmit } from '../../../components/decorators/bus.decorators.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { taggingXModule } from '../x-module.js';

/**
 * This component enables and manages the sending of information to the
 * {@link https://docs.empathy.co/develop-empathy-platform/api-reference/tagging-api.html |
 * Empathy Tagging API}. It allows to enable or disable the session id management through the
 * `consent` prop.
 *
 * @public
 */
let Tagging = class Tagging extends Vue {
    /**
     * The active consent, selected from the `consent` prop and the `snippetConfig.consent`
     * property. False by default.
     *
     * @remarks If the consent is undefined in the prop and in the snippetConfig, it will return
     * false.
     *
     * @returns A boolean that represents if the consent is accepted or not.
     */
    get activeConsent() {
        return this.consent ?? this.snippetConfig?.consent ?? false;
    }
    get config() {
        return {
            queryTaggingDebounceMs: this.queryTaggingDebounceMs,
            sessionTTLMs: this.sessionTTLMs,
            clickedResultStorageTTLMs: this.clickedResultStorageTTLMs,
            clickedResultStorageKey: this.clickedResultStorageKey
        };
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    render() { }
    /**
     * To emit the event that PDP is loaded  just when the snippet config includes a product id.
     */
    created() {
        this.emitEvents();
    }
    /**
     * Emits the {@link TaggingXEvents.PDPIsLoaded} XEvent if the snippet config contains
     * a product id.
     *
     * @internal
     */
    emitEvents() {
        if (this.snippetConfig?.productId) {
            this.$x.emit('PDPIsLoaded', this.snippetConfig.productId);
        }
    }
};
__decorate([
    Prop({ default: 30000 })
], Tagging.prototype, "clickedResultStorageTTLMs", void 0);
__decorate([
    Prop({ default: 'url' })
], Tagging.prototype, "clickedResultStorageKey", void 0);
__decorate([
    Inject('snippetConfig')
], Tagging.prototype, "snippetConfig", void 0);
__decorate([
    Prop()
], Tagging.prototype, "sessionTTLMs", void 0);
__decorate([
    Prop()
], Tagging.prototype, "queryTaggingDebounceMs", void 0);
__decorate([
    Prop()
], Tagging.prototype, "consent", void 0);
__decorate([
    XEmit('ConsentProvided')
], Tagging.prototype, "activeConsent", null);
__decorate([
    XEmit('TaggingConfigProvided')
], Tagging.prototype, "config", null);
Tagging = __decorate([
    Component({
        mixins: [xComponentMixin(taggingXModule)]
    })
], Tagging);
var script = Tagging;

export { script as default };
