// @empathyco/x-adapter/dist/cjs/empathy/mappers/request/empathy-section-recommendations-request.mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathySectionRecommendationsRequestMapper = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathySectionRecommendationsRequestMapper = /** @class */ (function () {
    function EmpathySectionRecommendationsRequestMapper() {
    }
    EmpathySectionRecommendationsRequestMapper.prototype.map = function (_a, request) {
        var section = _a.section, origin = _a.origin, rows = _a.rows, start = _a.start;
        return Object.assign(request, {
            origin: origin,
            section: section,
            rows: rows,
            start: start
        });
    };
    EmpathySectionRecommendationsRequestMapper = tslib_1.__decorate([
        (0, inversify_1.injectable)()
    ], EmpathySectionRecommendationsRequestMapper);
    return EmpathySectionRecommendationsRequestMapper;
}());
exports.EmpathySectionRecommendationsRequestMapper = EmpathySectionRecommendationsRequestMapper;
