// @empathyco/x-components/js/x-modules/device/components/device-detector.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import GlobalEvents from 'vue-global-events';
import { Prop, Watch, Component } from 'vue-property-decorator';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { XEmit } from '../../../components/decorators/bus.decorators.js';
import { throttle } from '../../../utils/throttle.js';
import { deviceXModule } from '../x-module.js';

/**
 * This component helps detecting or setting a device, that can be used later to create
 * different layouts optimized for different devices. This detected device is available under
 * the {@link XComponentAliasAPI.device} property.
 *
 * @public
 */
let DeviceDetector = class DeviceDetector extends Vue {
    constructor() {
        super(...arguments);
        /**
         * The width in pixels of the window where the app is being rendered.
         *
         * @internal
         */
        this.windowWidthPx = null;
        /**
         * Throttled version of {@link DeviceDetector.storeWindowWidth} function.
         *
         * @internal
         */
        // eslint-disable-next-line @typescript-eslint/unbound-method
        this.throttledStoreWindowWidth = this.storeWindowWidth;
    }
    /**
     * List of each of the entries of the breakpoints sorted from the smallest to the biggest
     * max width.
     *
     * @returns A list of the breakpoints sorted by its max width in ascending order.
     *
     * @internal
     */
    get sortedBreakpoints() {
        return Object.entries(this.breakpoints).sort(([, aWidth], [, bWidth]) => aWidth - bWidth);
    }
    /**
     * The device detected by this component, or the value provided in {@link DeviceDetector.force}
     * prop.
     *
     * @returns The detected device, or the value provided in {@link DeviceDetector.force}
     * prop.
     *
     * @internal
     */
    get detectedDevice() {
        if (this.force) {
            return this.force;
        }
        else if (this.windowWidthPx === null) {
            return null;
        }
        else {
            return (this.sortedBreakpoints.find(([, width]) => this.windowWidthPx <= width)?.[0] ?? null);
        }
    }
    /**
     * Stores the window width in {@link DeviceDetector.windowWidthPx}.
     *
     * @internal
     */
    storeWindowWidth() {
        this.windowWidthPx = window.innerWidth;
    }
    /**
     * Updates {@link DeviceDetector.throttledStoreWindowWidth} with the throttle time at
     * {@link DeviceDetector.throttleMs}.
     *
     * @param throttleMs - The new duration in milliseconds for the throttle.
     *
     * @internal
     */
    updateThrottledStoreWindowWidth(throttleMs) {
        // eslint-disable-next-line @typescript-eslint/unbound-method
        this.throttledStoreWindowWidth = throttle(this.storeWindowWidth, throttleMs);
    }
    /**
     * Initialises the store window width.
     *
     * @remarks This is done this way to ensure SSR compatibility.
     *
     * @internal
     */
    mounted() {
        this.storeWindowWidth();
    }
};
__decorate([
    Prop({ default: () => ({}) })
], DeviceDetector.prototype, "breakpoints", void 0);
__decorate([
    Prop()
], DeviceDetector.prototype, "force", void 0);
__decorate([
    Prop({ default: 100 })
], DeviceDetector.prototype, "throttleMs", void 0);
__decorate([
    XEmit('DeviceProvided')
], DeviceDetector.prototype, "detectedDevice", null);
__decorate([
    Watch('throttleMs', { immediate: true })
], DeviceDetector.prototype, "updateThrottledStoreWindowWidth", null);
DeviceDetector = __decorate([
    Component({
        mixins: [xComponentMixin(deviceXModule)],
        components: {
            GlobalEvents
        }
    })
], DeviceDetector);
var script = DeviceDetector;

export { script as default };
