// @empathyco/x-components/js/x-modules/facets/service/facets.service.js
import { isFacetFilter } from '@empathyco/x-types';
import '../../../plugins/x-bus.js';
import { XPlugin } from '../../../plugins/x-plugin.js';
import { isArrayEmpty, arrayToObject, groupItemsBy } from '../../../utils/array.js';
import '../../../utils/storage.js';
import { FilterEntityFactory } from '../entities/filter-entity.factory.js';

/**
 * Default implementation for the {@link FacetsService}.
 *
 * @public
 */
class DefaultFacetsService {
    constructor(filterEntityFactory = FilterEntityFactory.instance) {
        this.filterEntityFactory = filterEntityFactory;
    }
    /**
     * The {@link https://vuex.vuejs.org/ | Vuex} store to use in the service.
     *
     * @returns The store.
     * @internal
     */
    get store() {
        return XPlugin.store;
    }
    setFacets(facetsGroup) {
        const newFilters = this.updateStore(facetsGroup);
        /* Ensures that filters are selected with valid values. For example, you can't set a single
         select facet with 2 or more selected filters */
        this.updateFiltersSelectedState(newFilters);
    }
    updateFacets(facetsGroup) {
        const selectedFilters = this.getSelectedFilters();
        const newFilters = this.updateStore(facetsGroup);
        this.updateFiltersSelectedState(newFilters, selectedFilters);
    }
    updatePreselectedFilters(filters) {
        this.setPreselectedFilter(filters);
    }
    selectPreselectedFilters() {
        this.select(this.store.state.x.facets.preselectedFilters);
    }
    clearFilters(facetIds) {
        this.getSelectedFilters()
            .filter(filter => !facetIds || (isFacetFilter(filter) && facetIds.includes(filter.facetId)))
            .forEach(this.deselect.bind(this));
    }
    deselect(filter) {
        this.getFilterEntity(filter).deselect(filter);
    }
    select(filterOrFilters) {
        const filters = Array.isArray(filterOrFilters) ? filterOrFilters : [filterOrFilters];
        filters.forEach(filter => this.getFilterEntity(filter).select(filter));
    }
    toggle(filter) {
        if (filter.selected) {
            this.deselect(filter);
        }
        else {
            this.select(filter);
        }
    }
    /**
     * Creates an entity from a filter DTO.
     *
     * @param filter - The filter to create an entity from.
     * @returns The filter entity.
     * @internal
     */
    getFilterEntity(filter) {
        return this.filterEntityFactory.getFilterEntity(this.store, filter);
    }
    /**
     * Sets in the store the Facets, the Filters and the FacetsGroup, without applying any logic
     * to the selected state.
     *
     * @param facetsGroup - The {@link FacetsGroup} to set into the store state.
     * @returns An array with the new filters.
     * @internal
     */
    updateStore(facetsGroup) {
        this.removeGroupFacets(facetsGroup.id);
        this.removeGroupFilters(facetsGroup.id);
        facetsGroup.facets.forEach(facet => {
            this.setFacetGroup({ facetId: facet.id, groupId: facetsGroup.id });
            this.setFacet(facet);
        });
        const newFilters = facetsGroup.facets.flatMap(facet => facet.filters);
        this.setFilters(newFilters);
        return newFilters;
    }
    /**
     * Retrieves the selected filters from the store.
     *
     * @returns The list of selected filters of the store.
     * @internal
     */
    getSelectedFilters() {
        return this.store.getters['x/facets/selectedFilters'];
    }
    /**
     * Changes the filters selection state to match the store.
     *
     * @param newFilters - The list of filters to save.
     * @param previousFilters - (Optional) The list of old filters, used to set the `newFilters`
     * selected state.
     */
    updateFiltersSelectedState(newFilters, previousFilters) {
        if (!isArrayEmpty(newFilters)) {
            const newStateFiltersMap = arrayToObject(previousFilters ?? newFilters, 'id');
            newFilters.forEach(filter => {
                const filterEntity = this.getFilterEntity(filter);
                if (newStateFiltersMap[filter.id]?.selected) {
                    filterEntity.select(filter);
                }
                else {
                    filterEntity.deselect(filter);
                }
            });
        }
    }
    /**
     * Removes the filters that belong to the given group.
     *
     * @param groupId - The id of the group from whom remove the filters that are in the store.
     *
     * @returns The removed filters.
     *
     * @internal
     */
    removeGroupFilters(groupId) {
        const filtersToRemove = groupItemsBy(Object.values(this.store.state.x.facets.filters), filter => isFacetFilter(filter)
            ? this.store.state.x.facets.groups[filter.facetId]
            : '__unknown-group__')[groupId] ?? [];
        this.removeFilters(filtersToRemove);
        return filtersToRemove;
    }
    /**
     * Removes the facets that belong to the given group.
     *
     * @param groupId - The id of the group from whom remove the facets that are in the store.
     * @returns The removed facets.
     * @internal
     */
    removeGroupFacets(groupId) {
        const facetsToRemove = Object.values(this.store.state.x.facets.facets).filter(facet => this.store.state.x.facets.groups[facet.id] === groupId);
        facetsToRemove.forEach(this.removeFacet.bind(this));
        return facetsToRemove;
    }
    /**
     * Sets the group that a facet belongs to.
     *
     * @param facetGroup - The id of the facet, and the group it belongs to.
     * @internal
     */
    setFacetGroup(facetGroup) {
        this.store.commit('x/facets/setFacetGroup', facetGroup);
    }
    /**
     * Sets the Facet to the store facets record.
     *
     * @param facet - The facet to store.
     *
     * @internal
     */
    setFacet({ filters, ...restFacet }) {
        this.store.commit('x/facets/setFacet', restFacet);
    }
    /**
     * Removes a facet from the store.
     *
     * @param facet - The facet to remove.
     * @internal
     */
    removeFacet(facet) {
        this.store.commit('x/facets/removeFacet', facet);
    }
    /**
     * Saves a list of filters to the store without any state change logic applied.
     *
     * @param filters - The filters to save.
     * @internal
     */
    setFilters(filters) {
        this.store.commit('x/facets/setFilters', filters);
    }
    /**
     * Saves a list of preselected filters to the store without any state change logic applied.
     *
     * @param filters - The filters to save.
     * @internal
     */
    setPreselectedFilter(filters) {
        this.store.commit('x/facets/setPreselectedFilters', filters);
    }
    /**
     * Removes a list of filters from the store.
     *
     * @param filters - The filters to remove.
     * @internal
     */
    removeFilters(filters) {
        this.store.commit('x/facets/removeFilters', filters);
    }
}
/**
 * Global instance of the {@link FacetsService}.
 */
DefaultFacetsService.instance = new DefaultFacetsService();

export { DefaultFacetsService };
