// @empathyco/x-components/js/x-modules/popular-searches/store/module.js
import { setStatus } from '../../../store/utils/status-store.utils.js';
import { cancelFetchAndSaveSuggestions, fetchAndSaveSuggestions } from './actions/fetch-and-save-suggestions.action.js';
import { fetchSuggestions } from './actions/fetch-suggestions.action.js';
import { popularSearches } from './getters/popular-searches.getter.js';
import { request } from './getters/request.getter.js';

/**
 * {@link XStoreModule} For the next-queries module.
 *
 * @internal
 */
const popularSearchesXStoreModule = {
    state: () => ({
        popularSearches: [],
        searchedQueries: [],
        status: 'initial',
        config: {
            hideSessionQueries: true,
            maxItemsToRequest: 20,
            showExtraSuggestionWithoutFilter: false
        },
        params: {}
    }),
    getters: {
        request,
        popularSearches
    },
    mutations: {
        setSuggestions(state, suggestions) {
            state.popularSearches = suggestions;
        },
        setSearchedQueries(state, searchedQueries) {
            state.searchedQueries = searchedQueries;
        },
        setStatus,
        setParams(state, params) {
            state.params = params;
        }
    },
    actions: {
        cancelFetchAndSaveSuggestions,
        fetchSuggestions,
        fetchAndSaveSuggestions
    }
};

export { popularSearchesXStoreModule };
