// @empathyco/x-components/js/x-modules/identifier-results/store/module.js
import { setQuery } from '../../../store/utils/query.utils.js';
import { setStatus } from '../../../store/utils/status-store.utils.js';
import { cancelFetchAndSaveIdentifierResults, fetchAndSaveIdentifierResults } from './actions/fetch-and-save-identifier-results.action.js';
import { fetchIdentifierResults } from './actions/fetch-identifier-results.action.js';
import { saveOrigin } from './actions/save-origin.action.js';
import { saveQuery } from './actions/save-query.action.js';
import { identifierDetectionRegexp } from './getters/identifier-detection-regexp.getter.js';
import { identifierHighlightRegexp } from './getters/identifier-highlight-regexp.getter.js';
import { identifierResultsRequest } from './getters/identifier-results-request.getter.js';

/**
 * {@link XStoreModule} For the identifier results module.
 *
 * @internal
 */
const identifierResultsXStoreModule = {
    state: () => ({
        config: {
            debounceInMs: 600,
            maxItemsToRequest: 10,
            identifierDetectionRegexp: '^[0-9]{2,}$',
            separatorChars: '-/ '
        },
        identifierResults: [],
        origin: null,
        query: '',
        params: {},
        status: 'initial'
    }),
    getters: {
        identifierResultsRequest,
        identifierHighlightRegexp,
        identifierDetectionRegexp
    },
    mutations: {
        setIdentifierResults(state, identifierResults) {
            state.identifierResults = identifierResults;
        },
        setOrigin(state, origin = null) {
            state.origin = origin;
        },
        setParams(state, params) {
            state.params = params;
        },
        setQuery,
        setStatus
    },
    actions: {
        cancelFetchAndSaveIdentifierResults,
        fetchIdentifierResults,
        fetchAndSaveIdentifierResults,
        saveOrigin,
        saveQuery
    }
};

export { identifierResultsXStoreModule };
