// @empathyco/x-components/js/plugins/x-bus.js
import { ReplaySubject } from 'rxjs';
import { map } from 'rxjs/operators';

/**
 * Default {@link XBus} implementation.
 *
 * @public
 */
class BaseXBus {
    constructor() {
        /**
         * Dictionary to store the created event emitters.
         *
         * @internal
         */
        this.emitters = {};
    }
    /**
     * Emits an event. See {@link XBus.(emit:2)}.
     *
     * @param event - Event to be emitted.
     * @param payload - Event payload.
     * @param metadata - Information of who emits the event.
     */
    emit(event, payload, metadata = { moduleName: null }) {
        // Payload is defined here as an optional argument (which is wrong), but as this
        // implementation must be used with the type XBus there is no problem
        const value = {
            eventPayload: payload,
            metadata
        };
        const emitter = this.getOrCreateEmitter(event);
        emitter.next(value);
    }
    /**
     * Retrieves an observable. See {@link XBus.(on:3)}.
     *
     * @public
     * @param event - Event to listener.
     * @param withMetadata - Option to listener with info about event emitter.
     * @returns The emitter for the event passed.
     */
    on(event, withMetadata = false) {
        return withMetadata
            ? this.getOrCreateEmitter(event)
            : this.getOrCreateEmitter(event).pipe(map(value => value.eventPayload));
    }
    /**
     * Retrieves an event emitter for a given event.
     *
     * @param event - The event to retrieve the emitter for.
     * @returns The emitter for the event passed.
     * @remarks The emitter is implemented with a
     * {@Link https://www.learnrxjs.io/learn-rxjs/subjects/replaysubject | ReplaySubject} to allow any
     * new subscriber receive the last emitted value.
     * @internal
     */
    getOrCreateEmitter(event) {
        // TODO I Don't get why the types are not working here
        return (this.emitters[event] ??
            (this.emitters[event] = new ReplaySubject(1)));
    }
}
/** @internal The bus instance. Will be replaced by injection */
const bus = new BaseXBus();
// TODO Remove this instantiation and replace with injection where used

export { BaseXBus, bus };
