// @empathyco/x-components/js/plugins/x-plugin.js
import { deepMerge } from '@empathyco/x-deep-merge';
import { forEach } from '@empathyco/x-utils';
import Vuex, { Store } from 'vuex';
import { FILTERS_REGISTRY } from '../filters/filters.registry.js';
import { cleanGettersProxyCache } from '../store/utils/getters-proxy.utils.js';
import { RootXStoreModule } from '../store/x.module.js';
import { bus } from './x-bus.js';
import { registerStoreEmitters } from './x-emitters.js';
import { createXComponentAPIMixin } from './x-plugin.mixin.js';
import { assertXPluginOptionsAreValid } from './x-plugin.utils.js';

/**
 * Vue plugin that initializes the properties needed by the x-components, and exposes the events bus
 * and the adapter after it has been installed.
 *
 * @public
 */
class XPlugin {
    /**
     * Creates a new instance of the XPlugin with the given bus passed as parameter.
     *
     * @param bus - The {@link XBus} implementation to use for the plugin.
     *
     * @public
     */
    constructor(bus) {
        /**
         * Set of the already installed {@link XModule | XModules} to avoid re-registering them.
         *
         * @internal
         */
        this.installedXModules = new Set();
        /**
         * True if the plugin has been installed in a Vue instance, in this case
         * {@link XModule |Xmodules} will be installed immediately. False otherwise, in this case
         * {@link XModule | XModules} will be installed lazily when the {@link XPlugin#install} method
         * is called.
         *
         * @internal
         */
        this.isInstalled = false;
        this.bus = bus;
    }
    /**
     * {@link @empathyco/x-adapter#SearchAdapter | SearchAdapter} Is the middleware between
     * the components and our API where data can be mapped to client needs.
     * This property is only available after installing the plugin.
     *
     * @returns The installed adapter.
     * @throws If this property is accessed before calling `Vue.use(xPlugin)`.
     * @public
     */
    static get adapter() {
        return this.getInstance().adapter;
    }
    /**
     * Exposed {@link XBus}, so any kind of application can subscribe to {@link XEventsTypes}
     * without having to pass through a component.
     * This property is only available after installing the plugin.
     *
     * @returns The installed bus.
     * @throws If this property is accessed before calling `Vue.use(xPlugin)`.
     * @public
     */
    static get bus() {
        return this.getInstance().bus;
    }
    /**
     * {@link https://vuex.vuejs.org | Vuex Store} Is the place where all shared data
     * is saved.
     *
     * @returns The installed store.
     * @throws If this property is accessed before calling `Vue.use(xPlugin)`.
     * @public
     */
    static get store() {
        return this.getInstance().store;
    }
    /**
     * Safely retrieves the installed instance of the XPlugin.
     *
     * @returns The installed instance of the XPlugin.
     * @throws If this method is called before calling `Vue.use(xPlugin)`.
     * @internal
     */
    static getInstance() {
        if (!this.instance) {
            throw Error("XPlugin must be installed before accessing it's API.");
        }
        return this.instance;
    }
    /**
     * If the plugin has already been installed, it immediately registers a {@link XModule}. If it
     * has not been installed yet, it stores the module in a list until the plugin is installed.
     *
     * @param xModule - The module to register.
     *
     * @public
     */
    static registerXModule(xModule) {
        if (this.instance) {
            this.instance.registerXModule(xModule);
        }
        else {
            this.lazyRegisterXModule(xModule);
        }
    }
    /**
     * Utility method for resetting the installed instance of the plugin.
     *
     * @remarks Use only for testing.
     *
     * @internal
     */
    static resetInstance() {
        cleanGettersProxyCache();
        this.instance = undefined;
    }
    /**
     * Stores the {@link XModule} in a dictionary, so it can be registered later in the install
     * process.
     *
     * @param xModule - The module to register.
     *
     * @internal
     */
    static lazyRegisterXModule(xModule) {
        this.pendingXModules[xModule.name] = xModule;
    }
    /**
     * Installs the plugin into the Vue instance.
     *
     * @param vue - The GlobalVue object.
     * @param options - The options to install this plugin with.
     * @throws If the XPlugin has already been installed, or the options are not valid.
     *
     * @internal
     */
    install(vue, options) {
        if (this.isInstalled) {
            throw new Error('XPlugin has already been installed');
        }
        assertXPluginOptionsAreValid(options);
        XPlugin.instance = this;
        this.vue = vue;
        this.options = options;
        this.adapter = options.adapter;
        this.createAdapterConfigChangedListener();
        this.registerStore();
        this.applyMixins();
        this.registerFilters();
        this.registerInitialModules();
        this.registerPendingXModules();
        this.isInstalled = true;
    }
    /**
     * Performs the registration of a {@link XModule}.
     *
     * @param xModule - The module to register.
     *
     * @internal
     */
    registerXModule(xModule) {
        if (!this.installedXModules.has(xModule.name)) {
            const customizedXModule = this.customizeXModule(xModule);
            this.registerStoreModule(customizedXModule);
            this.registerStoreEmitters(customizedXModule);
            // The wiring must be registered after the store emitters
            // to allow lazy loaded modules work properly.
            this.registerWiring(customizedXModule);
            this.installedXModules.add(xModule.name);
        }
    }
    /**
     * Performs a customization of a {@link XModule} using the XPlugin public and private options.
     *
     * @param xModule - The module to customize.
     * @returns The customized xModule.
     *
     * @internal
     */
    customizeXModule({ name, wiring, storeModule, storeEmitters, ...restXModule }) {
        const { wiring: wiringOptions, config } = this.options.xModules?.[name] ?? {};
        const { storeModule: storeModuleOptions, storeEmitters: emittersOptions } = this.options.__PRIVATE__xModules?.[name] ?? {};
        return {
            name,
            wiring: wiringOptions ? deepMerge({}, wiring, wiringOptions) : wiring,
            storeModule: this.customizeStoreModule(storeModule, storeModuleOptions ?? {}, config),
            storeEmitters: emittersOptions
                ? deepMerge({}, storeEmitters, emittersOptions)
                : storeEmitters,
            ...restXModule
        };
    }
    /**
     * Performs the registration of the wiring, retrieving the observable for each event, and
     * executing each wire.
     *
     * @param xModule - The {@link XModule} to register its wiring.
     *
     * @internal
     */
    registerWiring({ wiring }) {
        forEach(wiring, (event, wires) => {
            // Obtain the observable
            const observable = this.bus.on(event, true);
            // Register event wires
            forEach(wires, (_, wire) => {
                wire(observable, this.store, this.bus.on.bind(this.bus));
            });
        });
    }
    /**
     * Registers a {@link https://vuex.vuejs.org/ | Vuex} store module under the 'x' module.
     *
     * @param xModule - The {@link XModule} to register its Store Module.
     *
     * @internal
     */
    registerStoreModule({ name, storeModule }) {
        storeModule.namespaced = true;
        this.store.registerModule(['x', name], storeModule);
    }
    /**
     * Overrides a {@link https://vuex.vuejs.org/ | Vuex} store module definition.
     *
     * Priority of configuration merging.
     * 1st {@link XPluginOptions.xModules | xModules XPlugin option}.
     * 2nd {@link XPluginOptions.__PRIVATE__xModules | Private xModules XPlugin option}.
     * 3rd {@link XStoreModule.state | Default state of the xModule}.
     *
     * @param defaultModule - The default store module to override.
     * @param moduleOptions - The state, actions, mutations and getters to override the defaultModule.
     * @param configOptions - The state config to override the moduleOptions.
     * @returns The {@link XStoreModule} customized.
     *
     * @internal
     */
    customizeStoreModule({ state: defaultState, ...actionsGettersMutations }, { state: xModuleState, ...newActionsGettersMutations }, configOptions) {
        const configOptionsObject = configOptions ? { config: configOptions } : {};
        const customizedModule = deepMerge({}, actionsGettersMutations, newActionsGettersMutations);
        customizedModule.state = deepMerge(defaultState(), xModuleState, configOptionsObject);
        return customizedModule;
    }
    /**
     * Registers the store emitters, making them emit the event when the part of the state selected
     * changes.
     *
     * @param xModule - The {@link XModule} to register its Store Emitters.
     *
     * @internal
     */
    registerStoreEmitters(xModule) {
        registerStoreEmitters(xModule, this.bus, this.store);
    }
    /**
     * Registers the {@link https://vuex.vuejs.org/ | Vuex} store. If the store has not been passed
     * through the {@link XPluginOptions} object, it creates one, and injects it in the Vue
     * prototype. Then it register an x module in the store, to safe scope all the
     * {@link XModule | XModules} dynamically installed.
     *
     * @internal
     */
    registerStore() {
        this.vue.use(Vuex); // We can safely install Vuex because if it is already installed Vue
        // will simply ignore it
        this.store =
            this.options.store ??
                new Store({
                    strict: process.env.NODE_ENV !== 'production'
                });
        if (!this.options.store) {
            this.vue.prototype.$store = this.store;
        }
        this.store.registerModule('x', RootXStoreModule);
    }
    /**
     * Applies the {@link createXComponentAPIMixin} mixin in the global Vue.
     *
     * @internal
     */
    applyMixins() {
        this.vue.mixin(createXComponentAPIMixin(this.bus));
    }
    /**
     * Registers the initial {@link XModule | XModules} during the {@link XPlugin} installation.
     *
     * @internal
     */
    registerInitialModules() {
        this.options.initialXModules?.forEach(xModule => {
            this.registerXModule(xModule);
        });
    }
    /**
     * Registers the pending {@link XModule | XModules}, that requested to be registered before the
     * installation of the plugin.
     *
     * @internal
     */
    registerPendingXModules() {
        forEach(XPlugin.pendingXModules, (_, xModule) => {
            this.registerXModule(xModule);
        });
        XPlugin.pendingXModules = {};
    }
    /**
     * If the received adapter supports it, it registers a listener to emit the
     * {@link XEventsTypes.AdapterConfigChanged} event whenever the config of it changes.
     *
     * @internal
     */
    createAdapterConfigChangedListener() {
        this.options.adapter.addConfigChangedListener?.(newAdapterConfig => {
            this.bus.emit('AdapterConfigChanged', newAdapterConfig);
        });
    }
    /**
     * Registers filters globally.
     *
     * @internal
     */
    registerFilters() {
        forEach(FILTERS_REGISTRY, (filterName, filterFunction) => this.vue.filter(filterName, filterFunction));
    }
}
/**
 * Record of modules that have been tried to be installed before the installation of the plugin.
 *
 * @internal
 */
XPlugin.pendingXModules = {};
/**
 * Vue plugin that modifies each component instance, extending them with the
 * {@link XComponentAPI | X Component API }.
 *
 * @example
 * Minimal installation example. A search adapter is needed for the plugin to work, and connect to
 * the API.
 * ```typescript
 * const adapter = new EmpathyAdapterBuilder()
 *  .withConfiguration({instance: 'my-instance-id'})
 *  .build();
 * Vue.use(xPlugin, { adapter });
 * ```
 *
 * @example
 * If you are using {@link https://vuex.vuejs.org/ | Vuex} in your project you must install its
 *   plugin, and instantiate an store before installing the XPlugin:
 * ```typescript
 * Vue.use(Vuex);
 * const store = new Store({ ... });
 * Vue.use(xPlugin, { adapter, store });
 * ```
 * @public
 */
const xPlugin = new XPlugin(bus);

export { XPlugin, xPlugin };
