// @empathyco/x-components/js/plugins/x-plugin.alias.js
import { getGetterPath } from './x-plugin.utils.js';

/**
 * Creates an object containing the alias part of {@link XComponentAPI}.
 *
 * @param component - The component with the store from which retrieve the data.
 * @returns An object containing the alias part of the {@link XComponentAPI}.
 *
 * @internal
 */
function getAliasAPI(component) {
    const queryModules = [
        'searchBox',
        'nextQueries',
        'querySuggestions',
        'relatedTags',
        'search'
    ];
    const statusModules = [
        'identifierResults',
        'nextQueries',
        'popularSearches',
        'querySuggestions',
        'recommendations',
        'relatedTags',
        'search'
    ];
    const query = queryModules.reduce((acc, moduleName) => {
        return Object.defineProperty(acc, moduleName, {
            get() {
                return component.$store.state.x[moduleName]?.query ?? '';
            },
            enumerable: true
        });
    }, {});
    const status = statusModules.reduce((acc, moduleName) => {
        return Object.defineProperty(acc, moduleName, {
            get() {
                return component.$store.state.x[moduleName]?.status;
            },
            enumerable: true
        });
    }, {});
    return {
        query,
        status,
        get device() {
            return component.$store.state.x.device?.name ?? null;
        },
        get facets() {
            return component.$store.getters[getGetterPath('facets', 'facets')] ?? {};
        },
        get historyQueries() {
            return component.$store.getters[getGetterPath('historyQueries', 'historyQueries')] ?? [];
        },
        get identifierResults() {
            return component.$store.state.x.identifierResults?.identifierResults ?? [];
        },
        get isEmpathizeOpen() {
            return component.$store.state.x.empathize?.isOpen ?? false;
        },
        get nextQueries() {
            return component.$store.getters[getGetterPath('nextQueries', 'nextQueries')] ?? [];
        },
        get noResults() {
            return !this.totalResults && !!this.query.search && this.status.search === 'success';
        },
        get partialResults() {
            return component.$store.state.x.search?.partialResults ?? [];
        },
        get popularSearches() {
            return component.$store.state.x.popularSearches?.popularSearches ?? [];
        },
        get querySuggestions() {
            return component.$store.state.x.querySuggestions?.suggestions ?? [];
        },
        get recommendations() {
            return component.$store.state.x.recommendations?.recommendations ?? [];
        },
        get redirections() {
            return component.$store.state.x.search?.redirections ?? [];
        },
        get relatedTags() {
            return component.$store.getters[getGetterPath('relatedTags', 'relatedTags')] ?? [];
        },
        get results() {
            return component.$store.state.x.search?.results ?? [];
        },
        get scroll() {
            return component.$store.state.x.scroll?.data ?? {};
        },
        get selectedFilters() {
            return component.$store.getters[getGetterPath('facets', 'selectedFilters')] ?? [];
        },
        get selectedRelatedTags() {
            return component.$store.state.x.relatedTags?.selectedRelatedTags ?? [];
        },
        get spellcheckedQuery() {
            return component.$store.state.x.search?.spellcheckedQuery ?? null;
        },
        get totalResults() {
            return component.$store.state.x.search?.totalResults ?? 0;
        },
        get selectedSort() {
            return component.$store.state.x.search?.sort ?? '';
        }
    };
}

export { getAliasAPI };
