// @empathyco/x-components/js/plugins/x-plugin.mixin.js
import { getXComponentXModuleName, isXComponent } from '../components/x-component.utils.js';
import { getAliasAPI } from './x-plugin.alias.js';

/**
 * Vue global mixin that adds a `$x` object to every component with the {@link XComponentAPI}.
 *
 * @remarks It adds an injection property `origin` which value is included in the metadata of each
 * event emitted with `$x.emit`.
 *
 * @param bus - The {@link XBus} to use inside the components for emitting events.
 * @returns Mixin options which registers the component as X-Component and the $x.
 * @internal
 */
const createXComponentAPIMixin = (bus) => ({
    inject: {
        $location: {
            from: 'location',
            default: undefined
        }
    },
    beforeCreate() {
        this.xComponent = getRootXComponent(this);
        const aliasAPI = getAliasAPI(this);
        const busAPI = getBusAPI(bus, this);
        this.$x = Object.assign(aliasAPI, busAPI);
    }
});
/**
 * Creates an object containing the API related to the {@link XBus}.
 *
 * @param bus - The global {@link XBus}.
 * @param component - The component instance.
 *
 * @returns An object containing the {@link XComponentBusAPI}.
 * @internal
 */
function getBusAPI(bus, component) {
    return {
        emit: (event, payload, metadata = {}) => {
            bus.emit(event, payload, createWireMetadata(component, metadata));
            component.xComponent?.$emit(event, payload);
        },
        on: bus.on.bind(bus)
    };
}
/**
 * Creates a wire metadata object based on the component and the provided metadata.
 *
 * @param component - The component this metadata belongs to.
 * @param metadata - Additional metadata emitted by the component.
 * @returns A {@link WireMetadata} object.
 * @internal
 */
function createWireMetadata(component, metadata) {
    return {
        moduleName: getXComponentXModuleName(component.xComponent),
        location: component.$location,
        ...metadata
    };
}
/**
 * Given a component, finds the root XComponent in the ancestors hierarchy.
 *
 * @param component - The component to find its root XComponent.
 * @returns The root XComponent or undefined if it has not.
 * @public
 */
function getRootXComponent(component) {
    let xComponent;
    while (component !== undefined && component !== null) {
        if (isXComponent(component)) {
            xComponent = component;
        }
        component = component.$parent;
    }
    return xComponent;
}

export { createXComponentAPIMixin, getBusAPI, getRootXComponent };
