// @empathyco/x-components/js/x-modules/search/store/emitters.js
import { createStoreEmitters } from '../../../store/utils/store-emitters.utils.js';
import { isStringEmpty } from '../../../utils/string.js';
import { searchXStoreModule } from './module.js';

/**
 * {@link StoreEmitters} For the search module.
 *
 * @internal
 */
const searchEmitters = createStoreEmitters(searchXStoreModule, {
    FacetsChanged: {
        selector: state => state.facets,
        filter(newValue, oldValue) {
            return newValue.length !== 0 || oldValue.length !== 0;
        }
    },
    PageChanged: state => state.page,
    ResultsChanged: state => state.results,
    SearchRequestChanged: (_, getters) => getters.request,
    SearchRequestUpdated: {
        selector: (_, getters) => getters.request,
        filter(newValue, oldValue) {
            return !!newValue && !!oldValue;
        }
    },
    SearchTaggingChanged: {
        selector: state => state.queryTagging,
        filter: ({ url }) => !isStringEmpty(url)
    },
    SpellcheckChanged: state => state.spellcheckedQuery,
    SortChanged: state => state.sort
});

export { searchEmitters };
