// @empathyco/x-adapter/dist/cjs/empathy/mappers/response/filters/empathy-number-range-filter.mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyNumberRangeFilterMapper = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
/**
 * Number Range Empathy filter mapper.
 *
 * @public
 */
var EmpathyNumberRangeFilterMapper = /** @class */ (function () {
    function EmpathyNumberRangeFilterMapper() {
    }
    EmpathyNumberRangeFilterMapper.prototype.map = function (rawFilter, filter, context) {
        var _a = rawFilter.value.split(':'), min = _a[0], max = _a[1];
        return Object.assign(filter, {
            modelName: 'NumberRangeFilter',
            range: {
                min: Number.parseFloat(min) || null,
                max: Number.parseFloat(max) || null
            }
        });
    };
    EmpathyNumberRangeFilterMapper = tslib_1.__decorate([
        (0, inversify_1.injectable)()
    ], EmpathyNumberRangeFilterMapper);
    return EmpathyNumberRangeFilterMapper;
}());
exports.EmpathyNumberRangeFilterMapper = EmpathyNumberRangeFilterMapper;
