// @empathyco/x-components/js/x-installer/api/base-api.js
/**
 * Default implementation for {@link XAPI}.
 *
 * @public
 */
class BaseXAPI {
    constructor() {
        /**
         * Flag to check if the initialization was already done.
         *
         * @internal
         */
        this.isXInitialized = false;
    }
    /**
     * Tracks that a product was added to cart from PDP.
     *
     * @param productId - The product id that was added to cart.
     */
    addProductToCart(productId) {
        this.bus?.emit('UserClickedPDPAddToCart', productId);
    }
    /**
     * Setter for the {@link XBus}.
     *
     * @param bus - The {@link XBus} received to emit events.
     *
     * @internal
     */
    setBus(bus) {
        this.bus = bus;
    }
    /**
     * Setter for the callback to call in the init method.
     *
     * @param initCallback - The callback to call.
     */
    setInitCallback(initCallback) {
        this.initCallback = initCallback;
    }
    /**
     * Setter for the callback to modify the snippet config.
     *
     * @param snippetCallback - The callback to call.
     *
     * @internal
     */
    setSnippetConfigCallback(snippetCallback) {
        this.snippetCallback = snippetCallback;
    }
    /**
     * Sets or updates the snippet config.
     *
     * @param config - A part or all the snippet config.
     *
     * @public
     */
    setSnippetConfig(config) {
        this?.snippetCallback(config);
    }
    /**
     * Searches the query parameter as user query.
     *
     * @param query - Query to be searched.
     *
     * @public
     */
    search(query) {
        if (query) {
            this.bus?.emit('UserAcceptedAQuery', query);
        }
        this.bus?.emit('UserClickedOpenX');
    }
    /**
     * Initializes the Application passing the {@link SnippetConfig}.
     *
     * @param config - The config coming from the customer snippet.
     *
     * @public
     */
    init(config) {
        if (!this.isXInitialized) {
            this.isXInitialized = true;
            this?.initCallback(config);
        }
        else {
            //eslint-disable-next-line no-console
            console.warn('We know X is awesome, but you only need to initialize it once.');
        }
    }
}

export { BaseXAPI };
