// @empathyco/x-components/js/components/decorators/bus.decorators.js
import { Subscription } from 'rxjs';
import { filter } from 'rxjs/operators';
import { createDecorator } from 'vue-class-component';
import { clone } from '../../utils/clone.js';

/**
 * Creates a subscription to an {@link XEvent}, an array of {@link XEvent} or a component property (
 * reacting to its changes via a watcher) filtering out the passed metadata, if any, and
 * un-subscribes on the beforeDestroy hook.
 *
 * @remarks
 * The decorated property needs to be public for type inference to work.
 *
 * @param xEvent - The {@link XEvent}, an array of {@link XEvent} or a component property.
 * @param metadataFilteringOptions - The {@link WireMetadata} options to filter out a callback
 * execution.
 * @returns Decorator that creates a subscription to an {@link XEvent} and un-subscribes on the
 * beforeDestroy hook.
 * @public
 */
function XOn(xEvent, metadataFilteringOptions = {}) {
    return createDecorator((options, key) => {
        // eslint-disable-next-line @typescript-eslint/unbound-method
        const originalCreated = options.created;
        Object.assign(options, {
            created() {
                originalCreated?.apply(this);
                const componentCreateSubscription = createSubscription.bind(this);
                const subscriptionMetadata = {
                    event: [],
                    callback: this[key],
                    metadataFilteringOptions
                };
                let subscription;
                if (typeof xEvent === 'function') {
                    this.$watch(() => xEvent(this), newEvents => {
                        subscription?.unsubscribe();
                        subscriptionMetadata.event = newEvents;
                        subscription = componentCreateSubscription(subscriptionMetadata);
                    }, { immediate: true });
                }
                else {
                    subscriptionMetadata.event = xEvent;
                    subscription = componentCreateSubscription(subscriptionMetadata);
                }
                this.$on('hook:beforeDestroy', () => subscription.unsubscribe()); // Using Vue
                // bus to subscribe to the lifecycle hook 'beforeDestroy' instead of 'capturing' the
                // original component's 'beforeDestroy' method to override it plus calling
                // originalBeforeDestroy.apply(this) to preserve the existing original hook functionality
            }
        });
    });
}
/**
 * Create a subscription for the given events executing the passed callback and filtering out based
 * on the passed metadata options.
 *
 * @param this - The vue component.
 * @param subscriptionMetadata - The {@link SubscriptionMetadata}.
 * @returns A
 * {@link https://www.learnrxjs.io/learn-rxjs/concepts/rxjs-primer#subscription | subscription}.
 * @internal
 */
function createSubscription(subscriptionMetadata) {
    const { event, callback, metadataFilteringOptions } = subscriptionMetadata;
    const eventArray = Array.isArray(event) ? event : [event];
    const subscription = new Subscription();
    eventArray.forEach(xEvent => {
        subscription.add(this.$x
            .on(xEvent, true)
            .pipe(filter(({ metadata }) => filterMetadataOptions(metadataFilteringOptions, metadata)))
            .subscribe(({ eventPayload, metadata }) => callback(eventPayload, metadata)));
    });
    return subscription;
}
/**
 * Checks if the metadata options passed to the decorator match the metadata of the emitted event.
 *
 * @param filteringOptions - The decorator options.
 * @param eventOptions - The emitted event metadata.
 * @returns True if options are empty or match the metadata.
 * @internal
 */
function filterMetadataOptions(filteringOptions, eventOptions) {
    return Object.keys(filteringOptions)
        .filter(key => filteringOptions[key] !== undefined)
        .every(key => filteringOptions[key] === eventOptions[key]);
}
/**
 * Emits the provided event whenever the decorated property changes.
 *
 * @param xEvent - The event to emit.
 * @param watcherOptions - Options for Vue's watcher.
 * @returns Decorator that makes the component emit an event when the decorated property changes.
 * @public
 */
function XEmit(xEvent, { immediate = true, deep = false } = {}) {
    return createDecorator((options, key) => {
        // eslint-disable-next-line @typescript-eslint/unbound-method
        const originalCreated = options.created;
        options.created = function () {
            originalCreated?.apply(this);
            if (this[key] !== undefined) {
                this.$watch(key, newValue => {
                    this.$x.emit(xEvent, typeof newValue === 'object' && newValue !== null ? clone(newValue) : newValue);
                }, { immediate, deep });
            }
        };
    });
}

export { XEmit, XOn };
