// @empathyco/x-components/js/x-modules/url/wiring.js
import { namespacedWireCommit } from '../../wiring/namespaced-wires.factory.js';
import { createWiring } from '../../wiring/wiring.utils.js';

/**
 * WireCommit for {@link UrlXModule}.
 *
 * @internal
 */
const wireCommit = namespacedWireCommit('url');
/**
 * Sets the {@link RelatedTag | related tags }.
 *
 * @public
 */
const setUrlRelatedTags = wireCommit('setRelatedTags');
/**
 * Sets the query of the url module.
 *
 * @public
 */
const setUrlQuery = wireCommit('setQuery');
/**
 * Sets the page of the url module.
 *
 * @public
 */
const setUrlPage = wireCommit('setPage');
/**
 * Sets the sort of the url module.
 *
 * @public
 */
const setUrlSort = wireCommit('setSort');
/**
 * Sets the extra params of the url module.
 *
 * @public
 */
const setParams = wireCommit('setParams');
/**
 * Sets the scroll of the url module.
 *
 * @public
 */
const setUrlScroll = wireCommit('setScroll');
/**
 * Sets the initial provided extra params.
 *
 * @public
 */
const setInitialExtraParams = wireCommit('setInitialExtraParams');
/**
 * Sets the filters of the url module.
 *
 * @public
 */
const setUrlFilters = wireCommit('setFilters');
/**
 * Wiring configuration for the {@link UrlXModule | url module}.
 *
 * @internal
 */
const urlWiring = createWiring({
    ParamsLoadedFromUrl: {
        setParams
    },
    UserAcceptedAQuery: {
        setUrlQuery
    },
    UserClearedQuery: {
        setUrlQuery
    },
    SelectedRelatedTagsChanged: {
        setUrlRelatedTags
    },
    SelectedFiltersChanged: {
        setUrlFilters
    },
    PageChanged: {
        setUrlPage
    },
    SortChanged: {
        setUrlSort
    },
    ExtraParamsChanged: {
        setParams
    },
    ExtraParamsInitialized: {
        setInitialExtraParams
    },
    UserScrolledToElement: {
        setUrlScroll
    }
});

export { setInitialExtraParams, setParams, setUrlFilters, setUrlPage, setUrlQuery, setUrlRelatedTags, setUrlScroll, setUrlSort, urlWiring };
