// @empathyco/x-components/js/x-modules/search-box/components/clear-search-input.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Component } from 'vue-property-decorator';
import __vue_component__ from '../../../components/base-event-button.vue.js';
import { State } from '../../../components/decorators/store.decorators.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { searchBoxXModule } from '../x-module.js';

/**
 * This component renders a button to delete the current query.
 *
 * @remarks
 * A button that when pressed emits the {@link SearchBoxXEvents.UserPressedClearSearchBoxButton}
 * and {@link SearchBoxXEvents.UserClearedQuery} events, expressing the user intention to clear
 * the current query.
 * It also adds `x-clear-search-input--has-empty-query` as class when there is no query.
 *
 * @public
 */
let ClearSearchInput = class ClearSearchInput extends Vue {
    constructor() {
        super(...arguments);
        /**
         * The events dictionary that are going to be emitted when the button is pressed.
         *
         * @internal
         */
        this.clearSearchInputEvents = {
            UserPressedClearSearchBoxButton: undefined
        };
    }
    get isQueryEmpty() {
        return this.query.length === 0;
    }
    get dynamicClasses() {
        return {
            'x-clear-search-input--has-empty-query': this.isQueryEmpty
        };
    }
};
__decorate([
    State('searchBox', 'query')
], ClearSearchInput.prototype, "query", void 0);
ClearSearchInput = __decorate([
    Component({
        components: { BaseEventButton: __vue_component__ },
        mixins: [xComponentMixin(searchBoxXModule)]
    })
], ClearSearchInput);
var script = ClearSearchInput;

export { script as default };
