// @empathyco/x-components/js/x-modules/search-box/components/search-input.vue.js
import script from './search-input.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c("input", {
    ref: "input",
    staticClass: "x-input x-search-input",
    attrs: {
      maxlength: _vm.maxLength,
      autocomplete: "off",
      enterkeyhint: "search",
      inputmode: "search",
      type: "search",
      "data-test": "search-input",
      "aria-label": _vm.searchInputMessage,
    },
    domProps: { value: _vm.query },
    on: {
      blur: _vm.emitUserBlurredSearchBox,
      click: _vm.emitUserClickedSearchBox,
      focus: _vm.emitUserFocusedSearchBox,
      input: _vm.emitUserIsTypingAQueryEvents,
      keydown: [
        function ($event) {
          if (
            !$event.type.indexOf("key") &&
            _vm._k($event.keyCode, "enter", 13, $event.key, "Enter")
          ) {
            return null
          }
          return _vm.emitUserPressedEnterKey.apply(null, arguments)
        },
        function ($event) {
          if (
            !$event.type.indexOf("key") &&
            _vm._k($event.keyCode, "up", 38, $event.key, ["Up", "ArrowUp"]) &&
            _vm._k($event.keyCode, "down", 40, $event.key, [
              "Down",
              "ArrowDown",
            ])
          ) {
            return null
          }
          $event.preventDefault();
          return _vm.emitUserPressedArrowKey.apply(null, arguments)
        },
      ],
    },
  })
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-d46f5de0_0", { source: ".x-search-input[data-v-d46f5de0]::-webkit-search-decoration,\n.x-search-input[data-v-d46f5de0]::-webkit-search-cancel-button,\n.x-search-input[data-v-d46f5de0]::-webkit-search-results-button,\n.x-search-input[data-v-d46f5de0]::-webkit-search-results-decoration {\n  -webkit-appearance: none;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-d46f5de0";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
