// @empathyco/x-components/js/x-modules/search-box/components/search-button.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Component } from 'vue-property-decorator';
import { State } from '../../../components/decorators/store.decorators.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { searchBoxXModule } from '../x-module.js';

/**
 * This component renders a button to submit the query.
 *
 * @remarks
 * If query is not empty, it emits {@link XEventsTypes.UserAcceptedAQuery} and
 * {@link SearchBoxXEvents.UserPressedSearchButton} events with the query as payload.
 * It also adds `x-search-button--has-empty-query` as class when there is no query.
 *
 * @public
 */
let SearchButton = class SearchButton extends Vue {
    get isQueryEmpty() {
        return this.query.length === 0;
    }
    /**
     * Generates the {@link WireMetadata | event metadata} object omitting the moduleName.
     *
     * @returns The {@link WireMetadata} object omitting the moduleName.
     * @internal
     */
    createEventMetadata() {
        return {
            target: this.$el,
            feature: 'search_box'
        };
    }
    /**
     * Emits events when the button is clicked.
     *
     * @public
     */
    emitEvents() {
        if (!this.isQueryEmpty) {
            this.$x.emit('UserAcceptedAQuery', this.query, this.createEventMetadata());
            this.$x.emit('UserPressedSearchButton', this.query, this.createEventMetadata());
        }
    }
    get dynamicClasses() {
        return {
            'x-search-button--has-empty-query': this.isQueryEmpty
        };
    }
};
__decorate([
    State('searchBox', 'query')
], SearchButton.prototype, "query", void 0);
SearchButton = __decorate([
    Component({
        mixins: [xComponentMixin(searchBoxXModule)]
    })
], SearchButton);
var script = SearchButton;

export { script as default };
