// @empathyco/x-components/js/x-modules/search-box/components/search-input.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { XOn } from '../../../components/decorators/bus.decorators.js';
import { State } from '../../../components/decorators/store.decorators.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { debounce } from '../../../utils/debounce.js';
import { searchBoxXModule } from '../x-module.js';

/**
 * This component renders an input field that allows the user to type a query. It also reacts to
 * query changes through event listening.
 *
 * @public
 */
let SearchInput = class SearchInput extends Vue {
    constructor() {
        super(...arguments);
        this.searchInputMessage = 'type your query here';
    }
    /**
     * When event {@link XEventsTypes.UserReachedEmpathizeTop} or
     * {@link SearchBoxXEvents.UserPressedClearSearchBoxButton}
     * are emitted the search in put is focused.
     *
     * @internal
     */
    focusInput() {
        this.$refs.input?.focus();
    }
    /**
     * When event {@link XEventsTypes.UserAcceptedAQuery} or
     * {@link SearchBoxXEvents.UserClearedQuery} are emitted the pending debounced emit
     * {@link XEvent} `UserAcceptedAQuery` is canceled.
     *
     * @internal
     */
    cancelDebouncedUserAcceptedAQuery() {
        this.debouncedUserAcceptedAQuery?.cancel();
    }
    mounted() {
        if (this.autofocus) {
            this.focusInput();
        }
    }
    /**
     * Emits {@link XEventsTypes.UserAcceptedAQuery} event with a debounce configured in
     * `instantDebounceInMs` prop.
     *
     * @internal
     * @param query - The query that will be emitted.
     */
    emitDebouncedUserAcceptedAQuery(query) {
        if (this.instant) {
            if (!this.debouncedUserAcceptedAQuery) {
                this.debouncedUserAcceptedAQuery = debounce(this.emitUserAcceptedAQuery.bind(this), this.instantDebounceInMs);
            }
            this.debouncedUserAcceptedAQuery(query);
        }
    }
    /**
     * Generates the {@link WireMetadata | event metadata} object omitting the moduleName.
     *
     * @returns The {@link WireMetadata} object omitting the moduleName.
     * @internal
     */
    createEventMetadata() {
        return {
            target: this.$refs.input,
            feature: 'search_box'
        };
    }
    /**
     * Emits event {@link SearchBoxXEvents.UserBlurredSearchBox} when search box loses focus.
     *
     * @internal
     */
    emitUserBlurredSearchBox() {
        this.$x.emit('UserBlurredSearchBox', undefined, { target: this.$refs.input });
    }
    /**
     * Emits event {@link SearchBoxXEvents.UserClickedSearchBox} when user clicks  the search input.
     *
     * @internal
     */
    emitUserClickedSearchBox() {
        this.$x.emit('UserClickedSearchBox', undefined, { target: this.$refs.input });
    }
    /**
     * Emits event {@link SearchBoxXEvents.UserFocusedSearchBox} when search box gains focus.
     *
     * @internal
     */
    emitUserFocusedSearchBox() {
        this.$x.emit('UserFocusedSearchBox', undefined, { target: this.$refs.input });
    }
    /**
     * Emits event {@link SearchBoxXEvents.UserIsTypingAQuery} when the user typed/pasted something
     * into the search-box. Also emits event {@link SearchBoxXEvents.UserClearedQuery} when the user
     * removes all characters in the search-box.
     *
     * @internal
     */
    emitUserIsTypingAQueryEvents() {
        const query = this.$refs.input.value;
        this.$x.emit('UserIsTypingAQuery', query, { target: this.$refs.input });
        if (query.trim()) {
            this.emitDebouncedUserAcceptedAQuery(query);
        }
        else {
            this.cancelDebouncedUserAcceptedAQuery();
        }
    }
    /**
     * Emits event {@link XEventsTypes.UserPressedArrowKey} when the user pressed an arrow key.
     *
     * @param event - The keyboard event with the arrow key pressed.
     * @internal
     */
    emitUserPressedArrowKey(event) {
        this.$x.emit('UserPressedArrowKey', event.key, this.createEventMetadata());
    }
    /**
     * Emits multiple events when the user pressed the enter key.
     *
     * @remarks
     * Emitted events are:
     * {@link SearchBoxXEvents.UserPressedEnterKey}
     * {@link XEventsTypes.UserAcceptedAQuery}
     *
     * @internal
     */
    emitUserPressedEnterKey() {
        const query = this.$refs.input.value.trim();
        if (query.length > 0) {
            this.$x.emit('UserPressedEnterKey', query, this.createEventMetadata());
            this.emitUserAcceptedAQuery(query);
        }
        this.$refs.input?.blur();
    }
    /**
     * Emits {@link XEventsTypes.UserAcceptedAQuery} event.
     *
     * @remarks It is necessary in a separated method to use it as the parameter of debounce in
     * emitDebouncedUserAcceptedAQuery method.
     * @internal
     * @param query - The query that will be emitted.
     */
    emitUserAcceptedAQuery(query) {
        this.$x.emit('UserAcceptedAQuery', query, this.createEventMetadata());
    }
};
__decorate([
    Prop({ default: 64 })
], SearchInput.prototype, "maxLength", void 0);
__decorate([
    Prop({ default: true })
], SearchInput.prototype, "autofocus", void 0);
__decorate([
    Prop({ default: true })
], SearchInput.prototype, "instant", void 0);
__decorate([
    Prop({ default: 500 })
], SearchInput.prototype, "instantDebounceInMs", void 0);
__decorate([
    Prop({ default: () => ['ArrowRight'] })
], SearchInput.prototype, "autocompleteKeyboardKeys", void 0);
__decorate([
    Prop({ default: 'QuerySuggestionsChanged' })
], SearchInput.prototype, "autocompleteSuggestionsEvent", void 0);
__decorate([
    State('searchBox', 'query')
], SearchInput.prototype, "query", void 0);
__decorate([
    XOn(['UserReachedEmpathizeTop', 'UserPressedClearSearchBoxButton'])
], SearchInput.prototype, "focusInput", null);
__decorate([
    XOn(['UserAcceptedAQuery', 'UserClearedQuery'])
], SearchInput.prototype, "cancelDebouncedUserAcceptedAQuery", null);
SearchInput = __decorate([
    Component({
        mixins: [xComponentMixin(searchBoxXModule)]
    })
], SearchInput);
var script = SearchInput;

export { script as default };
