// @empathyco/x-adapter/dist/esm/empathy/empathy.adapter.js
import { __decorate, __metadata, __param } from "tslib";
import { deepMerge } from '@empathyco/x-deep-merge';
import { inject, injectable, optional } from 'inversify';
import { DEPENDENCIES } from './container/container.const';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyAdapter = /** @class */ (function () {
    function EmpathyAdapter(config, requestors, cache) {
        this.config = config;
        this.requestors = requestors;
        this.cache = cache;
        this.configChangedListeners = new Set();
    }
    EmpathyAdapter.prototype.getNextQueries = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.nextQueries.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getTopRecommendations = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.topRecommendations.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getClicksRecommendations = function (request, requestOptions) {
        return this.requestors.clicksRecommendations.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getQueriesRecommendations = function (request, requestOptions) {
        return this.requestors.queriesRecommendations.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getSectionRecommendations = function (request, requestOptions) {
        return this.requestors.sectionRecommendations.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getUserRecommendations = function (request, requestOptions) {
        return this.requestors.userRecommendations.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getRelatedTags = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.relatedTags.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.getSuggestions = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.suggestions.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.search = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.search.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.searchById = function (request, requestOptions) {
        if (requestOptions === void 0) { requestOptions = {}; }
        return this.requestors.searchById.request(request, requestOptions);
    };
    EmpathyAdapter.prototype.track = function (trackingRequest) {
        return this.requestors.track.request(trackingRequest);
    };
    EmpathyAdapter.prototype.invalidateCache = function () {
        if (this.cache) {
            this.cache.clear();
        }
    };
    EmpathyAdapter.prototype.setConfig = function (newConfig) {
        deepMerge(this.config, newConfig);
        this.notifyConfigChangedListeners();
    };
    EmpathyAdapter.prototype.addConfigChangedListener = function (listener) {
        this.configChangedListeners.add(listener);
    };
    EmpathyAdapter.prototype.removeConfigChangedListener = function (listener) {
        this.configChangedListeners.delete(listener);
    };
    EmpathyAdapter.prototype.notifyConfigChangedListeners = function () {
        var _this = this;
        this.configChangedListeners.forEach(function (listener) {
            listener(_this.config);
        });
    };
    EmpathyAdapter = __decorate([
        injectable(),
        __param(0, inject(DEPENDENCIES.config)),
        __param(1, inject(DEPENDENCIES.requestors)),
        __param(2, optional()),
        __param(2, inject(DEPENDENCIES.cacheService)),
        __metadata("design:paramtypes", [Object, Object, Object])
    ], EmpathyAdapter);
    return EmpathyAdapter;
}());
export { EmpathyAdapter };
