// @empathyco/x-components/js/x-modules/scroll/components/main-scroll.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Watch, Component } from 'vue-property-decorator';
import { XEmit } from '../../../components/decorators/bus.decorators.js';
import { XProvide } from '../../../components/decorators/injection.decorators.js';
import { State } from '../../../components/decorators/store.decorators.js';
import { NoElement } from '../../../components/no-element.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { scrollXModule } from '../x-module.js';
import { ScrollObserverKey } from './scroll.const.js';

/**
 * Extends the scroll making it able to sync the first visible element, and allowing
 * the children position to be restored.
 *
 * Each child element that wants to have this support must be wrapped in a {@link MainScrollItem}
 * component.
 *
 * @public
 */
let MainScroll = 
/* eslint-disable @typescript-eslint/unbound-method */
class MainScroll extends Vue {
    constructor() {
        super(...arguments);
        /**
         * The elements that are currently considered visible.
         *
         * @internal
         */
        this.intersectingElements = [];
        /**
         * Intersection observer to determine visibility of the elements.
         *
         * @returns An intersection observer to detect elements visibility.
         * @internal
         */
        this.intersectionObserver = null;
    }
    /**
     * Creates an `IntersectionObserver` to detect the first visible elements. Children of this
     * component should register themselves if they want to be observed.
     *
     * @returns The intersection observer.
     * @public
     */
    get visibleElementsObserver() {
        const observer = this.intersectionObserver;
        return observer
            ? {
                observe: observer.observe.bind(observer),
                unobserve: element => {
                    this.removeVisibleElement(element);
                    observer.unobserve(element);
                }
            }
            : null;
    }
    /**
     * The first visible element contained in this component.
     *
     * @returns The first visible element in this component.
     * @internal
     */
    get firstVisibleElement() {
        if (this.intersectingElements.length === 0) {
            return '';
        }
        const firstVisibleElement = this.intersectingElements.reduce((firstVisibleElement, anotherElement) => {
            // FIXME: This algorithm only takes into account LTR layouts
            const firstVisibleElementBounds = firstVisibleElement.getBoundingClientRect();
            const anotherElementBounds = anotherElement.getBoundingClientRect();
            return anotherElementBounds.left <= firstVisibleElementBounds.left &&
                anotherElementBounds.top <= firstVisibleElementBounds.top
                ? anotherElement
                : firstVisibleElement;
        });
        return firstVisibleElement === this.$el.querySelector('[data-scroll]')
            ? ''
            : firstVisibleElement.dataset.scroll;
    }
    /**
     * Initialise the observer after mounting the component.
     */
    mounted() {
        this.intersectionObserver = new IntersectionObserver(this.updateVisibleElements, {
            root: this.useWindow ? document : this.$el,
            threshold: this.threshold,
            rootMargin: this.margin
        });
    }
    /**
     * Disconnects the intersection observer.
     *
     * @internal
     */
    beforeDestroy() {
        this.intersectionObserver?.disconnect();
        this.$x.emit('UserScrolledToElement', '');
    }
    /**
     * Disconnects the previous observer.
     *
     * @param _new - The new `IntersectionObserver`.
     * @param old - The new `IntersectionObserver`.
     * @internal
     */
    disconnectPreviousObserver(_new, old) {
        old?.disconnect();
    }
    /**
     * If there is a pending scroll, starts a countdown to stop trying to restore the scroll.
     *
     * @param pendingScrollTo - The position the scroll should be restored to.
     * @internal
     */
    failRestoringScroll(pendingScrollTo) {
        // TODO Move this logic to the wiring. A cancelable delay operator is needed
        clearTimeout(this.restoreScrollFailTimeoutId);
        if (pendingScrollTo) {
            this.restoreScrollFailTimeoutId = setTimeout(() => {
                this.$x.emit('ScrollRestoreFailed');
            }, this.restoreScrollTimeoutMs);
        }
    }
    /**
     * Removes an element from the {@link MainScroll.intersectingElements} list.
     *
     * @param element - The element to remove from the visible elements.
     * @internal
     */
    removeVisibleElement(element) {
        const index = this.intersectingElements.indexOf(element);
        if (index !== -1) {
            this.intersectingElements.splice(index, 1);
        }
    }
    /**
     * Updates the visible elements given a list of intersection observer entries.
     *
     * @param entries - The entries from whom update the visibility.
     * @internal
     */
    updateVisibleElements(entries) {
        entries.forEach(entry => {
            const target = entry.target;
            if (entry.isIntersecting) {
                this.intersectingElements.push(target);
            }
            else {
                this.removeVisibleElement(target);
            }
        });
    }
    /**
     * Adds the dynamic css classes to the component.
     *
     * @returns The class to be added to the component.
     *
     * @internal
     */
    get dynamicClasses() {
        return {
            'x-main-scroll--no-transition': !!this.pendingScrollTo
        };
    }
};
__decorate([
    Prop({ default: false, type: Boolean })
], MainScroll.prototype, "useWindow", void 0);
__decorate([
    Prop({ default: 5000 })
], MainScroll.prototype, "restoreScrollTimeoutMs", void 0);
__decorate([
    Prop({ default: 0.3 })
], MainScroll.prototype, "threshold", void 0);
__decorate([
    Prop({ default: '0px' })
], MainScroll.prototype, "margin", void 0);
__decorate([
    State('scroll', 'pendingScrollTo')
], MainScroll.prototype, "pendingScrollTo", void 0);
__decorate([
    XProvide(ScrollObserverKey)
], MainScroll.prototype, "visibleElementsObserver", null);
__decorate([
    XEmit('UserScrolledToElement')
], MainScroll.prototype, "firstVisibleElement", null);
__decorate([
    Watch('intersectionObserver')
], MainScroll.prototype, "disconnectPreviousObserver", null);
__decorate([
    Watch('pendingScrollTo')
], MainScroll.prototype, "failRestoringScroll", null);
MainScroll = __decorate([
    Component({
        components: { NoElement },
        mixins: [xComponentMixin(scrollXModule)]
    })
    /* eslint-disable @typescript-eslint/unbound-method */
], MainScroll);
var script = MainScroll;

export { script as default };
