// @empathyco/x-components/js/x-modules/scroll/components/scroll-to-top.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import '../../../components/animations/animate-width.vue.js';
import '../../../components/animations/collapse-height.vue.js';
import '../../../components/animations/collapse-width.vue.js';
import '../../../components/animations/cross-fade.vue.js';
import '../../../components/animations/fade-and-slide.vue.js';
import '../../../components/animations/staggered-fade-and-slide.vue.js';
import '../../../components/animations/staggering-transition-group.vue_rollup-plugin-vue_script.vue.js';
import 'vue-runtime-helpers/dist/normalize-component.mjs';
import 'vue-runtime-helpers/dist/inject-style/browser.mjs';
import '../../../components/animations/animate-clip-path/animate-clip-path.style.scss.js';
import '../../../components/animations/animate-scale/animate-scale.style.scss.js';
import '../../../components/animations/animate-translate/animate-translate.style.scss.js';
import '../../../components/column-picker/base-column-picker-dropdown.vue.js';
import '../../../components/column-picker/base-column-picker-list.vue.js';
import '../../../components/column-picker/column-picker.mixin.js';
import '../../../components/currency/base-currency.vue.js';
import '../../../components/filters/labels/base-price-filter-label.vue_rollup-plugin-vue_script.vue.js';
import '../../../components/filters/labels/base-rating-filter-label.vue.js';
import '../../../components/icons/arrow-down.vue.js';
import '../../../components/icons/arrow-left.vue.js';
import '../../../components/icons/arrow-right.vue.js';
import '../../../components/icons/arrow-up.vue.js';
import '../../../components/icons/bag.vue.js';
import '../../../components/icons/bar-code.vue.js';
import '../../../components/icons/barcode-tiny.vue.js';
import '../../../components/icons/cart.vue.js';
import '../../../components/icons/check-tiny.vue.js';
import '../../../components/icons/check.vue.js';
import '../../../components/icons/checkbox-card-selected.vue.js';
import '../../../components/icons/checkbox-card-unselected.vue.js';
import '../../../components/icons/checkbox-selected.vue.js';
import '../../../components/icons/checkbox-unselected.vue.js';
import '../../../components/icons/chevron-down.vue.js';
import '../../../components/icons/chevron-left.vue.js';
import '../../../components/icons/chevron-right.vue.js';
import '../../../components/icons/chevron-tiny-down.vue.js';
import '../../../components/icons/chevron-tiny-left.vue.js';
import '../../../components/icons/chevron-tiny-right.vue.js';
import '../../../components/icons/chevron-tiny-up.vue.js';
import '../../../components/icons/chevron-up.vue.js';
import '../../../components/icons/cross-tiny.vue.js';
import '../../../components/icons/cross.vue.js';
import '../../../components/icons/curated-check.vue.js';
import '../../../components/icons/filters.vue.js';
import '../../../components/icons/grid-1-col.vue.js';
import '../../../components/icons/grid-2-col.vue.js';
import '../../../components/icons/grid-2-rows.vue.js';
import '../../../components/icons/hide.vue.js';
import '../../../components/icons/history-tiny.vue.js';
import '../../../components/icons/history.vue.js';
import '../../../components/icons/light-bulb-off.vue.js';
import '../../../components/icons/light-bulb-on.vue.js';
import '../../../components/icons/menu.vue.js';
import '../../../components/icons/minus-tiny.vue.js';
import '../../../components/icons/minus.vue.js';
import '../../../components/icons/nq-1.vue.js';
import '../../../components/icons/nq-2.vue.js';
import '../../../components/icons/nq-3.vue.js';
import '../../../components/icons/nq-4.vue.js';
import '../../../components/icons/plus.vue.js';
import '../../../components/icons/search.vue.js';
import '../../../components/icons/show.vue.js';
import '../../../components/icons/star.vue.js';
import '../../../components/icons/search-tiny.vue.js';
import '../../../components/icons/tag.vue.js';
import '../../../components/icons/trash-open.vue.js';
import '../../../components/icons/trash.vue.js';
import '../../../components/icons/trending-tiny.vue.js';
import '../../../components/icons/trending.vue.js';
import '../../../components/icons/user.vue.js';
import '../../../components/layouts/fixed-header-and-asides-layout.vue.js';
import '../../../components/layouts/multi-column-max-width-layout.vue.js';
import '../../../components/layouts/single-column-layout.vue.js';
import '../../../components/modals/base-events-modal-close.vue.js';
import '../../../components/modals/base-events-modal.vue.js';
import '../../../components/modals/base-events-modal-open.vue.js';
import '../../../components/modals/base-modal.vue.js';
import '../../../components/modals/base-id-modal.vue.js';
import '../../../components/modals/base-id-modal-close.vue.js';
import '../../../components/modals/base-id-modal-open.vue.js';
import '../../../components/panels/base-header-toggle-panel.vue.js';
import '../../../components/panels/base-toggle-panel.vue.js';
import '../../../components/panels/base-id-toggle-panel.vue.js';
import '../../../components/panels/base-id-toggle-panel-button.vue.js';
import '../../../components/result/base-result-add-to-cart.vue.js';
import '../../../components/result/base-result-current-price.vue.js';
import '../../../components/result/base-result-image.vue.js';
import '../../../components/result/base-result-link.vue.js';
import '../../../components/result/base-result-previous-price.vue.js';
import '../../../components/result/base-result-fallback-image.vue.js';
import '../../../components/result/base-result-placeholder-image.vue.js';
import '../../../components/result/base-result-rating.vue.js';
import '../../../components/scroll/base-scroll.vue.js';
import '../../../components/scroll/scroll.mixin.js';
import '../../../components/suggestions/base-suggestion.vue.js';
import '../../../components/suggestions/base-suggestions.vue.js';
import '../../../components/auto-progress-bar.vue.js';
import '../../../components/base-dropdown.vue.js';
import __vue_component__ from '../../../components/base-event-button.vue.js';
import '../../../components/base-grid.vue.js';
import '../../../components/base-keyboard-navigation.vue.js';
import '../../../components/base-rating.vue.js';
import '../../../components/base-variable-column-grid.vue.js';
import '../../../components/global-x-bus.vue_rollup-plugin-vue_script.vue.js';
import '../../../components/items-list.vue.js';
import '../../../components/location-provider.vue_rollup-plugin-vue_script.vue.js';
import { NoElement } from '../../../components/no-element.js';
import '../../../components/sliding-panel.vue.js';
import '../../../components/snippet-callbacks.vue.js';
import 'rxjs';
import 'rxjs/operators';
import 'vue-class-component';
import { State } from '../../../components/decorators/store.decorators.js';
import '../../../components/items-list-injection.mixin.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { scrollXModule } from '../x-module.js';
import { MainScrollId } from './scroll.const.js';

/**
 * The `ScrollToTop` component is a button that the user can click to make a container scroll
 * up to its initial position.
 *
 * @public
 */
let ScrollToTop = class ScrollToTop extends Vue {
    /**
     * The scroll data retrieved for this component.
     *
     * @returns The scroll data for this component if a valid {@link ScrollToTop.scrollId} has been
     * passed. Otherwise it returns `null`.
     * @internal
     */
    get scrollData() {
        return this.scrollId && this.scrollPositionsMap[this.scrollId]
            ? this.scrollPositionsMap[this.scrollId]
            : {
                position: 0,
                direction: 'UP',
                hasReachedStart: false,
                hasAlmostReachedEnd: false,
                hasReachedEnd: false
            };
    }
    /**
     * Event that will be emitted when the scroll to top is clicked.
     *
     * @returns The event to be emitted when the scroll to top is clicked. The id as a payload.
     * @internal
     */
    get events() {
        return { UserClickedScrollToTop: this.scrollId };
    }
    /**
     * Checks if the thresholdPx prop has been provided and if it is a number.
     *
     * @returns If the thresholdPx is a number or not.
     * @internal
     */
    get useThresholdStrategy() {
        return typeof this.thresholdPx === 'number';
    }
    /**
     * Checks if the threshold has been reached in case the threshold strategy is in use.
     *
     * @returns If the scrollTop is bigger than the thresholdPx.
     * @internal
     */
    get isThresholdReached() {
        return this.useThresholdStrategy && this.scrollData.position > this.thresholdPx;
    }
    /**
     * Whether if the button is visible or not depending on the strategy being used.
     *
     * @returns If the button should be visible or not.
     * @internal
     */
    get isVisible() {
        return this.useThresholdStrategy ? this.isThresholdReached : this.hasAlmostReachedScrollEnd;
    }
    /**
     * Returns if the scroll has almost reached its end or not.
     *
     * @returns True if the scroll has almost reached the end and the user is still scrolling down.
     * @internal
     */
    get hasAlmostReachedScrollEnd() {
        return this.scrollData.hasAlmostReachedEnd && this.scrollData.direction === 'DOWN';
    }
};
__decorate([
    Prop({ default: () => NoElement })
], ScrollToTop.prototype, "animation", void 0);
__decorate([
    Prop()
], ScrollToTop.prototype, "thresholdPx", void 0);
__decorate([
    Prop({ default: MainScrollId })
], ScrollToTop.prototype, "scrollId", void 0);
__decorate([
    State('scroll', 'data')
], ScrollToTop.prototype, "scrollPositionsMap", void 0);
ScrollToTop = __decorate([
    Component({
        mixins: [xComponentMixin(scrollXModule)],
        components: { BaseEventButton: __vue_component__ }
    })
], ScrollToTop);
var script = ScrollToTop;

export { script as default };
