// @empathyco/x-components/js/x-modules/scroll/components/scroll.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { XOn } from '../../../components/decorators/bus.decorators.js';
import __vue_component__ from '../../../components/scroll/base-scroll.vue.js';
import { scrollXModule } from '../x-module.js';
import { MainScrollId } from './scroll.const.js';

/**
 * Scrollable container that emits scroll related X Events. It exposes all the listeners
 * and props from the {@link BaseScroll} component.
 *
 * @public
 */
let Scroll = class Scroll extends Vue {
    /**
     * Emits the `UserScrolled` event.
     *
     * @param position - The new position of scroll.
     * @internal
     */
    emitScroll(position) {
        this.$x.emit('UserScrolled', position, this.createEventMetadata());
    }
    /**
     * Emits the `UserChangedScrollDirection` event when the scrolling direction has changed.
     *
     * @param direction - The new direction of scroll.
     * @internal
     */
    emitScrollDirectionChange(direction) {
        this.$x.emit('UserChangedScrollDirection', direction, this.createEventMetadata());
    }
    /**
     * Emits the 'UserReachedScrollStart' event when the user reaches the start.
     *
     * @param isAtStart - A boolean indicating if the scroll is at the ending position.
     * @internal
     */
    emitScrollAtStart(isAtStart) {
        this.$x.emit('UserReachedScrollStart', isAtStart, this.createEventMetadata());
    }
    /**
     * Emits the 'UserAlmostReachedScrollEnd' event when the user is about to reach to end.
     *
     * @param isAlmostAtEnd - A boolean indicating if the scroll is almost at its ending position.
     * @internal
     */
    emitScrollAlmostAtEnd(isAlmostAtEnd) {
        this.$x.emit('UserAlmostReachedScrollEnd', isAlmostAtEnd, this.createEventMetadata());
    }
    /**
     * Emits the 'UserReachedScrollEnd' event when the user is about to reach to end.
     *
     * @param isAtEnd - A boolean indicating if the scroll is at the ending position.
     * @internal
     */
    emitScrollAtEnd(isAtEnd) {
        this.$x.emit('UserReachedScrollEnd', isAtEnd, this.createEventMetadata());
    }
    /**
     * Creates a {@link WireMetadata} metadata object for all the emitted events.
     *
     * @internal
     * @returns A new {@link WireMetadata} object.
     */
    createEventMetadata() {
        return { target: this.$el, id: this.id };
    }
    /**
     * Scrolls to initial position when the user has clicked the scroll to top button.
     *
     * @param scrollId - {@link XEventsTypes.UserClickedScrollToTop}.
     * @internal
     */
    scrollToTop(scrollId) {
        if (scrollId === this.id && this.$el) {
            this.$el?.scrollTo({ top: 0, behavior: 'smooth' });
        }
    }
};
__decorate([
    Prop({ default: MainScrollId })
], Scroll.prototype, "id", void 0);
__decorate([
    XOn('UserClickedScrollToTop')
], Scroll.prototype, "scrollToTop", null);
Scroll = __decorate([
    Component({
        mixins: [xComponentMixin(scrollXModule)],
        components: { BaseScroll: __vue_component__ }
    })
], Scroll);
var script = Scroll;

export { script as default };
