// @empathyco/x-components/js/x-modules/scroll/components/window-scroll.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import { mixins } from 'vue-class-component';
import { Prop, Component } from 'vue-property-decorator';
import '../../../components/animations/animate-width.vue.js';
import '../../../components/animations/collapse-height.vue.js';
import '../../../components/animations/collapse-width.vue.js';
import '../../../components/animations/cross-fade.vue.js';
import '../../../components/animations/fade-and-slide.vue.js';
import '../../../components/animations/staggered-fade-and-slide.vue.js';
import '../../../components/animations/staggering-transition-group.vue_rollup-plugin-vue_script.vue.js';
import 'vue-runtime-helpers/dist/normalize-component.mjs';
import 'vue-runtime-helpers/dist/inject-style/browser.mjs';
import 'vue';
import '../../../components/animations/animate-clip-path/animate-clip-path.style.scss.js';
import '../../../components/animations/animate-scale/animate-scale.style.scss.js';
import '../../../components/animations/animate-translate/animate-translate.style.scss.js';
import '../../../components/column-picker/base-column-picker-dropdown.vue.js';
import '../../../components/column-picker/base-column-picker-list.vue.js';
import '../../../components/column-picker/column-picker.mixin.js';
import '../../../components/currency/base-currency.vue.js';
import '../../../components/filters/labels/base-price-filter-label.vue_rollup-plugin-vue_script.vue.js';
import '../../../components/filters/labels/base-rating-filter-label.vue.js';
import '../../../components/icons/arrow-down.vue.js';
import '../../../components/icons/arrow-left.vue.js';
import '../../../components/icons/arrow-right.vue.js';
import '../../../components/icons/arrow-up.vue.js';
import '../../../components/icons/bag.vue.js';
import '../../../components/icons/bar-code.vue.js';
import '../../../components/icons/barcode-tiny.vue.js';
import '../../../components/icons/cart.vue.js';
import '../../../components/icons/check-tiny.vue.js';
import '../../../components/icons/check.vue.js';
import '../../../components/icons/checkbox-card-selected.vue.js';
import '../../../components/icons/checkbox-card-unselected.vue.js';
import '../../../components/icons/checkbox-selected.vue.js';
import '../../../components/icons/checkbox-unselected.vue.js';
import '../../../components/icons/chevron-down.vue.js';
import '../../../components/icons/chevron-left.vue.js';
import '../../../components/icons/chevron-right.vue.js';
import '../../../components/icons/chevron-tiny-down.vue.js';
import '../../../components/icons/chevron-tiny-left.vue.js';
import '../../../components/icons/chevron-tiny-right.vue.js';
import '../../../components/icons/chevron-tiny-up.vue.js';
import '../../../components/icons/chevron-up.vue.js';
import '../../../components/icons/cross-tiny.vue.js';
import '../../../components/icons/cross.vue.js';
import '../../../components/icons/curated-check.vue.js';
import '../../../components/icons/filters.vue.js';
import '../../../components/icons/grid-1-col.vue.js';
import '../../../components/icons/grid-2-col.vue.js';
import '../../../components/icons/grid-2-rows.vue.js';
import '../../../components/icons/hide.vue.js';
import '../../../components/icons/history-tiny.vue.js';
import '../../../components/icons/history.vue.js';
import '../../../components/icons/light-bulb-off.vue.js';
import '../../../components/icons/light-bulb-on.vue.js';
import '../../../components/icons/menu.vue.js';
import '../../../components/icons/minus-tiny.vue.js';
import '../../../components/icons/minus.vue.js';
import '../../../components/icons/nq-1.vue.js';
import '../../../components/icons/nq-2.vue.js';
import '../../../components/icons/nq-3.vue.js';
import '../../../components/icons/nq-4.vue.js';
import '../../../components/icons/plus.vue.js';
import '../../../components/icons/search.vue.js';
import '../../../components/icons/show.vue.js';
import '../../../components/icons/star.vue.js';
import '../../../components/icons/search-tiny.vue.js';
import '../../../components/icons/tag.vue.js';
import '../../../components/icons/trash-open.vue.js';
import '../../../components/icons/trash.vue.js';
import '../../../components/icons/trending-tiny.vue.js';
import '../../../components/icons/trending.vue.js';
import '../../../components/icons/user.vue.js';
import '../../../components/layouts/fixed-header-and-asides-layout.vue.js';
import '../../../components/layouts/multi-column-max-width-layout.vue.js';
import '../../../components/layouts/single-column-layout.vue.js';
import '../../../components/modals/base-events-modal-close.vue.js';
import '../../../components/modals/base-events-modal.vue.js';
import '../../../components/modals/base-events-modal-open.vue.js';
import '../../../components/modals/base-modal.vue.js';
import '../../../components/modals/base-id-modal.vue.js';
import '../../../components/modals/base-id-modal-close.vue.js';
import '../../../components/modals/base-id-modal-open.vue.js';
import '../../../components/panels/base-header-toggle-panel.vue.js';
import '../../../components/panels/base-toggle-panel.vue.js';
import '../../../components/panels/base-id-toggle-panel.vue.js';
import '../../../components/panels/base-id-toggle-panel-button.vue.js';
import '../../../components/result/base-result-add-to-cart.vue.js';
import '../../../components/result/base-result-current-price.vue.js';
import '../../../components/result/base-result-image.vue.js';
import '../../../components/result/base-result-link.vue.js';
import '../../../components/result/base-result-previous-price.vue.js';
import '../../../components/result/base-result-fallback-image.vue.js';
import '../../../components/result/base-result-placeholder-image.vue.js';
import '../../../components/result/base-result-rating.vue.js';
import '../../../components/scroll/base-scroll.vue.js';
import ScrollMixin from '../../../components/scroll/scroll.mixin.js';
import '../../../components/suggestions/base-suggestion.vue.js';
import '../../../components/suggestions/base-suggestions.vue.js';
import '../../../components/auto-progress-bar.vue.js';
import '../../../components/base-dropdown.vue.js';
import '../../../components/base-event-button.vue.js';
import '../../../components/base-grid.vue.js';
import '../../../components/base-keyboard-navigation.vue.js';
import '../../../components/base-rating.vue.js';
import '../../../components/base-variable-column-grid.vue.js';
import '../../../components/global-x-bus.vue_rollup-plugin-vue_script.vue.js';
import '../../../components/items-list.vue.js';
import '../../../components/location-provider.vue_rollup-plugin-vue_script.vue.js';
import '../../../components/no-element.js';
import '../../../components/sliding-panel.vue.js';
import '../../../components/snippet-callbacks.vue.js';
import 'rxjs';
import 'rxjs/operators';
import '../../../components/items-list-injection.mixin.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { scrollXModule } from '../x-module.js';
import { MainScrollId } from './scroll.const.js';

/**
 * The `WindowScroll` component listens to either the `html` or `body` DOM scroll events, and
 * re-emits them as X Events. Additionally it also emits events related to the direction or
 * current position of these elements scroll.
 *
 * @public
 */
let WindowScroll = class WindowScroll extends mixins(ScrollMixin) {
    mounted() {
        this.initAndListenElement();
        this.$on('scroll', (position) => {
            this.$x.emit('UserScrolled', position, this.createXEventMetadata());
        });
        this.$on('scroll:direction-change', (direction) => {
            this.$x.emit('UserChangedScrollDirection', direction, this.createXEventMetadata());
        });
        this.$on('scroll:at-start', (hasReachedStart) => {
            this.$x.emit('UserReachedScrollStart', hasReachedStart, this.createXEventMetadata());
        });
        this.$on('scroll:almost-at-end', (hasAlmostReachedEnd) => {
            this.$x.emit('UserAlmostReachedScrollEnd', hasAlmostReachedEnd, this.createXEventMetadata());
        });
        this.$on('scroll:at-end', (hasReachedEnd) => {
            this.$x.emit('UserReachedScrollEnd', hasReachedEnd, this.createXEventMetadata());
        });
    }
    /**
     * Sets the HTML element depending on {@link WindowScroll.scrollableElement}, and initialises
     * its events.
     *
     * @internal
     */
    initAndListenElement() {
        this.$el = this.scrollableElement === 'body' ? document.body : document.documentElement;
        this.$el.addEventListener('scroll', this.throttledStoreScrollData);
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    render() { }
    /**
     * Cleanup listeners.
     */
    beforeDestroy() {
        this.$el.removeEventListener('scroll', this.throttledStoreScrollData);
    }
    /**
     * Creates the metadata for the events of this component.
     *
     * @returns A {@link WireMetadata} for the events emitted by this component.
     * @internal
     */
    createXEventMetadata() {
        return { target: this.$el, id: this.id };
    }
};
__decorate([
    Prop({ default: 'html' })
], WindowScroll.prototype, "scrollableElement", void 0);
__decorate([
    Prop({ default: MainScrollId })
], WindowScroll.prototype, "id", void 0);
WindowScroll = __decorate([
    Component({
        mixins: [xComponentMixin(scrollXModule)]
    })
], WindowScroll);
var script = WindowScroll;

export { script as default };
