// @empathyco/x-adapter/dist/esm/empathy/mappers/response.mappers.js
import { __decorate, __metadata, __param } from "tslib";
import { injectable, multiInject } from 'inversify';
import { DEPENDENCIES } from '../container/container.const';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var ResponseMappers = /** @class */ (function () {
    function ResponseMappers(banners, facets, nextQueries, partialResults, promoteds, queryTagging, redirections, relatedTags, results, showTagging, spellcheck, suggestions, totalResults) {
        this.banners = banners;
        this.facets = facets;
        this.nextQueries = nextQueries;
        this.partialResults = partialResults;
        this.promoteds = promoteds;
        this.queryTagging = queryTagging;
        this.redirections = redirections;
        this.relatedTags = relatedTags;
        this.results = results;
        this.showTagging = showTagging;
        this.spellcheck = spellcheck;
        this.suggestions = suggestions;
        this.totalResults = totalResults;
    }
    ResponseMappers = __decorate([
        injectable(),
        __param(0, multiInject(DEPENDENCIES.ResponseMappers.banners)),
        __param(1, multiInject(DEPENDENCIES.ResponseMappers.facets)),
        __param(2, multiInject(DEPENDENCIES.ResponseMappers.nextQueries)),
        __param(3, multiInject(DEPENDENCIES.ResponseMappers.partialResults)),
        __param(4, multiInject(DEPENDENCIES.ResponseMappers.promoteds)),
        __param(5, multiInject(DEPENDENCIES.ResponseMappers.queryTagging)),
        __param(6, multiInject(DEPENDENCIES.ResponseMappers.redirections)),
        __param(7, multiInject(DEPENDENCIES.ResponseMappers.relatedTags)),
        __param(8, multiInject(DEPENDENCIES.ResponseMappers.results)),
        __param(9, multiInject(DEPENDENCIES.ResponseMappers.showTagging)),
        __param(10, multiInject(DEPENDENCIES.ResponseMappers.spellcheck)),
        __param(11, multiInject(DEPENDENCIES.ResponseMappers.suggestions)),
        __param(12, multiInject(DEPENDENCIES.ResponseMappers.totalResults)),
        __metadata("design:paramtypes", [Array, Array, Array, Array, Array, Array, Array, Array, Array, Array, Array, Array, Array])
    ], ResponseMappers);
    return ResponseMappers;
}());
export { ResponseMappers };
