// @empathyco/x-adapter/dist/esm/empathy/mappers/response/facets/empathy-hierarchical-facet.mapper.js
import { __assign, __decorate, __metadata, __param } from "tslib";
import { injectable, multiInject } from 'inversify';
import { DEPENDENCIES } from '../../../container/container.const';
import { pipeMappers } from '../../pipe-mappers';
/**
 * Hierarchical Empathy facet mapper.
 *
 * @public
 */
var EmpathyHierarchicalFacetMapper = /** @class */ (function () {
    function EmpathyHierarchicalFacetMapper(filterMappers) {
        this.mapFilter = pipeMappers.apply(void 0, filterMappers);
    }
    EmpathyHierarchicalFacetMapper.prototype.map = function (rawFacet, facet, context) {
        if (facet.modelName === 'HierarchicalFacet') {
            Object.assign(facet, {
                filters: this.mapFiltersTree(facet, rawFacet.values, context)
            });
        }
        return facet;
    };
    EmpathyHierarchicalFacetMapper.prototype.mapFiltersTree = function (facet, rawFilters, context) {
        var _this = this;
        var filters = [];
        var initFilterProps = { facetId: facet.id, parentId: null };
        rawFilters.forEach(function (rawFilter) {
            _this.mapDeepFilter(rawFilter, initFilterProps, context, filters);
        });
        return filters;
    };
    EmpathyHierarchicalFacetMapper.prototype.mapDeepFilter = function (rawFilter, initFilterProps, context, filters) {
        var _this = this;
        var _a;
        var filter = this.mapFilter(rawFilter, __assign({}, initFilterProps), context);
        filter.children =
            (_a = rawFilter.values) === null || _a === void 0 ? void 0 : _a.map(function (rawFilterChild) { return _this.mapDeepFilter(rawFilterChild, __assign(__assign({}, initFilterProps), { parentId: filter.id }), context, filters); });
        filters.push(filter);
        return filter.id;
    };
    EmpathyHierarchicalFacetMapper = __decorate([
        injectable(),
        __param(0, multiInject(DEPENDENCIES.ResponseMappers.hierarchicalFilter)),
        __metadata("design:paramtypes", [Array])
    ], EmpathyHierarchicalFacetMapper);
    return EmpathyHierarchicalFacetMapper;
}());
export { EmpathyHierarchicalFacetMapper };
