// @empathyco/x-adapter/dist/esm/empathy/mappers/response/facets/empathy-number-range-facet.mapper.js
import { __decorate, __metadata, __param } from "tslib";
import { inject, injectable, multiInject } from 'inversify';
import { DEPENDENCIES } from '../../../container/container.const';
import { pipeMappers } from '../../pipe-mappers';
/**
 * Number Range Empathy facet mapper.
 *
 * @public
 */
var EmpathyNumberRangeFacetMapper = /** @class */ (function () {
    function EmpathyNumberRangeFacetMapper(filterMappers, config) {
        this.mapFilter = pipeMappers.apply(void 0, filterMappers);
        this.facetsConfig = config.mappings.facets;
    }
    EmpathyNumberRangeFacetMapper.prototype.map = function (rawFacet, facet, context) {
        var isDynamic = (this.facetsConfig.named[facet.id] || this.facetsConfig.default).isDynamic;
        if (facet.modelName === 'NumberRangeFacet') {
            Object.assign(facet, {
                filters: this.mapFilters(facet, rawFacet.values, isDynamic, context)
            });
        }
        return facet;
    };
    EmpathyNumberRangeFacetMapper.prototype.mapFilters = function (facet, rawFilters, isDynamic, context) {
        var _this = this;
        if (rawFilters === void 0) { rawFilters = []; }
        var rawFiltersToMap = isDynamic ? [rawFilters[0]] : rawFilters;
        return rawFiltersToMap.map(function (rawFilter) { return _this.mapFilter(rawFilter, { facetId: facet.id }, context); });
    };
    EmpathyNumberRangeFacetMapper = __decorate([
        injectable(),
        __param(0, multiInject(DEPENDENCIES.ResponseMappers.numberRangeFilter)),
        __param(1, inject(DEPENDENCIES.config)),
        __metadata("design:paramtypes", [Array, Object])
    ], EmpathyNumberRangeFacetMapper);
    return EmpathyNumberRangeFacetMapper;
}());
export { EmpathyNumberRangeFacetMapper };
