// @empathyco/x-adapter/dist/cjs/empathy/http-clients/fetch-http-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FetchHttpClient = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
var container_const_1 = require("../container/container.const");
var request_error_1 = require("./errors/request.error");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var FetchHttpClient = /** @class */ (function () {
    function FetchHttpClient(cache) {
        this.cache = cache;
        this.requestsAbortControllers = {};
    }
    FetchHttpClient.prototype.get = function (endpoint, params, _a) {
        if (params === void 0) { params = {}; }
        var _b = _a === void 0 ? {} : _a, _c = _b.requestId, requestId = _c === void 0 ? endpoint : _c, _d = _b.headers, headers = _d === void 0 ? {} : _d, _e = _b.ttlInMinutes, ttlInMinutes = _e === void 0 ? 0 : _e;
        this.cancelPreviousRequest(requestId);
        var url = this.buildUrl(endpoint, params);
        var requestOptions = this.getRequestOptions(requestId, headers);
        var cachedResponse = this.cache && this.cache.getItem(url);
        return cachedResponse
            ? Promise.resolve(cachedResponse)
            : fetch(url, requestOptions)
                .then(this.parseResponse)
                .then(this.storeInCache(url, ttlInMinutes));
    };
    FetchHttpClient.prototype.parseResponse = function (response) {
        if (response.ok) {
            return response.json();
        }
        else {
            throw new request_error_1.RequestError('Request failed', response);
        }
    };
    FetchHttpClient.prototype.buildUrl = function (endpoint, params) {
        var url = new URL(endpoint);
        Object.entries(params).forEach(function (_a) {
            var key = _a[0], param = _a[1];
            if (param !== undefined) {
                if (Array.isArray(param)) {
                    param.forEach(function (arrayParamValue) { return url.searchParams.append(key, arrayParamValue); });
                }
                else {
                    url.searchParams.set(key, param);
                }
            }
        });
        return url.href;
    };
    FetchHttpClient.prototype.cancelPreviousRequest = function (requestId) {
        var previousRequest = this.requestsAbortControllers[requestId];
        if (previousRequest) {
            previousRequest.abort();
        }
    };
    FetchHttpClient.prototype.getRequestOptions = function (requestId, headers) {
        if (typeof AbortController !== 'undefined') {
            var signal = this.createAbortController(requestId);
            return { signal: signal, headers: headers };
        }
        return { headers: headers };
    };
    FetchHttpClient.prototype.createAbortController = function (requestId) {
        var controller = new AbortController();
        this.requestsAbortControllers[requestId] = controller;
        return controller.signal;
    };
    FetchHttpClient.prototype.storeInCache = function (url, ttlInMinutes) {
        var _this = this;
        return function (response) {
            if (_this.cache && ttlInMinutes) {
                _this.cache.setItem(url, response, ttlInMinutes);
            }
            return response;
        };
    };
    FetchHttpClient = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.optional)()),
        tslib_1.__param(0, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.cacheService)),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], FetchHttpClient);
    return FetchHttpClient;
}());
exports.FetchHttpClient = FetchHttpClient;
