// @empathyco/x-components/js/components/scroll/scroll.mixin.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Watch, Component } from 'vue-property-decorator';
import { throttle } from '../../utils/throttle.js';
import { XOn } from '../decorators/bus.decorators.js';

/**
 * Mixin to share Scroll logic.
 *
 * @public
 */
let ScrollMixin = 
/* eslint-disable @typescript-eslint/unbound-method */
class ScrollMixin extends Vue {
    constructor() {
        super(...arguments);
        /**
         * Property for getting the client height of scroll.
         *
         * @internal
         */
        this.clientHeight = 0;
        /**
         * Property for getting the current position of scroll.
         *
         * @internal
         */
        this.currentPosition = 0;
        /**
         * Property for getting the previous position of scroll.
         *
         * @internal
         */
        this.previousPosition = 0;
        /**
         * Property for getting the scroll height.
         *
         * @internal
         */
        this.scrollHeight = 0;
    }
    /**
     * Throttled version of the function that stores the DOM scroll related properties.
     * The duration of the throttle is configured through the
     * {@link ScrollMixin.throttleMs}.
     *
     * @returns A throttled version of the function to store the scroll data.
     * @internal
     */
    get throttledStoreScrollData() {
        return throttle(this.storeScrollData, this.throttleMs);
    }
    /**
     * Returns distance missing to end position position.
     *
     * @returns A number for knowing the distance missing to end position position.
     * @internal
     */
    get distanceToEnd() {
        return this.scrollEndPosition - this.currentPosition;
    }
    /**
     * Returns `true` when the amount of pixels scrolled is greater than
     * the {@link ScrollMixin.distanceToBottom}.
     *
     * @returns A boolean for knowing if the user is about to reaching to the end.
     * @internal
     */
    get hasScrollAlmostReachedEnd() {
        return !this.hasScrollReachedStart && this.distanceToBottom > this.distanceToEnd;
    }
    /**
     * Returns `true` when there is no more content to scroll.
     *
     * @returns A boolean for knowing if the user scrolls to the end.
     * @internal
     */
    get hasScrollReachedEnd() {
        return this.currentPosition === this.scrollEndPosition;
    }
    /**
     * Returns `true` when the scroll is at the initial position.
     *
     * @returns A boolean for knowing if the user scrolls to the start.
     * @internal
     */
    get hasScrollReachedStart() {
        return this.currentPosition === 0;
    }
    /**
     * Returns direction of scroll based in {@link ScrollDirection}.
     *
     * @returns The scroll direction.
     * @internal
     */
    get scrollDirection() {
        return this.currentPosition > this.previousPosition ? 'DOWN' : 'UP';
    }
    /**
     * Returns end position of scroll.
     *
     * @returns A number for knowing end position of scroll.
     * @internal
     */
    get scrollEndPosition() {
        return this.scrollHeight - this.clientHeight;
    }
    /**
     * Initialises DOM dependant scroll properties.
     *
     * @internal
     */
    mounted() {
        this.$nextTick().then(() => {
            if (!this.$el) {
                // TODO Replace with Empathy's logger
                // eslint-disable-next-line no-console
                console.warn('[ScrollMixin]', 'Components using this mixin must set `this.$el` to the HTML node that is scrolling.');
            }
            else {
                this.storeScrollData();
            }
        });
    }
    /**
     * Resets the scroll position.
     *
     * @internal
     */
    resetScroll() {
        this.$nextTick().then(() => {
            if (this.resetOnChange) {
                this.$el.scrollTo({ top: 0 });
            }
        });
    }
    /**
     * Emits the `scroll` event.
     *
     * @param _newScrollPosition - The new position of scroll.
     * @param oldScrollPosition - The old position of scroll.
     * @internal
     */
    emitScroll(_newScrollPosition, oldScrollPosition) {
        this.$emit('scroll', this.currentPosition);
        this.previousPosition = oldScrollPosition;
    }
    /**
     * Emits the 'scroll:at-start' event when the user reaches the start.
     *
     * @param isScrollAtStart - For knowing if the user reaches at start.
     * @internal
     */
    emitScrollReachedAtStart(isScrollAtStart) {
        this.$emit('scroll:at-start', isScrollAtStart);
    }
    /**
     * Emits the 'scroll:almost-at-end' event when the user is about to reach to end.
     *
     * @param isScrollAlmostAtEnd - For knowing if the user is about to reach to end.
     * @internal
     */
    emitScrollAlmostAtEnd(isScrollAlmostAtEnd) {
        this.$emit('scroll:almost-at-end', isScrollAlmostAtEnd);
    }
    /**
     * Emits the 'scroll:at-end' event when the user reaches the end.
     *
     * @param isScrollAtEnd - For knowing if the user reaches at end.
     * @internal
     */
    emitScrollAtEnd(isScrollAtEnd) {
        this.$emit('scroll:at-end', isScrollAtEnd);
    }
    /**
     * Emits the `scroll:direction-change` event when the scrolling direction has changed.
     *
     * @param direction - The new direction of scroll.
     * @internal
     */
    emitScrollDirection(direction) {
        this.$emit('scroll:direction-change', direction);
    }
    /**
     * Updates scroll related properties.
     *
     * @internal
     */
    storeScrollData() {
        if (this.$el) {
            this.currentPosition = this.$el.scrollTop;
            this.scrollHeight = this.$el.scrollHeight;
            this.clientHeight = this.$el.clientHeight;
        }
    }
};
__decorate([
    Prop({ default: 100 })
], ScrollMixin.prototype, "distanceToBottom", void 0);
__decorate([
    Prop({ default: 100 })
], ScrollMixin.prototype, "firstElementThresholdPx", void 0);
__decorate([
    Prop({ default: 100 })
], ScrollMixin.prototype, "throttleMs", void 0);
__decorate([
    Prop({ type: Boolean, default: true })
], ScrollMixin.prototype, "resetOnChange", void 0);
__decorate([
    Prop({
        default: () => [
            'SearchBoxQueryChanged',
            'SortChanged',
            'SelectedFiltersChanged',
            'SelectedRelatedTagsChanged',
            'UserChangedExtraParams'
        ]
    })
], ScrollMixin.prototype, "resetOn", void 0);
__decorate([
    XOn(instance => instance.resetOn)
], ScrollMixin.prototype, "resetScroll", null);
__decorate([
    Watch('currentPosition')
], ScrollMixin.prototype, "emitScroll", null);
__decorate([
    Watch('hasScrollReachedStart')
], ScrollMixin.prototype, "emitScrollReachedAtStart", null);
__decorate([
    Watch('hasScrollAlmostReachedEnd')
], ScrollMixin.prototype, "emitScrollAlmostAtEnd", null);
__decorate([
    Watch('hasScrollReachedEnd')
], ScrollMixin.prototype, "emitScrollAtEnd", null);
__decorate([
    Watch('scrollDirection')
], ScrollMixin.prototype, "emitScrollDirection", null);
ScrollMixin = __decorate([
    Component
    /* eslint-disable @typescript-eslint/unbound-method */
], ScrollMixin);
var ScrollMixin$1 = ScrollMixin;
/*  eslint-enable @typescript-eslint/unbound-method */

export { ScrollMixin$1 as default };
