// @empathyco/x-adapter/dist/esm/empathy/mappers/response/results/empathy-result-query-tagging.mapper.js
import { __decorate, __metadata, __param } from "tslib";
import { getSafePropertyChain } from '@empathyco/x-get-safe-property-chain';
import { inject, injectable, multiInject } from 'inversify';
import { DEPENDENCIES } from '../../../container/container.const';
import { Logger } from '../../../logger';
import { pipeMappers } from '../../pipe-mappers';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyResultQueryTaggingMapper = /** @class */ (function () {
    function EmpathyResultQueryTaggingMapper(config, queryTaggingMappers) {
        this.logger = Logger.child('EmpathyResultQueryTaggingMapper');
        this.mapQueryTagging = pipeMappers.apply(void 0, queryTaggingMappers);
        this.queryTrackingPath = config.features.search.responsePaths.queryTagging;
    }
    EmpathyResultQueryTaggingMapper.prototype.map = function (_rawResult, result, context) {
        Object.assign(result.tagging, { query: context.resultsQueryTagging || this.mapNewQueryTagging(context) });
        return result;
    };
    EmpathyResultQueryTaggingMapper.prototype.mapNewQueryTagging = function (context) {
        var queryTagging = getSafePropertyChain(context.rawResponse, this.queryTrackingPath);
        if (queryTagging) {
            return context.resultsQueryTagging = this.mapQueryTagging(queryTagging, {}, context);
        }
        this.logger.warn("Query tagging at path \"".concat(this.queryTrackingPath, "\" not found when mapping for feature \"").concat(context.feature, "\""));
        return null;
    };
    EmpathyResultQueryTaggingMapper = __decorate([
        injectable(),
        __param(0, inject(DEPENDENCIES.config)),
        __param(1, multiInject(DEPENDENCIES.ResponseMappers.queryTagging)),
        __metadata("design:paramtypes", [Object, Array])
    ], EmpathyResultQueryTaggingMapper);
    return EmpathyResultQueryTaggingMapper;
}());
export { EmpathyResultQueryTaggingMapper };
