// @empathyco/x-adapter/dist/esm/empathy/mappers/response/results/empathy-result.mapper.js
import { __decorate, __metadata, __param } from "tslib";
import { getSafePropertyChain } from '@empathyco/x-get-safe-property-chain';
import { inject, injectable, multiInject } from 'inversify';
import { DEPENDENCIES } from '../../../container/container.const';
import { Logger } from '../../../logger';
import { pipeMappers } from '../../pipe-mappers';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyResultMapper = /** @class */ (function () {
    function EmpathyResultMapper(config, taggingMappers) {
        this.logger = Logger.child('EmpathyResultMapper');
        this.mapTagging = pipeMappers.apply(void 0, taggingMappers);
        this.trackingResultConfig = config.mappings.tracking.result;
    }
    EmpathyResultMapper.prototype.map = function (rawResult, result, context) {
        var value = Number.parseFloat(rawResult.price);
        var originalValue = Number.parseFloat((rawResult.originalPrice || rawResult.price));
        var sku = rawResult.eb_sku || '';
        return Object.assign(result, {
            modelName: 'Result',
            type: 'default',
            id: rawResult.id,
            name: rawResult.name || '',
            url: rawResult.url,
            images: rawResult.image ? [rawResult.image] : [],
            price: {
                value: value,
                originalValue: originalValue,
                hasDiscount: value < originalValue
            },
            rating: {
                value: rawResult.rating !== undefined ? Number.parseFloat(rawResult.rating) : null
            },
            identifier: {
                value: sku
            },
            tagging: this.createResultTagging(rawResult, result.tagging || {}, context),
            isWishlisted: rawResult.isWishlisted || false
        });
    };
    EmpathyResultMapper.prototype.createResultTagging = function (rawResult, initialResultTagging, context) {
        var _this = this;
        return Object.entries(this.trackingResultConfig).reduce(function (resultTagging, _a) {
            var key = _a[0], path = _a[1];
            var rawTagging = getSafePropertyChain(rawResult, path);
            if (rawTagging) {
                resultTagging[key] = _this.mapTagging(rawTagging, {}, context);
            }
            else {
                if (process.env.NODE_ENV !== 'production') {
                    _this.logger.warn("Tagging for event \"".concat(key, "\" at path \"").concat(path, "\" not found when mapping for feature \"").concat(context.feature, "\""));
                }
            }
            return resultTagging;
        }, initialResultTagging);
    };
    EmpathyResultMapper = __decorate([
        injectable(),
        __param(0, inject(DEPENDENCIES.config)),
        __param(1, multiInject(DEPENDENCIES.ResponseMappers.Helpers.tagging)),
        __metadata("design:paramtypes", [Object, Array])
    ], EmpathyResultMapper);
    return EmpathyResultMapper;
}());
export { EmpathyResultMapper };
