// @empathyco/x-components/js/x-modules/related-tags/store/module.js
import { setQuery } from '../../../store/utils/query.utils.js';
import { setStatus } from '../../../store/utils/status-store.utils.js';
import { cancelFetchAndSaveRelatedTags, fetchAndSaveRelatedTags } from './actions/fetch-and-save-related-tags.action.js';
import { fetchRelatedTags } from './actions/fetch-related-tags.action.js';
import { setUrlParams } from './actions/set-url-params.action.js';
import { toggleRelatedTag } from './actions/toggle-related-tag.action.js';
import { query } from './getters/query.getter.js';
import { relatedTags } from './getters/related-tags.getter.js';
import { request } from './getters/request.getter.js';

/**
 * {@link XStoreModule} For the related tags module.
 *
 * @internal
 */
const relatedTagsXStoreModule = {
    state: () => ({
        query: '',
        relatedTags: [],
        selectedRelatedTags: [],
        status: 'initial',
        config: {
            maxItemsToRequest: 10
        },
        params: {}
    }),
    getters: {
        request,
        relatedTags,
        query
    },
    mutations: {
        setQuery,
        setRelatedTags(state, relatedTags) {
            state.relatedTags = relatedTags;
        },
        setSelectedRelatedTags(state, selectedRelatedTags) {
            state.selectedRelatedTags = selectedRelatedTags;
        },
        setStatus,
        setParams(state, params) {
            state.params = params;
        }
    },
    actions: {
        cancelFetchAndSaveRelatedTags,
        fetchRelatedTags,
        fetchAndSaveRelatedTags,
        toggleRelatedTag,
        setUrlParams
    }
};

export { relatedTagsXStoreModule };
