// @empathyco/x-adapter/dist/cjs/empathy/services/empathy-cache.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyCacheService = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
var container_const_1 = require("../container/container.const");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyCacheService = /** @class */ (function () {
    function EmpathyCacheService(storageService) {
        this.storageService = storageService;
    }
    EmpathyCacheService.prototype.setItem = function (key, item, ttlInMinutes) {
        this.storageService.setItem(this.transformKey(key), item, this.minutesToMilliseconds(ttlInMinutes));
    };
    EmpathyCacheService.prototype.getItem = function (key) {
        return this.storageService.getItem(this.transformKey(key));
    };
    EmpathyCacheService.prototype.removeItem = function (key) {
        return this.storageService.removeItem(this.transformKey(key));
    };
    EmpathyCacheService.prototype.clear = function () {
        return this.storageService.clear();
    };
    EmpathyCacheService.prototype.transformKey = function (key) {
        return "cache-".concat(this.hash(key));
    };
    EmpathyCacheService.prototype.hash = function (key) {
        // https://gist.github.com/hyamamoto/fd435505d29ebfa3d9716fd2be8d42f0#gistcomment-2775538
        var h = 0;
        for (var i = 0; i < key.length; i++) {
            // tslint:disable-next-line:no-bitwise
            h = (Math.imul(31, h) + key.charCodeAt(i)) | 0;
        }
        return h.toString();
    };
    EmpathyCacheService.prototype.minutesToMilliseconds = function (minutes) {
        return minutes * 60 * 1000;
    };
    EmpathyCacheService = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.storageService)),
        tslib_1.__metadata("design:paramtypes", [Object])
    ], EmpathyCacheService);
    return EmpathyCacheService;
}());
exports.EmpathyCacheService = EmpathyCacheService;
