// @empathyco/x-utils/dist/cjs/object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.flatObject = exports.every = exports.getNewAndUpdatedKeys = exports.objectFilter = exports.cleanUndefined = exports.map = exports.reduce = exports.forEach = void 0;
const typeguards_1 = require("./typeguards");
/**
 * Iterates over every non-undefined property of the object calling the callback passed as
 * parameter.
 *
 * @param obj - The object to iterate through each property.
 * @param callbackFn - The callback function to call for each property.
 * @public
 */
function forEach(obj, callbackFn) {
    if (obj == null) {
        return;
    }
    let index = 0;
    for (const key in obj) {
        if (Object.prototype.hasOwnProperty.call(obj, key) && obj[key] !== undefined) {
            callbackFn(key, obj[key], index++);
        }
    }
}
exports.forEach = forEach;
/**
 * Iterates through the obj properties calling the reducer function.
 *
 * @param obj - The object to iterate through each property.
 * @param reducer - A function that will be called for each property, modifying the initialValue
 * object.
 * @param initialValue - The initial value of the accumulator property of the reducer function.
 * @returns Result of the reducer function.
 * @public
 */
function reduce(obj, reducer, initialValue) {
    let accumulator = initialValue;
    forEach(obj, (key, value, index) => {
        accumulator = reducer(accumulator, key, value, index);
    });
    return accumulator;
}
exports.reduce = reduce;
/**
 * Creates an object from another object transforming each property value.
 *
 * @param obj - The object to transform each property value.
 * @param mapper - The mapper function which will transform each value.
 * @returns A record with the result of the mapper.
 * @public
 */
function map(obj, mapper) {
    return reduce(obj, (accumulator, key, value, index) => {
        accumulator[key] = mapper(key, value, index);
        return accumulator;
    }, {});
}
exports.map = map;
/**
 * Creates an object picking only the not undefined properties.
 *
 * @param obj - The object from whom pick the values.
 * @returns A new object with the not undefined properties of the source object.
 * @public
 */
function cleanUndefined(obj) {
    return typeof obj !== 'object' || obj === null || Array.isArray(obj)
        ? obj
        : reduce(obj, (pickedObject, key, value) => {
            pickedObject[key] = cleanUndefined(value);
            return pickedObject;
        }, {});
}
exports.cleanUndefined = cleanUndefined;
/**
 * Creates an object picking only the ones that pass the test implemented by the
 * provided function isIncluded.
 *
 * @param obj - T object to be filtered.
 * @param isIncluded - Test function that every obj item must pass.
 * @returns A filtered object.
 * @public
 */
function objectFilter(obj, isIncluded) {
    return reduce(obj, (accumulator, key, value, index) => {
        if (isIncluded(key, value, index)) {
            accumulator[key] = value;
        }
        return accumulator;
    }, {});
}
exports.objectFilter = objectFilter;
/**
 * Compares two objects of the same type, checking the values of their keys and retrieving
 * those that were not present in the old value and/or those whose value has changed.
 *
 * @param newValue - The new object value.
 * @param oldValue - The old object value.
 *
 * @returns An array of keys.
 * @public
 */
function getNewAndUpdatedKeys(newValue, oldValue) {
    if (newValue === oldValue || !newValue || !oldValue) {
        return [];
    }
    return Object.keys(newValue).filter(key => !(key in oldValue) || newValue[key] !== oldValue[key]);
}
exports.getNewAndUpdatedKeys = getNewAndUpdatedKeys;
/**
 * Ensures that the given condition is met in all the non-undefined entries of the object.
 *
 * @param object - The object to check if every item meets the given condition.
 * @param condition - The condition to check in each one of the entries of the object.
 *
 * @returns True when all the entries pass the condition. False otherwise.
 * @public
 */
function every(object, condition) {
    return Object.entries(object)
        .filter(([, value]) => value !== undefined)
        .every(([key, value], index) => condition(key, value, index));
}
exports.every = every;
/**
 * Flattens recursively the passed object to a one level object.
 *
 * @param object - The object to flatten.
 * @returns The flattened object.
 */
function flatObject(object) {
    const flattenedObject = {};
    forEach(object, (key, value) => {
        if ((0, typeguards_1.isObject)(value)) {
            Object.assign(flattenedObject, flatObject(value));
        }
        else {
            flattenedObject[key] = value;
        }
    });
    return flattenedObject;
}
exports.flatObject = flatObject;
