// @empathyco/x-components/js/components/base-keyboard-navigation.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { DirectionalFocusNavigationService } from '../services/directional-focus-navigation.service.js';
import { XOn } from './decorators/bus.decorators.js';

/**
 * Base component to handle keyboard navigation for elements inside it. It has a required slot to
 * include the navigable elements.
 *
 * @remarks
 * The component can be customized through props: an array of navigationHijacker objects, which
 * contains: the xEvent to listen to, the moduleName in charge of emitting the event and to which
 * direction it should react to; to take control of the navigation and eventsForDirectionLimit, to
 * emit an xEvent when reaching the navigation limit in any direction.
 *
 * @public
 */
let BaseKeyboardNavigation = class BaseKeyboardNavigation extends Vue {
    mounted() {
        // TODO Replace this with injection
        this.navigationService = new DirectionalFocusNavigationService(this.$el);
    }
    /**
     * Get the navigation hijacker events.
     *
     * @remarks
     * If the same {@link XEvent} is defined multiple times it is only inserted once.
     *
     * @returns The events to hijack the navigation.
     */
    get navigationHijackerEvents() {
        const eventsSet = this.navigationHijacker.map(({ xEvent }) => xEvent);
        return Array.from(new Set(eventsSet));
    }
    /**
     * Trigger navigation if this component is in control of it.
     *
     * @param eventPayload - The {@link WirePayload.eventPayload}.
     * @param metadata - The {@link WirePayload.metadata}.
     * @public
     */
    triggerNavigation(eventPayload, metadata) {
        if (this.hasToTakeNavigationControl(eventPayload, metadata)) {
            this.focusNextNavigableElement(eventPayload);
        }
    }
    /**
     * Checks if the component has to take control of the keyboard navigation.
     *
     * @param eventPayload - The {@link ArrowKey}.
     * @param metadata - The {@link WireMetadata}.
     *
     * @returns Whether the component needs to take control of the keyboard navigation or not.
     * @internal
     */
    hasToTakeNavigationControl(eventPayload, metadata) {
        return this.navigationHijacker.some(({ moduleName, direction }) => moduleName === metadata.moduleName && direction === eventPayload);
    }
    /**
     * Focus the next navigable element returned by the navigation service.
     *
     * @param direction - The navigation direction.
     * @internal
     */
    focusNextNavigableElement(direction) {
        const dir = typeof direction === 'object' ? direction.key : direction;
        const nextElementToFocus = this.navigationService?.navigateTo(dir);
        if (this.elementToFocus !== nextElementToFocus) {
            this.elementToFocus = nextElementToFocus;
            this.elementToFocus.focus();
        }
        else {
            this.emitDirectionalLimitReached(dir);
            this.elementToFocus = undefined;
        }
    }
    /**
     * Emit the {@link XEvent} associated to the navigation's direction when reaching its limit.
     *
     * @param direction - The navigation direction.
     * @internal
     */
    emitDirectionalLimitReached(direction) {
        const xEvent = this.eventsForDirectionLimit?.[direction];
        if (xEvent) {
            this.$x.emit(xEvent, undefined, { target: this.elementToFocus });
        }
    }
};
__decorate([
    Prop({
        default: () => [
            { xEvent: 'UserPressedArrowKey', moduleName: 'searchBox', direction: 'ArrowDown' }
        ]
    })
], BaseKeyboardNavigation.prototype, "navigationHijacker", void 0);
__decorate([
    Prop({ default: () => ({ ArrowUp: 'UserReachedEmpathizeTop' }) })
], BaseKeyboardNavigation.prototype, "eventsForDirectionLimit", void 0);
__decorate([
    XOn(component => component.navigationHijackerEvents)
], BaseKeyboardNavigation.prototype, "triggerNavigation", null);
BaseKeyboardNavigation = __decorate([
    Component
], BaseKeyboardNavigation);
var script = BaseKeyboardNavigation;

export { script as default };
