// @empathyco/x-components/js/components/base-grid.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { toKebabCase } from '../utils/string.js';
import { XInject } from './decorators/injection.decorators.js';
import { LIST_ITEMS_KEY } from './decorators/injection.consts.js';

/**
 * Grid component that is able to render different items based on their modelName value. In order
 * to achieve this, it exposes a scopedSlot for each different modelName. In case the items used
 * do not have modelName property, the default slot is used instead. It has a required property:
 * the `items` to render; and an optional one: the number `columns` the grid is divided in. If the
 * number of columns is not specified, the grid automatically fills the rows with as many columns
 * as it can fit.
 *
 * @public
 */
let BaseGrid = class BaseGrid extends Vue {
    /**
     * It returns the items passed as props or the injected ones.
     *
     * @returns List of grid items.
     *
     * @public
     */
    get computedItems() {
        return (this.items ??
            this.injectedListItems ??
            //TODO: add here logger
            //eslint-disable-next-line no-console
            console.warn('It is necessary to pass a prop or inject the list of filters'));
    }
    /**
     * CSS class based on the column property value so items inside the grid can fill different
     * amount of columns or rows based on how many columns the grid is divided into.
     *
     * @returns CSS class with the column property value.
     *
     * @internal
     */
    get cssClasses() {
        return this.columns ? `x-grid--cols-${this.columns}` : 'x-grid--cols-auto';
    }
    /**
     * CSSStyleDeclaration object specifying the number of columns the grid is divided into based on
     * the column property value.
     *
     * @returns A CSSStyleDeclaration to use as the style attribute.
     *
     * @internal
     */
    get style() {
        return {
            gridTemplateColumns: this.columns
                ? `repeat(${this.columns}, minmax(0, 1fr))`
                : 'repeat(auto-fill, minmax(var(--x-size-min-width-grid-item, 150px), 1fr))'
        };
    }
    /**
     * Maps the item to an object containing: the `item`, its `CSS class` and its slot name.
     *
     * @returns An array of objects containing the item and its CSS class.
     *
     * @internal
     */
    get gridItems() {
        return this.computedItems.map(item => {
            const slotName = toKebabCase(item.modelName);
            return {
                slotName,
                item,
                cssClass: `x-base-grid__${slotName}`
            };
        });
    }
};
__decorate([
    Prop({ default: 'ul' })
], BaseGrid.prototype, "animation", void 0);
__decorate([
    Prop({ default: 0 })
], BaseGrid.prototype, "columns", void 0);
__decorate([
    Prop()
], BaseGrid.prototype, "items", void 0);
__decorate([
    XInject(LIST_ITEMS_KEY)
], BaseGrid.prototype, "injectedListItems", void 0);
BaseGrid = __decorate([
    Component({})
], BaseGrid);
var script = BaseGrid;

export { script as default };
