// @empathyco/x-components/js/components/base-dropdown.vue.js
import script from './base-dropdown.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(
    "div",
    {
      staticClass: "x-dropdown",
      class: _vm.dropdownCSSClasses,
      on: {
        keydown: [
          _vm.updateSearchBuffer,
          function ($event) {
            if (
              !$event.type.indexOf("key") &&
              _vm._k($event.keyCode, "down", 40, $event.key, [
                "Down",
                "ArrowDown",
              ])
            ) {
              return null
            }
            $event.preventDefault();
            return _vm.highlightNextItem.apply(null, arguments)
          },
          function ($event) {
            if (
              !$event.type.indexOf("key") &&
              _vm._k($event.keyCode, "up", 38, $event.key, ["Up", "ArrowUp"])
            ) {
              return null
            }
            $event.preventDefault();
            return _vm.highlightPreviousItem.apply(null, arguments)
          },
        ],
      },
    },
    [
      _c(
        "button",
        {
          staticClass: "x-dropdown__toggle",
          attrs: { "aria-haspopup": "listbox", "data-test": "dropdown-toggle" },
          on: {
            click: _vm.toggle,
            keydown: function ($event) {
              if (
                !$event.type.indexOf("key") &&
                _vm._k($event.keyCode, "up", 38, $event.key, [
                  "Up",
                  "ArrowUp",
                ]) &&
                _vm._k($event.keyCode, "down", 40, $event.key, [
                  "Down",
                  "ArrowDown",
                ])
              ) {
                return null
              }
              $event.preventDefault();
              $event.stopPropagation();
              return _vm.open.apply(null, arguments)
            },
          },
        },
        [
          _vm.$scopedSlots.toggle
            ? _vm._t(
                "toggle",
                function () {
                  return [_vm._v("\n      " + _vm._s(_vm.value) + "\n    ")]
                },
                { isOpen: _vm.isOpen, item: _vm.value }
              )
            : _vm._t(
                "item",
                function () {
                  return [_vm._v(_vm._s(_vm.value))]
                },
                { item: _vm.value }
              ),
        ],
        2
      ),
      _vm._v(" "),
      _c(_vm.animation, { tag: "component" }, [
        _vm.isOpen
          ? _c(
              "ul",
              {
                staticClass: "x-dropdown__items-list",
                attrs: {
                  "aria-expanded": _vm.isOpen.toString(),
                  role: "listbox",
                  tabIndex: "0",
                },
                on: {
                  keydown: [
                    function ($event) {
                      if (
                        !$event.type.indexOf("key") &&
                        _vm._k(
                          $event.keyCode,
                          "end",
                          undefined,
                          $event.key,
                          undefined
                        )
                      ) {
                        return null
                      }
                      return _vm.highlightLastItem.apply(null, arguments)
                    },
                    function ($event) {
                      if (
                        !$event.type.indexOf("key") &&
                        _vm._k($event.keyCode, "esc", 27, $event.key, [
                          "Esc",
                          "Escape",
                        ])
                      ) {
                        return null
                      }
                      return _vm.close.apply(null, arguments)
                    },
                    function ($event) {
                      if (
                        !$event.type.indexOf("key") &&
                        _vm._k(
                          $event.keyCode,
                          "home",
                          undefined,
                          $event.key,
                          undefined
                        )
                      ) {
                        return null
                      }
                      return _vm.highlightFirstItem.apply(null, arguments)
                    },
                  ],
                },
              },
              _vm._l(_vm.items, function (item, index) {
                return _c(
                  "li",
                  {
                    key: item.id || item,
                    staticClass: "x-dropdown__list-item",
                  },
                  [
                    _c(
                      "button",
                      {
                        ref: "itemButtons",
                        refInFor: true,
                        staticClass: "x-dropdown__item",
                        class: _vm.itemsCSSClasses[index],
                        attrs: {
                          "aria-selected": (
                            index === _vm.highlightedItemIndex
                          ).toString(),
                          "aria-current": (item === _vm.value).toString(),
                          "data-test": "dropdown-item",
                          role: "option",
                          tabindex: "-1",
                        },
                        on: {
                          click: function ($event) {
                            return _vm.emitSelectedItemChanged(item)
                          },
                        },
                      },
                      [
                        _vm._t(
                          "item",
                          function () {
                            return [
                              _vm._v(
                                "\n            " + _vm._s(item) + "\n          "
                              ),
                            ]
                          },
                          {
                            isHighlighted: index === _vm.highlightedItemIndex,
                            isSelected: item === _vm.value,
                            item: item,
                          }
                        ),
                      ],
                      2
                    ),
                  ]
                )
              }),
              0
            )
          : _vm._e(),
      ]),
    ],
    1
  )
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-7f6f2f40_0", { source: ".x-dropdown[data-v-7f6f2f40] {\n  position: relative;\n}\n.x-dropdown__items-list[data-v-7f6f2f40] {\n  z-index: 1;\n  list-style: none;\n  position: absolute;\n  padding: 0;\n  margin: 0;\n  top: calc(100% + var(--x-size-gap-dropdown-default, 0));\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-7f6f2f40";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
