// @empathyco/x-components/js/components/base-dropdown.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import { Prop, Watch, Component } from 'vue-property-decorator';
import Vue from 'vue';
import { normalizeString } from '../utils/normalize.js';
import { isInRange } from '../utils/number.js';
import { debounce } from '../utils/debounce.js';
import { NoElement } from './no-element.js';

/**
 * Dropdown component that mimics a Select element behavior, but with the option
 * to customize the toggle button and each item contents.
 *
 * @public
 */
let BaseDropdown = class BaseDropdown extends Vue {
    constructor() {
        super(...arguments);
        /**
         * Property to track whether the dropdown is expanded and displaying the full
         * list of items, or closed.
         *
         * @internal
         */
        this.isOpen = false;
        /**
         * Index of the element that has the focus in the list. -1 means no element has focus.
         *
         * @internal
         */
        this.highlightedItemIndex = -1;
        /**
         * String to search for the first element that starts with it.
         *
         * @internal
         */
        this.searchBuffer = '';
    }
    /**
     * Dynamic CSS classes to add to the dropdown root element.
     *
     * @returns An object containing the CSS classes to add to the dropdown root element.
     * @internal
     */
    get dropdownCSSClasses() {
        return {
            'x-dropdown--is-open': this.isOpen
        };
    }
    /**
     * Dynamic CSS classes to add to each one of the items.
     *
     * @returns An object containing the CSS classes to add to each item.
     * @internal
     */
    get itemsCSSClasses() {
        return this.items.map((item, index) => {
            return {
                'x-dropdown__item--is-selected': this.value === item,
                'x-dropdown__item--is-highlighted': this.highlightedItemIndex === index
            };
        });
    }
    /**
     * If the dropdown is destroyed before removing the document listeners, it ensures that they
     * are removed too.
     *
     * @internal
     */
    beforeDestroy() {
        this.removeDocumentCloseListeners();
    }
    /**
     * Opens the dropdown.
     *
     * @internal
     */
    open() {
        this.isOpen = true;
    }
    /**
     * Closes the dropdown.
     *
     * @internal
     */
    close() {
        this.isOpen = false;
    }
    /**
     * If the dropdown is opened it closes it. If it is closed it opens it.
     *
     * @internal
     */
    toggle() {
        this.isOpen = !this.isOpen;
    }
    /**
     * Emits the event that the selected item has changed.
     *
     * @param item - The new selected item.
     * @internal
     */
    emitSelectedItemChanged(item) {
        this.$emit('change', item);
        this.close();
    }
    /**
     * Highlights the item after the one that is currently highlighted.
     *
     * @internal
     */
    highlightNextItem() {
        this.open();
        this.highlightedItemIndex = (this.highlightedItemIndex + 1) % this.items.length;
    }
    /**
     * Highlights the item before the one that is currently highlighted.
     *
     * @internal
     */
    highlightPreviousItem() {
        this.open();
        this.highlightedItemIndex =
            this.highlightedItemIndex > 0 ? this.highlightedItemIndex - 1 : this.items.length - 1;
    }
    /**
     * Highlights the first of the provided items.
     *
     * @internal
     */
    highlightFirstItem() {
        this.highlightedItemIndex = 0;
    }
    /**
     * Highlights the last of the provided items.
     *
     * @internal
     */
    highlightLastItem() {
        this.highlightedItemIndex = this.items.length - 1;
    }
    /**
     * Updates the variable that is used to search in the filters.
     *
     * @param event - The event coming from the user typing.
     * @internal
     */
    updateSearchBuffer(event) {
        if (/^\w$/.test(event.key)) {
            const key = event.key;
            this.searchBuffer += key;
            this.restartResetSearchTimeout();
        }
    }
    /**
     * Highlights the item that matches the search buffer. To do so it checks the list buttons
     * text content. It highlights items folowing this priority:
     * - If an element is already highlighted, it starts searching from that element.
     * - If no element is found starting from the previously highlighted, it returns the first one.
     * - If no element is found matching the search query it highlights the first element.
     *
     * @param search - The search string to find in the HTML.
     * @internal
     */
    highlightMatchingItem(search) {
        if (search) {
            const normalizedSearch = normalizeString(search);
            const matchingIndices = this.$refs.itemButtons.reduce((matchingIndices, button, index) => {
                const safeButtonWordCharacters = button.textContent.replace(/[^\w]/g, '');
                const normalizedButtonText = normalizeString(safeButtonWordCharacters);
                if (normalizedButtonText.startsWith(normalizedSearch)) {
                    matchingIndices.push(index);
                }
                return matchingIndices;
            }, []);
            this.highlightedItemIndex =
                // First matching item starting to search from the current highlighted element
                matchingIndices.find(index => index >= this.highlightedItemIndex) ??
                    // First matching item
                    matchingIndices[0] ??
                    // First item
                    0;
        }
    }
    /**
     * Resets the search buffer.
     *
     * @internal
     */
    resetSearch() {
        this.searchBuffer = '';
    }
    /**
     * Updates the debounced function to reset the search.
     *
     * @param searchTimeoutMs - The new milliseconds that have to pass without typing before
     * resetting the search.
     * @internal
     */
    updateSearchTimeout(searchTimeoutMs) {
        // eslint-disable-next-line @typescript-eslint/unbound-method
        this.restartResetSearchTimeout = debounce(this.resetSearch, searchTimeoutMs);
    }
    /**
     * Focuses the DOM element which matches the `highlightedItemIndex`.
     *
     * @param highlightedItemIndex - The index of the HTML element to focus.
     * @internal
     */
    focusHighlightedItem(highlightedItemIndex) {
        this.$nextTick(() => {
            if (this.$refs.itemButtons && isInRange(highlightedItemIndex, [0, this.items.length - 1])) {
                const newItem = this.$refs.itemButtons[this.highlightedItemIndex];
                newItem.focus();
            }
        });
    }
    /**
     * When the dropdown is open it sets the focused element to the one that is selected.
     *
     * @param isOpen - True if the dropdown is open, false otherwise.
     * @internal
     */
    updateHighlightedItem(isOpen) {
        if (isOpen) {
            this.highlightedItemIndex = this.value === null ? 0 : this.items.indexOf(this.value);
        }
        else {
            this.highlightedItemIndex = -1;
        }
    }
    /**
     * Adds and removes listeners to close the dropdown when it loses the focus.
     *
     * @param isOpen - True if the dropdown is open, false otherwise.
     * @internal
     */
    syncCloseListeners(isOpen) {
        /*
         * Because there is an issue with Firefox in macOS and Safari that doesn't focus the target
         * element of the `mousedown` events, the `focusout` event `relatedTarget` property can't be
         * used to detect whether or not the user has blurred the dropdown. The hack here is to use
         * document listeners that have the side effect of losing the focus.
         */
        if (isOpen) {
            this.addDocumentCloseListeners();
        }
        else {
            this.removeDocumentCloseListeners();
        }
    }
    /**
     * Adds listeners to the document element to detect if the focus has moved out from the
     * dropdown.
     *
     * @internal
     */
    addDocumentCloseListeners() {
        /* eslint-disable @typescript-eslint/unbound-method */
        document.addEventListener('mousedown', this.closeIfEventIsOutOfDropdown);
        document.addEventListener('touchstart', this.closeIfEventIsOutOfDropdown);
        document.addEventListener('focusin', this.closeIfEventIsOutOfDropdown);
        /* eslint-enable @typescript-eslint/unbound-method */
    }
    /**
     * Removes the listeners of the document element to detect if the focus has moved out from the
     * dropdown.
     *
     * @internal
     */
    removeDocumentCloseListeners() {
        /* eslint-disable @typescript-eslint/unbound-method */
        document.removeEventListener('mousedown', this.closeIfEventIsOutOfDropdown);
        document.removeEventListener('touchstart', this.closeIfEventIsOutOfDropdown);
        document.removeEventListener('focusin', this.closeIfEventIsOutOfDropdown);
        /* eslint-enable @typescript-eslint/unbound-method */
    }
    /**
     * Closes the dropdown if the passed event has happened on an element out of the dropdown.
     *
     * @param event - The event to check if it has happen out of the dropdown component.
     */
    closeIfEventIsOutOfDropdown(event) {
        if (!this.$el.contains(event.target)) {
            this.close();
        }
    }
};
__decorate([
    Prop({ required: true })
], BaseDropdown.prototype, "items", void 0);
__decorate([
    Prop({ required: true })
], BaseDropdown.prototype, "value", void 0);
__decorate([
    Prop({ default: 'NoElement' })
], BaseDropdown.prototype, "animation", void 0);
__decorate([
    Prop({ default: 1000 })
], BaseDropdown.prototype, "searchTimeoutMs", void 0);
__decorate([
    Watch('searchBuffer')
], BaseDropdown.prototype, "highlightMatchingItem", null);
__decorate([
    Watch('searchTimeoutMs', { immediate: true })
], BaseDropdown.prototype, "updateSearchTimeout", null);
__decorate([
    Watch('highlightedItemIndex', { immediate: true })
], BaseDropdown.prototype, "focusHighlightedItem", null);
__decorate([
    Watch('isOpen')
], BaseDropdown.prototype, "updateHighlightedItem", null);
__decorate([
    Watch('isOpen')
], BaseDropdown.prototype, "syncCloseListeners", null);
BaseDropdown = __decorate([
    Component({
        components: {
            NoElement
        },
        model: {
            event: 'change'
        }
    })
], BaseDropdown);
var script = BaseDropdown;

export { script as default };
