// @empathyco/x-components/js/components/sliding-panel.vue.js
import script from './sliding-panel.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _vm.$slots.default
    ? _c(
        "div",
        {
          staticClass: "x-sliding-panel",
          class: _vm.cssClasses,
          attrs: { "data-test": "sliding-panel" },
        },
        [
          _vm.showButtons
            ? _c(
                "button",
                {
                  staticClass:
                    "x-sliding-panel__button x-sliding-panel__button-left x-button",
                  class: _vm.buttonClass,
                  attrs: { "data-test": "sliding-panel-left-button" },
                  on: { click: _vm.scrollLeft },
                },
                [
                  _vm._t("sliding-panel-left-button", function () {
                    return [_vm._v("ᐸ")]
                  }),
                ],
                2
              )
            : _vm._e(),
          _vm._v(" "),
          _c(
            "div",
            {
              ref: "scrollContainer",
              staticClass: "x-list x-list--horizontal x-sliding-panel__scroll",
              attrs: { "data-test": "sliding-panel-scroll" },
              on: {
                scroll: _vm.debouncedUpdateScrollPosition,
                transitionend: _vm.debouncedUpdateScrollPosition,
                animationend: _vm.debouncedUpdateScrollPosition,
              },
            },
            [_vm._t("default")],
            2
          ),
          _vm._v(" "),
          _vm.showButtons
            ? _c(
                "button",
                {
                  staticClass:
                    "x-sliding-panel__button x-sliding-panel__button-right x-button",
                  class: _vm.buttonClass,
                  attrs: { "data-test": "sliding-panel-right-button" },
                  on: { click: _vm.scrollRight },
                },
                [
                  _vm._t("sliding-panel-right-button", function () {
                    return [_vm._v("ᐳ")]
                  }),
                ],
                2
              )
            : _vm._e(),
        ]
      )
    : _vm._e()
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-1c771718_0", { source: ".x-sliding-panel[data-v-1c771718] {\n  align-items: center;\n  display: flex;\n  flex-flow: row nowrap;\n  height: 100%;\n  position: relative;\n}\n.x-sliding-panel__button[data-v-1c771718] {\n  opacity: 0;\n  pointer-events: none;\n  position: absolute;\n  transition: all ease-out 0.2s;\n  z-index: 2;\n  /* To overlay the design system gradient with z-index:1 */\n}\n.x-sliding-panel__button-left[data-v-1c771718] {\n  left: 0;\n}\n.x-sliding-panel__button-right[data-v-1c771718] {\n  right: 0;\n}\n.x-sliding-panel__scroll[data-v-1c771718] {\n  display: flex;\n  flex: 100%;\n  flex-flow: row nowrap;\n  overflow: auto;\n  scrollbar-width: none;\n  -ms-overflow-style: none;\n}\n.x-sliding-panel__scroll[data-v-1c771718]::-webkit-scrollbar {\n  display: none;\n}\n.x-sliding-panel:not(.x-sliding-panel--show-buttons-on-hover):not(.x-sliding-panel--at-start) .x-sliding-panel__button-left[data-v-1c771718] {\n  opacity: 1;\n  pointer-events: all;\n}\n.x-sliding-panel:not(.x-sliding-panel--show-buttons-on-hover):not(.x-sliding-panel--at-end) .x-sliding-panel__button-right[data-v-1c771718] {\n  opacity: 1;\n  pointer-events: all;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-1c771718";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
