// @empathyco/x-components/js/components/sliding-panel.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { Debounce } from './decorators/debounce.decorators.js';

/**
 * This component allows for any other component or element inside it to be horizontally
 * navigable. It also implements customizable buttons as well as other minor customizations to its
 * general behavior.
 *
 * @public
 */
let SlidingPanel = class SlidingPanel extends Vue {
    constructor() {
        super(...arguments);
        /**
         * Indicates if the scroll is at the start of the sliding panel.
         *
         * @internal
         */
        this.isScrollAtStart = true;
        /**
         * Indicates if the scroll is at the end of the sliding panel.
         *
         * @internal
         */
        this.isScrollAtEnd = true;
    }
    /**
     * CSS classes to apply based on the scroll position.
     *
     * @returns The CSS classes to apply.
     *
     * @internal
     */
    get cssClasses() {
        return {
            'x-sliding-panel--at-start': this.isScrollAtStart,
            'x-sliding-panel--at-end': this.isScrollAtEnd
        };
    }
    /**
     * Initialises browser platform code:
     * - Creates a mutation observer to detect content changes and reset scroll position.
     * - Stores initial size and scroll position values.
     *
     * @internal
     */
    mounted() {
        // eslint-disable-next-line @typescript-eslint/unbound-method
        const resizeObserver = new ResizeObserver(this.debouncedUpdateScrollPosition);
        resizeObserver.observe(this.$el);
        // eslint-disable-next-line @typescript-eslint/unbound-method
        const contentChangedObserver = new MutationObserver(this.restoreAndUpdateScroll);
        this.$watch(() => this.resetOnContentChange, shouldReset => {
            if (shouldReset) {
                contentChangedObserver.observe(this.$refs.scrollContainer, {
                    attributes: false,
                    childList: true,
                    subtree: true,
                    characterData: false
                });
            }
            else {
                contentChangedObserver.disconnect();
            }
        }, { immediate: true });
        this.$on('hook:beforeDestroy', () => {
            contentChangedObserver.disconnect();
            resizeObserver.disconnect();
        });
        this.updateScrollPosition();
    }
    /**
     * Resets the scroll and updates the values of the scroll for the buttons to react.
     *
     * @internal
     */
    restoreAndUpdateScroll() {
        this.$refs.scrollContainer.scroll({ left: 0, behavior: 'smooth' });
        this.updateScrollPosition();
    }
    /**
     * Updates the values of the scroll positions to show or hide the buttons depending on it.
     *
     * @internal
     */
    updateScrollPosition() {
        const { scrollLeft, clientWidth, scrollWidth } = this.$refs.scrollContainer;
        this.isScrollAtStart = !scrollLeft;
        /* The 2 px extra is to fix some cases in some resolutions where the scroll + client size is
         *  less than the scroll width even when the scroll is at the end */
        this.isScrollAtEnd = scrollLeft + clientWidth + 2 >= scrollWidth;
    }
    /**
     * Debounced version of the {@link SlidingPanel.updateScrollPosition | updateScrollPosition}
     * method.
     *
     * @internal
     */
    debouncedUpdateScrollPosition() {
        this.updateScrollPosition();
    }
    /**
     * Scrolls the wrapper element to the left.
     *
     * @internal
     */
    scrollLeft() {
        this.scrollTo(-this.$refs.scrollContainer.clientWidth);
    }
    /**
     * Scrolls the wrapper element to the right.
     *
     * @internal
     */
    scrollRight() {
        this.scrollTo(this.$refs.scrollContainer.clientWidth);
    }
    /**
     * Scrolls the wrapper element towards the provided scroll value.
     *
     * @param scrollValue - The value the scroll will go towards.
     *
     * @internal
     */
    scrollTo(scrollValue) {
        this.$refs.scrollContainer.scrollBy({
            left: scrollValue * this.scrollFactor,
            behavior: 'smooth'
        });
    }
};
__decorate([
    Prop({ default: 0.7 })
], SlidingPanel.prototype, "scrollFactor", void 0);
__decorate([
    Prop({ default: true })
], SlidingPanel.prototype, "showButtons", void 0);
__decorate([
    Prop({ default: true })
], SlidingPanel.prototype, "resetOnContentChange", void 0);
__decorate([
    Prop()
], SlidingPanel.prototype, "buttonClass", void 0);
__decorate([
    Debounce(100, { leading: true })
], SlidingPanel.prototype, "restoreAndUpdateScroll", null);
__decorate([
    Debounce(100, { leading: true })
], SlidingPanel.prototype, "debouncedUpdateScrollPosition", null);
SlidingPanel = __decorate([
    Component
], SlidingPanel);
var script = SlidingPanel;

export { script as default };
