// @empathyco/x-components/js/x-modules/facets/wiring.js
import { createRawFilters } from '../../utils/filters.js';
import { wireService, wireServiceWithoutPayload } from '../../wiring/wires.factory.js';
import { mapWire } from '../../wiring/wires.operators.js';
import { createWiring } from '../../wiring/wiring.utils.js';
import { DefaultFacetsService } from './service/facets.service.js';

/**
 * Wires factory for {@link DefaultFacetsService}.
 */
const wireFacetsService = wireService(DefaultFacetsService.instance);
/**
 * Wires without payload factory for {@link DefaultFacetsService}.
 */
const wireFacetsServiceWithoutPayload = wireServiceWithoutPayload(DefaultFacetsService.instance);
/**
 * Saves the facets contained in the `search` group, removing the previous ones, and keeping the
 * previous filters selected state.
 *
 * @public
 */
const updateFacetsGroupWithSearchFacetsWire = mapWire(wireFacetsService('updateFacets'), (facets) => ({
    facets,
    id: 'search'
}));
/**
 * Saves the facets contained in the group, removing the previous ones, and keeping the new filters
 * selected state.
 *
 * @public
 */
const setFacetsGroupWire = wireFacetsService('setFacets');
/**
 * Toggles the selected state of a filter.
 *
 * @public
 */
const toggleFilterWire = wireFacetsService('toggle');
/**
 * Deselects all the filters. Optionally, it can accept a list of facets ids as payload, and it will
 * only deselect the filters from those facets.
 *
 * @public
 */
const clearFiltersWire = wireFacetsService('clearFilters');
/**
 * Deselects all selected filters.
 *
 * @public
 */
const clearAllFiltersWire = wireFacetsServiceWithoutPayload('clearFilters');
/**
 * Selects the filter passed by payload.
 *
 * @public
 */
const selectFilterWire = wireFacetsService('select');
/**
 * Saves the params from the url.
 *
 * @public
 */
const setFiltersFromUrl = mapWire(wireFacetsService('select'), ({ filter }) => createRawFilters(filter));
/**
 * Saves the preselected filters.
 *
 * @public
 */
const updatePreselectedFilters = wireFacetsService('updatePreselectedFilters');
/**
 * Selects the preselected filters stored in the state.
 *
 * @public
 */
const selectPreselectedFilterWire = wireFacetsService('selectPreselectedFilters');
/**
 * Wiring configuration for the {@link FacetsXModule | facets module}.
 *
 * @internal
 */
const facetsWiring = createWiring({
    ParamsLoadedFromUrl: {
        // TODO: move this logic to Facets Service
        clearAllFiltersWire,
        setFiltersFromUrl
    },
    PreselectedFiltersProvided: {
        updatePreselectedFilters
    },
    FacetsChanged: {
        updateFacetsGroupWithSearchFacetsWire
    },
    FacetsGroupProvided: {
        setFacetsGroupWire
    },
    UserChangedExtraParams: {
        clearAllFiltersWire
    },
    UserClickedAFilter: {
        toggleFilterWire
    },
    UserClickedClearAllFilters: {
        clearFiltersWire
    },
    UserModifiedEditableNumberRangeFilter: {
        selectFilterWire
    },
    UserClickedAllFilter: {
        clearFiltersWire
    },
    UserClearedQuery: {
        clearAllFiltersWire
    },
    UserClickedOpenX: {
        selectPreselectedFilterWire
    }
});

export { facetsWiring };
