// @empathyco/x-components/js/x-modules/facets/components/facets/facets.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import { map, objectFilter } from '@empathyco/x-utils';
import { Prop, Component } from 'vue-property-decorator';
import { Getter } from '../../../../components/decorators/store.decorators.js';
import { xComponentMixin } from '../../../../components/x-component.mixin.js';
import { toKebabCase } from '../../../../utils/string.js';
import FacetsMixin from '../facets.mixin.js';
import { facetsXModule } from '../../x-module.js';

/**
 * This component renders the list of facets stored in the Facets module. Facets can be rendered
 * differently based on their purpose and this can be achieved using the exposed slots:
 * - A default and required slot.
 * - A custom slot for each facet with the facetId as its name. This allows each facet to be
 * rendered differently based on its needs.
 *
 * @remarks It extends {@link FacetsMixin}.
 *
 * @public
 */
let Facets = class Facets extends FacetsMixin {
    /**
     * Transforms a dictionary of Facets including the slot name.
     *
     * @returns A dictionary of facets with the slot name.
     *
     * @internal
     */
    get mappedFacets() {
        return map(this.facetsToRender, (facetId, facet) => ({
            slotName: toKebabCase(facetId),
            facet
        }));
    }
    /**
     * The facets to be rendered after filtering {@link Facets.facets} by
     * {@link Facets.renderableFacets} content.
     *
     * @returns The list of facets to be rendered.
     *
     * @internal
     */
    get facetsToRender() {
        if (!this.renderableFacets) {
            return this.facets;
        }
        else {
            const excludedRegExp = /^!/;
            const facetIds = this.renderableFacets.split(',').map(facetId => facetId.trim());
            const included = [];
            const excluded = [];
            facetIds.forEach(facetId => {
                if (excludedRegExp.test(facetId)) {
                    excluded.push(facetId.replace(excludedRegExp, ''));
                }
                else {
                    included.push(facetId);
                }
            });
            return this.filterFacetsToRender(included, excluded);
        }
    }
    /**
     * Indicates if there are facets available to show.
     *
     * @returns True if there are facets available and false otherwise.
     * @internal
     */
    get hasFacets() {
        return !!Object.keys(this.facetsToRender).length;
    }
    /**
     * Filter facets dictionary retrieving those included and/or removing excluded.
     *
     * @param included - List of facets to render.
     * @param excluded - List of not renderable facets.
     *
     * @returns The filtered list of facets to render.
     *
     * @internal
     */
    filterFacetsToRender(included, excluded) {
        const hasAnyFacetIncluded = included.length > 0;
        return objectFilter(this.facets, facetKey => {
            const isIncluded = included.includes(String(facetKey));
            const isExcluded = excluded.includes(String(facetKey));
            return hasAnyFacetIncluded ? isIncluded && !isExcluded : !isExcluded;
        });
    }
};
__decorate([
    Prop({ default: 'ul' })
], Facets.prototype, "animation", void 0);
__decorate([
    Prop()
], Facets.prototype, "renderableFacets", void 0);
__decorate([
    Getter('facets', 'facets')
], Facets.prototype, "facets", void 0);
Facets = __decorate([
    Component({
        mixins: [xComponentMixin(facetsXModule)]
    })
], Facets);
var script = Facets;

export { script as default };
