// @empathyco/x-components/js/x-modules/facets/components/facets/facets-provider.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Watch, Component } from 'vue-property-decorator';
import '../../../../components/animations/animate-width.vue.js';
import '../../../../components/animations/collapse-height.vue.js';
import '../../../../components/animations/collapse-width.vue.js';
import '../../../../components/animations/cross-fade.vue.js';
import '../../../../components/animations/fade-and-slide.vue.js';
import '../../../../components/animations/staggered-fade-and-slide.vue.js';
import '../../../../components/animations/staggering-transition-group.vue_rollup-plugin-vue_script.vue.js';
import 'vue-runtime-helpers/dist/normalize-component.mjs';
import 'vue-runtime-helpers/dist/inject-style/browser.mjs';
import '../../../../components/animations/animate-clip-path/animate-clip-path.style.scss.js';
import '../../../../components/animations/animate-scale/animate-scale.style.scss.js';
import '../../../../components/animations/animate-translate/animate-translate.style.scss.js';
import '../../../../components/column-picker/base-column-picker-dropdown.vue.js';
import '../../../../components/column-picker/base-column-picker-list.vue.js';
import '../../../../components/column-picker/column-picker.mixin.js';
import '../../../../components/currency/base-currency.vue.js';
import '../../../../components/filters/labels/base-price-filter-label.vue_rollup-plugin-vue_script.vue.js';
import '../../../../components/filters/labels/base-rating-filter-label.vue.js';
import '../../../../components/icons/arrow-down.vue.js';
import '../../../../components/icons/arrow-left.vue.js';
import '../../../../components/icons/arrow-right.vue.js';
import '../../../../components/icons/arrow-up.vue.js';
import '../../../../components/icons/bag.vue.js';
import '../../../../components/icons/bar-code.vue.js';
import '../../../../components/icons/barcode-tiny.vue.js';
import '../../../../components/icons/cart.vue.js';
import '../../../../components/icons/check-tiny.vue.js';
import '../../../../components/icons/check.vue.js';
import '../../../../components/icons/checkbox-card-selected.vue.js';
import '../../../../components/icons/checkbox-card-unselected.vue.js';
import '../../../../components/icons/checkbox-selected.vue.js';
import '../../../../components/icons/checkbox-unselected.vue.js';
import '../../../../components/icons/chevron-down.vue.js';
import '../../../../components/icons/chevron-left.vue.js';
import '../../../../components/icons/chevron-right.vue.js';
import '../../../../components/icons/chevron-tiny-down.vue.js';
import '../../../../components/icons/chevron-tiny-left.vue.js';
import '../../../../components/icons/chevron-tiny-right.vue.js';
import '../../../../components/icons/chevron-tiny-up.vue.js';
import '../../../../components/icons/chevron-up.vue.js';
import '../../../../components/icons/cross-tiny.vue.js';
import '../../../../components/icons/cross.vue.js';
import '../../../../components/icons/curated-check.vue.js';
import '../../../../components/icons/filters.vue.js';
import '../../../../components/icons/grid-1-col.vue.js';
import '../../../../components/icons/grid-2-col.vue.js';
import '../../../../components/icons/grid-2-rows.vue.js';
import '../../../../components/icons/hide.vue.js';
import '../../../../components/icons/history-tiny.vue.js';
import '../../../../components/icons/history.vue.js';
import '../../../../components/icons/light-bulb-off.vue.js';
import '../../../../components/icons/light-bulb-on.vue.js';
import '../../../../components/icons/menu.vue.js';
import '../../../../components/icons/minus-tiny.vue.js';
import '../../../../components/icons/minus.vue.js';
import '../../../../components/icons/nq-1.vue.js';
import '../../../../components/icons/nq-2.vue.js';
import '../../../../components/icons/nq-3.vue.js';
import '../../../../components/icons/nq-4.vue.js';
import '../../../../components/icons/plus.vue.js';
import '../../../../components/icons/search.vue.js';
import '../../../../components/icons/show.vue.js';
import '../../../../components/icons/star.vue.js';
import '../../../../components/icons/search-tiny.vue.js';
import '../../../../components/icons/tag.vue.js';
import '../../../../components/icons/trash-open.vue.js';
import '../../../../components/icons/trash.vue.js';
import '../../../../components/icons/trending-tiny.vue.js';
import '../../../../components/icons/trending.vue.js';
import '../../../../components/icons/user.vue.js';
import '../../../../components/layouts/fixed-header-and-asides-layout.vue.js';
import '../../../../components/layouts/multi-column-max-width-layout.vue.js';
import '../../../../components/layouts/single-column-layout.vue.js';
import '../../../../components/modals/base-events-modal-close.vue.js';
import '../../../../components/modals/base-events-modal.vue.js';
import '../../../../components/modals/base-events-modal-open.vue.js';
import '../../../../components/modals/base-modal.vue.js';
import '../../../../components/modals/base-id-modal.vue.js';
import '../../../../components/modals/base-id-modal-close.vue.js';
import '../../../../components/modals/base-id-modal-open.vue.js';
import '../../../../components/panels/base-header-toggle-panel.vue.js';
import '../../../../components/panels/base-toggle-panel.vue.js';
import '../../../../components/panels/base-id-toggle-panel.vue.js';
import '../../../../components/panels/base-id-toggle-panel-button.vue.js';
import '../../../../components/result/base-result-add-to-cart.vue.js';
import '../../../../components/result/base-result-current-price.vue.js';
import '../../../../components/result/base-result-image.vue.js';
import '../../../../components/result/base-result-link.vue.js';
import '../../../../components/result/base-result-previous-price.vue.js';
import '../../../../components/result/base-result-fallback-image.vue.js';
import '../../../../components/result/base-result-placeholder-image.vue.js';
import '../../../../components/result/base-result-rating.vue.js';
import '../../../../components/scroll/base-scroll.vue.js';
import '../../../../components/scroll/scroll.mixin.js';
import '../../../../components/suggestions/base-suggestion.vue.js';
import '../../../../components/suggestions/base-suggestions.vue.js';
import '../../../../components/auto-progress-bar.vue.js';
import '../../../../components/base-dropdown.vue.js';
import '../../../../components/base-event-button.vue.js';
import '../../../../components/base-grid.vue.js';
import '../../../../components/base-keyboard-navigation.vue.js';
import '../../../../components/base-rating.vue.js';
import '../../../../components/base-variable-column-grid.vue.js';
import '../../../../components/global-x-bus.vue_rollup-plugin-vue_script.vue.js';
import '../../../../components/items-list.vue.js';
import '../../../../components/location-provider.vue_rollup-plugin-vue_script.vue.js';
import '../../../../components/no-element.js';
import '../../../../components/sliding-panel.vue.js';
import '../../../../components/snippet-callbacks.vue.js';
import { XOn } from '../../../../components/decorators/bus.decorators.js';
import 'vue-class-component';
import '../../../../components/items-list-injection.mixin.js';
import { xComponentMixin } from '../../../../components/x-component.mixin.js';
import { areFiltersDifferent } from '../../../../utils/filters.js';
import { facetsXModule } from '../../x-module.js';

/**
 * This component allows to provide facets by prop, to add them to the state of the
 * `Facets X-Module`. These facets will be added to the `Facets X-Module` state together with
 * the facets emitted by the `Search X-Module` through the {@link SearchXEvents.FacetsChanged}
 * event.
 *
 * @public
 */
let FacetsProvider = class FacetsProvider extends Vue {
    constructor() {
        super(...arguments);
        /**
         * Temporarily stores the selected filters from the {@link FacetsProvider.facets} prop.
         * This is necessary to handle the {@link FacetsXEvents.UserChangedSelectedFilters} event.
         *
         * @internal
         */
        this.selectedFilters = null;
    }
    /**
     * A computed property to group the facets and the groupId. This is used by the watcher.
     *
     * @returns The FacetGroup with the facets and the group id.
     *
     * @internal
     */
    get facetsGroup() {
        return { id: this.groupId, facets: this.facets };
    }
    /**
     * Emits the {@link FacetsXEvents.UserChangedSelectedFilters} event when the user changes
     * the selected filters.
     *
     * @param selectedFilters - The new list of selected filters.
     * @internal
     */
    emitSelectedFiltersChanged(selectedFilters) {
        if (this.selectedFilters === null ||
            areFiltersDifferent(this.selectedFilters, selectedFilters)) {
            this.$x.emit('UserChangedSelectedFilters', selectedFilters);
        }
        this.selectedFilters = null;
    }
    /**
     * Emits the {@link FacetsXEvents.FacetsGroupProvided} event with the
     * {@link FacetsProvider.facetsGroup} as payload. It also extracts and saves the selected
     * filters.
     */
    provideFacets() {
        if (this.facetsGroup.facets) {
            this.$x.emit('FacetsGroupProvided', this.facetsGroup);
            this.extractSelectedFilters(this.facets);
        }
    }
    /**
     * Extracts the selected filters from the facets and stores them in the
     * {@link FacetsProvider.selectedFilters} property.
     *
     * @param facets - The facets from whom extract the selected filters.
     * @internal
     */
    extractSelectedFilters(facets) {
        this.selectedFilters = facets
            .flatMap(facet => facet.filters)
            .filter(filter => filter.selected);
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    render() { }
};
__decorate([
    Prop({ default: 'provided-facets' })
], FacetsProvider.prototype, "groupId", void 0);
__decorate([
    Prop({ required: true })
], FacetsProvider.prototype, "facets", void 0);
__decorate([
    XOn('SelectedFiltersChanged')
], FacetsProvider.prototype, "emitSelectedFiltersChanged", null);
__decorate([
    Watch('facetsGroup', { immediate: true })
], FacetsProvider.prototype, "provideFacets", null);
FacetsProvider = __decorate([
    Component({
        mixins: [xComponentMixin(facetsXModule)]
    })
], FacetsProvider);
var script = FacetsProvider;

export { script as default };
