// @empathyco/x-components/js/x-modules/empathize/components/empathize.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import { XOn } from '../../../components/decorators/bus.decorators.js';
import { Debounce } from '../../../components/decorators/debounce.decorators.js';
import { NoElement } from '../../../components/no-element.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { empathizeXModule } from '../x-module.js';

/**
 * Component containing the empathize. It has a required slot to define its content and two props
 * to define when to open and close it: eventsToOpenEmpathize and eventsToCloseEmpathize.
 *
 * @public
 */
let Empathize = class Empathize extends Vue {
    constructor() {
        super(...arguments);
        /**
         * The modal container is open.
         *
         * @internal
         */
        this.isOpen = false;
    }
    /**
     * Open empathize. This method will be executed on any event in
     * {@link Empathize.eventsToOpenEmpathize} and on DOM event `focusin` on Empathize root element.
     *
     * @param payload - The payload of the {@link XEvent}, that is unused in this case.
     * @param metadata - The {@link WireMetadata} of the event, used to emit the Empathize XEvents.
     *
     * @internal
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    open(payload, metadata) {
        this.changeOpenState(true, metadata);
    }
    /**
     * Close empathize. This method will be executed on any event in
     * {@link Empathize.eventsToCloseEmpathize} and on DOM event `focusout` on Empathize root
     * element.
     *
     * @param payload - The payload of the {@link XEvent}, that is unused in this case.
     * @param metadata - The {@link WireMetadata} of the event, used to emit the Empathize XEvents.
     *
     * @internal
     */
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    close(payload, metadata) {
        this.changeOpenState(false, metadata);
    }
    /**
     * Changes the state of {@link Empathize.isOpen} assigning to it the value of `newOpenState`
     * parameter. Also emits the {@link XEvent | XEvents} `EmpathizeOpened` or `EmpathizeClosed` if
     * the state really changes.
     *
     * @param newOpenState - The new state to assign to {@link Empathize.isOpen}.
     * @param metadata - The {@link WireMetadata} to emit the {@link XEvent | XEvents}. If is
     * undefined, a this component is used as source of info for the metadata.
     *
     * @internal
     */
    changeOpenState(newOpenState, metadata) {
        if (this.isOpen !== newOpenState) {
            this.isOpen = newOpenState;
            this.$x.emit(this.isOpen ? 'EmpathizeOpened' : 'EmpathizeClosed', undefined, metadata ?? { moduleName: 'empathize', target: this.$el });
        }
    }
};
__decorate([
    Prop({ default: () => NoElement })
], Empathize.prototype, "animation", void 0);
__decorate([
    Prop({ default: () => ['UserFocusedSearchBox', 'UserIsTypingAQuery', 'UserClickedSearchBox'] })
], Empathize.prototype, "eventsToOpenEmpathize", void 0);
__decorate([
    Prop({
        default: () => [
            'UserClosedEmpathize',
            'UserSelectedASuggestion',
            'UserPressedEnter',
            'UserBlurredSearchBox'
        ]
    })
], Empathize.prototype, "eventsToCloseEmpathize", void 0);
__decorate([
    XOn(component => component.eventsToOpenEmpathize)
], Empathize.prototype, "open", null);
__decorate([
    XOn(component => component.eventsToCloseEmpathize)
], Empathize.prototype, "close", null);
__decorate([
    Debounce(0)
], Empathize.prototype, "changeOpenState", null);
Empathize = __decorate([
    Component({
        mixins: [xComponentMixin(empathizeXModule)]
    })
], Empathize);
var script = Empathize;

export { script as default };
