// @empathyco/x-adapter/dist/cjs/empathy/mappers/response/facets/empathy-number-range-facet.mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyNumberRangeFacetMapper = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
var container_const_1 = require("../../../container/container.const");
var pipe_mappers_1 = require("../../pipe-mappers");
/**
 * Number Range Empathy facet mapper.
 *
 * @public
 */
var EmpathyNumberRangeFacetMapper = /** @class */ (function () {
    function EmpathyNumberRangeFacetMapper(filterMappers, config) {
        this.mapFilter = pipe_mappers_1.pipeMappers.apply(void 0, filterMappers);
        this.facetsConfig = config.mappings.facets;
    }
    EmpathyNumberRangeFacetMapper.prototype.map = function (rawFacet, facet, context) {
        var isDynamic = (this.facetsConfig.named[facet.id] || this.facetsConfig.default).isDynamic;
        if (facet.modelName === 'NumberRangeFacet') {
            Object.assign(facet, {
                filters: this.mapFilters(facet, rawFacet.values, isDynamic, context)
            });
        }
        return facet;
    };
    EmpathyNumberRangeFacetMapper.prototype.mapFilters = function (facet, rawFilters, isDynamic, context) {
        var _this = this;
        if (rawFilters === void 0) { rawFilters = []; }
        var rawFiltersToMap = isDynamic ? [rawFilters[0]] : rawFilters;
        return rawFiltersToMap.map(function (rawFilter) { return _this.mapFilter(rawFilter, { facetId: facet.id }, context); });
    };
    EmpathyNumberRangeFacetMapper = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.numberRangeFilter)),
        tslib_1.__param(1, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.config)),
        tslib_1.__metadata("design:paramtypes", [Array, Object])
    ], EmpathyNumberRangeFacetMapper);
    return EmpathyNumberRangeFacetMapper;
}());
exports.EmpathyNumberRangeFacetMapper = EmpathyNumberRangeFacetMapper;
