// @empathyco/x-adapter/dist/cjs/empathy/mappers/response/facets/empathy-hierarchical-facet.mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyHierarchicalFacetMapper = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
var container_const_1 = require("../../../container/container.const");
var pipe_mappers_1 = require("../../pipe-mappers");
/**
 * Hierarchical Empathy facet mapper.
 *
 * @public
 */
var EmpathyHierarchicalFacetMapper = /** @class */ (function () {
    function EmpathyHierarchicalFacetMapper(filterMappers) {
        this.mapFilter = pipe_mappers_1.pipeMappers.apply(void 0, filterMappers);
    }
    EmpathyHierarchicalFacetMapper.prototype.map = function (rawFacet, facet, context) {
        if (facet.modelName === 'HierarchicalFacet') {
            Object.assign(facet, {
                filters: this.mapFiltersTree(facet, rawFacet.values, context)
            });
        }
        return facet;
    };
    EmpathyHierarchicalFacetMapper.prototype.mapFiltersTree = function (facet, rawFilters, context) {
        var _this = this;
        var filters = [];
        var initFilterProps = { facetId: facet.id, parentId: null };
        rawFilters.forEach(function (rawFilter) {
            _this.mapDeepFilter(rawFilter, initFilterProps, context, filters);
        });
        return filters;
    };
    EmpathyHierarchicalFacetMapper.prototype.mapDeepFilter = function (rawFilter, initFilterProps, context, filters) {
        var _this = this;
        var _a;
        var filter = this.mapFilter(rawFilter, tslib_1.__assign({}, initFilterProps), context);
        filter.children =
            (_a = rawFilter.values) === null || _a === void 0 ? void 0 : _a.map(function (rawFilterChild) { return _this.mapDeepFilter(rawFilterChild, tslib_1.__assign(tslib_1.__assign({}, initFilterProps), { parentId: filter.id }), context, filters); });
        filters.push(filter);
        return filter.id;
    };
    EmpathyHierarchicalFacetMapper = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.hierarchicalFilter)),
        tslib_1.__metadata("design:paramtypes", [Array])
    ], EmpathyHierarchicalFacetMapper);
    return EmpathyHierarchicalFacetMapper;
}());
exports.EmpathyHierarchicalFacetMapper = EmpathyHierarchicalFacetMapper;
