// @empathyco/x-components/js/x-modules/related-tags/wiring.js
import { namespacedWireCommit, namespacedWireDispatch, namespacedWireDispatchWithoutPayload } from '../../wiring/namespaced-wires.factory.js';
import { createWiring } from '../../wiring/wiring.utils.js';

/**
 * `relatedTags` {@link XModuleName | XModule name}.
 *
 * @internal
 */
const moduleName = 'relatedTags';
/**
 * WireCommit for {@link RelatedTagsXModule}.
 *
 * @internal
 */
const wireCommit = namespacedWireCommit(moduleName);
/**
 * WireDispatch for {@link RelatedTagsXModule}.
 *
 * @internal
 */
const wireDispatch = namespacedWireDispatch(moduleName);
/**
 * WireDispatchWithoutPayload for {@link RelatedTagsXModule}.
 *
 * @internal
 */
const wireDispatchWithoutPayload = namespacedWireDispatchWithoutPayload(moduleName);
/**
 * Sets the related tags state `query`.
 *
 * @public
 */
const setRelatedTagsQuery = wireCommit('setQuery');
/**
 * Sets the related tags state `params`.
 *
 * @public
 */
const setRelatedTagsExtraParams = wireCommit('setParams');
/**
 * Requests and stores the related tags.
 *
 * @public
 */
const fetchAndSaveRelatedTagsWire = wireDispatch('fetchAndSaveRelatedTags');
/**
 * Cancels the {@link RelatedTagsActions.fetchAndSaveRelatedTags} request promise.
 *
 * @public
 */
const cancelFetchAndSaveRelatedTagsWire = wireDispatchWithoutPayload('cancelFetchAndSaveRelatedTags');
/**
 * Sets the selected related tags.
 *
 * @public
 */
const toggleRelatedTagWire = wireDispatch('toggleRelatedTag');
/**
 * Clear the selected related tags.
 *
 * @public
 */
const clearSelectedRelatedTags = wireCommit('setSelectedRelatedTags', []);
/**
 * Clear the related tags query.
 *
 * @public
 */
const clearRelatedTagsQuery = wireCommit('setQuery', '');
/**
 * Saves the params from the url.
 *
 * @public
 */
const setUrlParamsWire = wireDispatch('setUrlParams');
/**
 * Wiring configuration for the {@link RelatedTagsXModule | related tags module}.
 *
 * @internal
 */
const relatedTagsWiring = createWiring({
    ParamsLoadedFromUrl: {
        setUrlParamsWire
    },
    UserAcceptedAQuery: {
        setRelatedTagsQuery,
        clearSelectedRelatedTags
    },
    UserPickedARelatedTag: {
        toggleRelatedTagWire
    },
    UserChangedExtraParams: {
        clearSelectedRelatedTags
    },
    RelatedTagsRequestChanged: {
        fetchAndSaveRelatedTagsWire
    },
    UserClearedQuery: {
        cancelFetchAndSaveRelatedTagsWire,
        clearRelatedTagsQuery,
        clearSelectedRelatedTags
    },
    ExtraParamsChanged: {
        setRelatedTagsExtraParams
    }
});

export { cancelFetchAndSaveRelatedTagsWire, clearRelatedTagsQuery, clearSelectedRelatedTags, fetchAndSaveRelatedTagsWire, relatedTagsWiring, setRelatedTagsExtraParams, setRelatedTagsQuery, setUrlParamsWire, toggleRelatedTagWire };
