// @empathyco/x-components/js/components/panels/base-id-toggle-panel.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Watch, Component } from 'vue-property-decorator';
import { XOn } from '../decorators/bus.decorators.js';
import { NoElement } from '../no-element.js';
import __vue_component__ from './base-toggle-panel.vue.js';

/**
 * Simple panel that could receives its initial open state via prop, if not the default is opens
 * and a required prop, named `panelId`, which are responsible of rendering default slot
 * inside a configurable transition.
 *
 * It reacts to `UserClickedPanelToggleButton` event, when their payload matches the component's
 * 'panelId' prop, to handle its open/close state.
 *
 * The default slot offers the possibility to customise the modal content.
 *
 * @public
 */
let BaseIdTogglePanel = class BaseIdTogglePanel extends Vue {
    constructor() {
        super(...arguments);
        /** Whether the toggle panel is open or not. */
        this.isOpen = this.startOpen;
    }
    /**
     * Method to subscribe to the {@link XEventsTypes.UserClickedPanelToggleButton} event.
     *
     * @param panelId - The payload of the {@link XEventsTypes.UserClickedPanelToggleButton} event.
     *
     * @public
     */
    togglePanel(panelId) {
        if (this.panelId === panelId) {
            this.isOpen = !this.isOpen;
        }
    }
    /**
     * Emits the {@link XEventsTypes.TogglePanelStateChanged} event, when the internal state
     * changes.
     *
     * @remarks This event is necessary to link the state with the {@link BaseIdTogglePanelButton}
     * component.
     * @public
     */
    emitStateEvent() {
        this.$x.emit('TogglePanelStateChanged', this.isOpen, {
            id: this.panelId,
            target: this.$el
        });
    }
};
__decorate([
    Prop({ default: true })
], BaseIdTogglePanel.prototype, "startOpen", void 0);
__decorate([
    Prop({ default: () => NoElement })
], BaseIdTogglePanel.prototype, "animation", void 0);
__decorate([
    Prop({ required: true })
], BaseIdTogglePanel.prototype, "panelId", void 0);
__decorate([
    XOn('UserClickedPanelToggleButton')
], BaseIdTogglePanel.prototype, "togglePanel", null);
__decorate([
    Watch('isOpen', { immediate: true })
], BaseIdTogglePanel.prototype, "emitStateEvent", null);
BaseIdTogglePanel = __decorate([
    Component({
        components: {
            BaseTogglePanel: __vue_component__
        }
    })
], BaseIdTogglePanel);
var script = BaseIdTogglePanel;

export { script as default };
