// @empathyco/x-adapter/dist/cjs/empathy/mappers/response/empathy-partial-result.mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyPartialResultMapper = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
var container_const_1 = require("../../container/container.const");
var pipe_mappers_1 = require("../pipe-mappers");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyPartialResultMapper = /** @class */ (function () {
    function EmpathyPartialResultMapper(resultMappers) {
        this.mapResult = pipe_mappers_1.pipeMappers.apply(void 0, resultMappers);
    }
    EmpathyPartialResultMapper.prototype.map = function (rawPartialResult, partialResult, context) {
        var _this = this;
        return Object.assign(partialResult, {
            query: rawPartialResult.suggestion,
            totalResults: rawPartialResult.numFound,
            results: rawPartialResult.docs.map(function (rawResult) { return _this.mapResult(rawResult, {}, context); })
        });
    };
    EmpathyPartialResultMapper = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.results)),
        tslib_1.__metadata("design:paramtypes", [Array])
    ], EmpathyPartialResultMapper);
    return EmpathyPartialResultMapper;
}());
exports.EmpathyPartialResultMapper = EmpathyPartialResultMapper;
