// @empathyco/x-adapter/dist/cjs/empathy/mappers/response/empathy-redirection.mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyRedirectionMapper = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
var container_const_1 = require("../../container/container.const");
var pipe_mappers_1 = require("../pipe-mappers");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyRedirectionMapper = /** @class */ (function () {
    function EmpathyRedirectionMapper(taggingMappers) {
        this.mapTagging = pipe_mappers_1.pipeMappers.apply(void 0, taggingMappers);
    }
    EmpathyRedirectionMapper.prototype.map = function (rawDirect, redirection, context) {
        return Object.assign(redirection, {
            modelName: 'Redirection',
            id: rawDirect.id,
            url: rawDirect.url,
            tagging: {
                click: this.mapTagging(rawDirect.trackable_url, {}, context)
            }
        });
    };
    EmpathyRedirectionMapper = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.queryTagging)),
        tslib_1.__metadata("design:paramtypes", [Array])
    ], EmpathyRedirectionMapper);
    return EmpathyRedirectionMapper;
}());
exports.EmpathyRedirectionMapper = EmpathyRedirectionMapper;
