// @empathyco/x-adapter/dist/cjs/empathy/mappers/response/empathy-promoted.mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyPromotedMapper = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
var container_const_1 = require("../../container/container.const");
var pipe_mappers_1 = require("../pipe-mappers");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyPromotedMapper = /** @class */ (function () {
    function EmpathyPromotedMapper(taggingMappers) {
        this.mapTagging = pipe_mappers_1.pipeMappers.apply(void 0, taggingMappers);
    }
    EmpathyPromotedMapper.prototype.map = function (rawPromoted, promoted, context) {
        return Object.assign(promoted, {
            modelName: 'Promoted',
            id: rawPromoted.id,
            title: rawPromoted.title,
            url: rawPromoted.url,
            image: rawPromoted.imagename,
            tagging: {
                click: this.mapTagging(rawPromoted.trackable_url, {}, context)
            }
        });
    };
    EmpathyPromotedMapper = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.queryTagging)),
        tslib_1.__metadata("design:paramtypes", [Array])
    ], EmpathyPromotedMapper);
    return EmpathyPromotedMapper;
}());
exports.EmpathyPromotedMapper = EmpathyPromotedMapper;
