// @empathyco/x-adapter/dist/cjs/empathy/mappers/response/empathy-banner.mapper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyBannerMapper = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
var container_const_1 = require("../../container/container.const");
var pipe_mappers_1 = require("../pipe-mappers");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyBannerMapper = /** @class */ (function () {
    function EmpathyBannerMapper(taggingMappers) {
        this.mapTagging = pipe_mappers_1.pipeMappers.apply(void 0, taggingMappers);
    }
    EmpathyBannerMapper.prototype.map = function (rawBanner, banner, context) {
        return Object.assign(banner, {
            modelName: 'Banner',
            id: rawBanner.id,
            title: rawBanner.title,
            url: rawBanner.url,
            image: rawBanner.imagename,
            tagging: {
                click: this.mapTagging(rawBanner.trackable_url, {}, context)
            }
        });
    };
    EmpathyBannerMapper = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.queryTagging)),
        tslib_1.__metadata("design:paramtypes", [Array])
    ], EmpathyBannerMapper);
    return EmpathyBannerMapper;
}());
exports.EmpathyBannerMapper = EmpathyBannerMapper;
