// @empathyco/x-logger/dist/cjs/simple-logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var logger_interfaces_1 = require("./logger.interfaces");
var SimpleLogger = /** @class */ (function () {
    function SimpleLogger() {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.instanceTags = [];
        this.instanceTags = args;
    }
    Object.defineProperty(SimpleLogger.prototype, "consoleLevel", {
        get: function () {
            return SimpleLogger.commonConsoleLevel;
        },
        set: function (level) {
            SimpleLogger.commonConsoleLevel = level;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(SimpleLogger.prototype, "serverLevel", {
        get: function () {
            return SimpleLogger.commonServerLevel;
        },
        set: function (level) {
            SimpleLogger.commonServerLevel = level;
        },
        enumerable: true,
        configurable: true
    });
    SimpleLogger.prototype.error = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.log(logger_interfaces_1.LogLevel.error, args);
    };
    SimpleLogger.prototype.warn = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.log(logger_interfaces_1.LogLevel.warn, args);
    };
    SimpleLogger.prototype.info = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.log(logger_interfaces_1.LogLevel.info, args);
    };
    SimpleLogger.prototype.debug = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.log(logger_interfaces_1.LogLevel.debug, args);
    };
    SimpleLogger.prototype.trace = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        this.log(logger_interfaces_1.LogLevel.trace, args);
    };
    SimpleLogger.prototype.child = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        return new (SimpleLogger.bind.apply(SimpleLogger, tslib_1.__spreadArrays([void 0], this.instanceTags, args)))();
    };
    SimpleLogger.prototype.log = function (level, args) {
        if (this.consoleLevel >= level) {
            this.sendLogToConsole.apply(this, tslib_1.__spreadArrays([level], this.instanceTags.concat(args)));
        }
        if (this.serverLevel >= level) {
            this.sendLogToServer.apply(this, tslib_1.__spreadArrays([level], this.instanceTags.concat(args)));
        }
    };
    SimpleLogger.prototype.sendLogToConsole = function (level) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        var consoleFunctionName = logger_interfaces_1.LogLevel[level];
        // To prevent failure on old browsers
        if (console[consoleFunctionName]) {
            console[consoleFunctionName].apply(console, args);
        }
        else {
            console.log.apply(console, args);
        }
    };
    SimpleLogger.prototype.sendLogToServer = function (level) {
        var args = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            args[_i - 1] = arguments[_i];
        }
        console.log.apply(console, tslib_1.__spreadArrays([logger_interfaces_1.LogLevel[level], 'sending to server...'], args));
    };
    SimpleLogger.commonConsoleLevel = logger_interfaces_1.LogLevel.warn;
    SimpleLogger.commonServerLevel = logger_interfaces_1.LogLevel.error;
    return SimpleLogger;
}());
exports.SimpleLogger = SimpleLogger;
