// @empathyco/x-components/js/x-modules/query-suggestions/store/module.js
import { setQuery } from '../../../store/utils/query.utils.js';
import { setStatus } from '../../../store/utils/status-store.utils.js';
import { cancelFetchAndSaveSuggestions, fetchAndSaveSuggestions } from './actions/fetch-and-save-suggestions.action.js';
import { fetchSuggestions } from './actions/fetch-suggestions.action.js';
import { setUrlParams } from './actions/set-url-params.action.js';
import { normalizedQuery } from './getters/normalized-query.getter.js';
import { querySuggestions } from './getters/query-suggestions.getter.js';
import { request } from './getters/request.getter.js';

/**
 * {@link XStoreModule} For the query-suggestions module.
 *
 * @internal
 */
const querySuggestionsXStoreModule = {
    state: () => ({
        query: '',
        suggestions: [],
        status: 'initial',
        config: {
            debounceInMs: 200,
            maxItemsToRequest: 10,
            showExtraSuggestionWithoutFilter: true,
            hideIfEqualsQuery: true
        },
        params: {}
    }),
    getters: {
        request,
        normalizedQuery,
        querySuggestions
    },
    mutations: {
        setQuery,
        setSuggestions(state, suggestions) {
            state.suggestions = suggestions;
        },
        setStatus,
        setParams(state, params) {
            state.params = params;
        }
    },
    actions: {
        cancelFetchAndSaveSuggestions,
        fetchSuggestions,
        fetchAndSaveSuggestions,
        setUrlParams
    }
};

export { querySuggestionsXStoreModule };
