// @empathyco/x-adapter/dist/cjs/empathy/builder/empathy-adapter.builder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmpathyAdapterBuilder = void 0;
var x_storage_service_1 = require("@empathyco/x-storage-service");
var x_deep_merge_1 = require("@empathyco/x-deep-merge");
var inversify_1 = require("inversify");
var container_config_parser_1 = require("../container/container-config-parser");
var container_bindings_1 = require("../container/container.bindings");
var container_const_1 = require("../container/container.const");
var empathy_adapter_1 = require("../empathy.adapter");
var logger_1 = require("../logger");
var empathy_cache_service_1 = require("../services/empathy-cache.service");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyAdapterBuilder = /** @class */ (function () {
    function EmpathyAdapterBuilder(container, defaultBindingsConfig, adapterClass) {
        if (container === void 0) { container = new inversify_1.Container(); }
        if (defaultBindingsConfig === void 0) { defaultBindingsConfig = container_bindings_1.BINDINGS; }
        if (adapterClass === void 0) { adapterClass = empathy_adapter_1.EmpathyAdapter; }
        this.container = container;
        this.defaultBindingsConfig = defaultBindingsConfig;
        this.adapterClass = adapterClass;
        this.initContainerBindings();
        // There is only one config object. As it has been binded with inversify toConstant() method, we are just getting a reference to it here
        this.config = container.get(container_const_1.DEPENDENCIES.config);
    }
    EmpathyAdapterBuilder.prototype.enableCache = function (_a) {
        var _b = _a === void 0 ? {} : _a, _c = _b.storageService, storageService = _c === void 0 ? typeof localStorage !== 'undefined'
            ? new x_storage_service_1.BrowserStorageService(localStorage)
            : undefined : _c, _d = _b.cacheService, cacheService = _d === void 0 ? empathy_cache_service_1.EmpathyCacheService : _d;
        if (storageService && cacheService) {
            this.container.bind(container_const_1.DEPENDENCIES.storageService).toConstantValue(storageService);
            this.container.bind(container_const_1.DEPENDENCIES.cacheService).to(cacheService);
        }
        else {
            logger_1.Logger.warn('Tried to enable cache with invalid options');
        }
        return this;
    };
    EmpathyAdapterBuilder.prototype.addClassMapper = function (mapper, entity, feature) {
        var mapperKey = container_const_1.DEPENDENCIES.ResponseMappers[entity];
        if (feature) {
            var requestorKey = container_const_1.DEPENDENCIES.Requestors[feature];
            this.container.bind(mapperKey).to(mapper).whenAnyAncestorIs(requestorKey);
        }
        else {
            this.container.bind(mapperKey).to(mapper);
        }
        return this;
    };
    EmpathyAdapterBuilder.prototype.addMapper = function (mapper, entity, feature) {
        var mapperClass = this.createMapperClass(mapper);
        this.addClassMapper(mapperClass, entity, feature);
        return this;
    };
    EmpathyAdapterBuilder.prototype.replaceMapper = function (mapper, entity) {
        var mapperClass = this.createMapperClass(mapper);
        this.replaceClassMapper(mapperClass, entity);
        return this;
    };
    EmpathyAdapterBuilder.prototype.replaceClassMapper = function (mapper, entity) {
        var mapperKey = container_const_1.DEPENDENCIES.ResponseMappers[entity];
        this.container.rebind(mapperKey).to(mapper);
        return this;
    };
    EmpathyAdapterBuilder.prototype.addClassRequestMapper = function (mapper, feature) {
        if (feature) {
            var requestorKey = container_const_1.DEPENDENCIES.Requestors[feature];
            this.container.bind(container_const_1.DEPENDENCIES.requestMappers).to(mapper).whenAnyAncestorIs(requestorKey);
        }
        else {
            this.container.bind(container_const_1.DEPENDENCIES.requestMappers).to(mapper);
        }
        return this;
    };
    EmpathyAdapterBuilder.prototype.addRequestMapper = function (mapper, feature) {
        var mapperClass = this.createRequestMapperClass(mapper);
        this.addClassRequestMapper(mapperClass, feature);
        return this;
    };
    EmpathyAdapterBuilder.prototype.replaceRequestMapper = function (mapper) {
        var mapperClass = this.createRequestMapperClass(mapper);
        this.replaceClassRequestMapper(mapperClass);
        return this;
    };
    EmpathyAdapterBuilder.prototype.replaceClassRequestMapper = function (mapper) {
        this.container.rebind(container_const_1.DEPENDENCIES.requestMappers).to(mapper);
        return this;
    };
    EmpathyAdapterBuilder.prototype.onResponseTransformed = function (hook, feature) {
        return this.addHook(hook, 'responseTransformed', feature);
    };
    EmpathyAdapterBuilder.prototype.onBeforeRequest = function (hook, feature) {
        return this.addHook(hook, 'beforeRequest', feature);
    };
    EmpathyAdapterBuilder.prototype.onBeforeResponseTransformed = function (hook, feature) {
        return this.addHook(hook, 'beforeResponseTransformed', feature);
    };
    EmpathyAdapterBuilder.prototype.configureContainer = function (configurator) {
        this.configurator = configurator;
        return this;
    };
    EmpathyAdapterBuilder.prototype.setRequestParams = function (requestParams) {
        Object.assign(this.config.requestParams, requestParams);
        return this;
    };
    EmpathyAdapterBuilder.prototype.setLang = function (lang) {
        this.config.requestParams.lang = lang;
        return this;
    };
    EmpathyAdapterBuilder.prototype.setScope = function (scope) {
        this.config.requestParams.scope = scope;
        return this;
    };
    EmpathyAdapterBuilder.prototype.setInstance = function (instance) {
        this.config.instance = instance;
        return this;
    };
    EmpathyAdapterBuilder.prototype.setUserInfo = function (user) {
        Object.assign(this.config.requestParams, user);
        return this;
    };
    EmpathyAdapterBuilder.prototype.setEnvironment = function (environment) {
        this.config.env = environment;
        return this;
    };
    EmpathyAdapterBuilder.prototype.setFeatureConfig = function (featureName, featureConfig) {
        (0, x_deep_merge_1.deepMerge)(this.config.features[featureName], featureConfig);
        return this;
    };
    EmpathyAdapterBuilder.prototype.setResultTrackingConfig = function (resultTrackingConfig) {
        (0, x_deep_merge_1.deepMerge)(this.config.mappings.tracking.result, resultTrackingConfig);
        return this;
    };
    EmpathyAdapterBuilder.prototype.setFacetConfig = function (config, facetId) {
        if (facetId) {
            var namedFacetsConfig = this.config.mappings.facets.named;
            if (!namedFacetsConfig[facetId]) {
                // We will complete the partial facet config in the build method, so we can cast this safely here.
                namedFacetsConfig[facetId] = config;
            }
            else {
                (0, x_deep_merge_1.deepMerge)(namedFacetsConfig[facetId], config);
            }
        }
        else {
            (0, x_deep_merge_1.deepMerge)(this.config.mappings.facets.default, config);
        }
        return this;
    };
    EmpathyAdapterBuilder.prototype.setQueryConfig = function (queryConfig) {
        (0, x_deep_merge_1.deepMerge)(this.config.mappings.query, queryConfig);
        return this;
    };
    EmpathyAdapterBuilder.prototype.withConfiguration = function (config) {
        (0, x_deep_merge_1.deepMerge)(this.config, config);
        return this;
    };
    EmpathyAdapterBuilder.prototype.build = function () {
        if (this.configurator) {
            this.configurator(this.container);
        }
        this.completePartialFacetsConfig();
        return this.container.resolve(this.adapterClass);
    };
    EmpathyAdapterBuilder.prototype.initContainerBindings = function () {
        new container_config_parser_1.ContainerConfigParser(this.defaultBindingsConfig, this.container).parse();
    };
    EmpathyAdapterBuilder.prototype.createMapperClass = function (mapFn) {
        return (0, inversify_1.injectable)()(/** @class */ (function () {
            function MockedMapper() {
                this.map = mapFn;
            }
            return MockedMapper;
        }()));
    };
    EmpathyAdapterBuilder.prototype.createRequestMapperClass = function (mapFn) {
        return (0, inversify_1.injectable)()(/** @class */ (function () {
            function MockedMapper() {
                this.map = mapFn;
            }
            return MockedMapper;
        }()));
    };
    EmpathyAdapterBuilder.prototype.addHook = function (hook, hookName, feature) {
        if (feature) {
            this.container
                .bind(container_const_1.DEPENDENCIES.Hooks[hookName])
                .toFunction(hook)
                .whenInjectedInto(container_const_1.DEPENDENCIES.Requestors[feature]);
        }
        else {
            this.container.bind(container_const_1.DEPENDENCIES.Hooks[hookName]).toFunction(hook);
        }
        return this;
    };
    EmpathyAdapterBuilder.prototype.completePartialFacetsConfig = function () {
        var config = this.container.get(container_const_1.DEPENDENCIES.config);
        var facetsConfig = config.mappings.facets;
        Object.entries(facetsConfig.named).forEach(function (_a) {
            var facetId = _a[0], namedFacetConfig = _a[1];
            facetsConfig.named[facetId] = (0, x_deep_merge_1.deepMerge)({}, facetsConfig.default, namedFacetConfig);
        });
    };
    return EmpathyAdapterBuilder;
}());
exports.EmpathyAdapterBuilder = EmpathyAdapterBuilder;
