// @empathyco/x-components/js/components/suggestions/base-suggestion.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import { forEach } from '@empathyco/x-utils';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { normalizeString } from '../../utils/normalize.js';
import { sanitize } from '../../utils/sanitize.js';

/**
 * Renders a button with a default slot. It receives a query, which should be the query of the
 * module using this component, a suggestion, the {@link XEvent | XEvents} that will be emitted
 * on click with a given feature.
 *
 * The default slot receives the suggestion and the matched query has props.
 *
 * @public
 */
let BaseSuggestion = class BaseSuggestion extends Vue {
    /**
     * The event handler that will be triggered when clicking on a suggestion.
     *
     * @remarks
     * * UserAcceptedAQuery: suggestion.query
     * * UserSelectedASuggestion: suggestion
     * * Merges the events defined in the suggestionSelectedEvents prop and also emits them
     *
     * @returns The {@link XEvent | XEvents} to emit.
     * @public
     */
    get events() {
        return {
            UserAcceptedAQuery: this.suggestion.query,
            UserSelectedASuggestion: this.suggestion,
            ...this.suggestionSelectedEvents
        };
    }
    /**
     * Emits the events when the button is clicked.
     *
     * @public
     */
    emitEvents() {
        forEach(this.events, (event, payload) => {
            this.$x.emit(event, payload, {
                target: this.$el,
                feature: this.feature
            });
        });
    }
    /**
     * Checks if the normalized suggestion query matches with the module's query so it has a
     * matching part.
     *
     * @returns If the query has a matching part or not.
     * @internal
     */
    get hasMatchingQuery() {
        return !!this.query && normalizeString(this.suggestion.query).includes(this.query);
    }
    /**
     * Checks if the suggestion is curated and if it should be highlighted.
     *
     * @returns True if the suggestion is curated and should be highlighted.
     *
     * @internal
     */
    get shouldHighlightCurated() {
        return this.highlightCurated && !!this.suggestion.isCurated;
    }
    /**
     * Generates css classes dynamically.
     *
     * @remarks
     * 'x-suggestion--matching added when the query should be matched.
     *
     * @returns The {@link VueCSSClasses} classes.
     * @public
     */
    get dynamicCSSClasses() {
        return {
            'x-suggestion--matching': this.hasMatchingQuery,
            'x-suggestion--is-curated': this.shouldHighlightCurated
        };
    }
    /**
     * Highlights the matching part of the suggestion query with the query passed as prop of the
     * component putting it inside a `<span>` tag.
     *
     * @remarks
     * The query prop should be normalized.
     *
     * @returns The suggestion's query with the matching part inside a `<span>` tag.
     * @public
     */
    get queryHTML() {
        if (this.hasMatchingQuery) {
            const matcherIndex = normalizeString(this.suggestion.query).indexOf(this.query);
            const [beginning, matching, end] = this.splitAt(this.suggestion.query, matcherIndex, this.query.length);
            const attrsMatching = 'data-test="matching-part" class="x-suggestion__matching-part"';
            return `${beginning}<span ${attrsMatching}>${matching}</span>${end}`;
        }
        return sanitize(this.suggestion.query);
    }
    /**
     * Splits the label in three parts based on two indexes.
     *
     * @param label - The string that will be divided in three parts.
     * @param start - The first index that the label will be divided by.
     * @param skip - The second index that the label will be divided by.
     *
     * @returns The three parts of the divided label.
     * @internal
     */
    splitAt(label, start, skip) {
        const startPart = label.substr(0, start);
        const matchingPart = label.substr(start, skip);
        const endPart = label.substr(start + skip);
        return [sanitize(startPart), sanitize(matchingPart), sanitize(endPart)];
    }
};
__decorate([
    Prop({ default: '' })
], BaseSuggestion.prototype, "query", void 0);
__decorate([
    Prop({ required: true })
], BaseSuggestion.prototype, "suggestion", void 0);
__decorate([
    Prop() //TODO: set to true when the suggestions components pass it.
], BaseSuggestion.prototype, "feature", void 0);
__decorate([
    Prop({ required: true })
], BaseSuggestion.prototype, "suggestionSelectedEvents", void 0);
__decorate([
    Prop({ default: false, type: Boolean })
], BaseSuggestion.prototype, "highlightCurated", void 0);
BaseSuggestion = __decorate([
    Component
], BaseSuggestion);
var script = BaseSuggestion;

export { script as default };
