// @empathyco/x-adapter/dist/esm/empathy/container/container-config-parser.js
import { Container } from 'inversify';
import { Logger } from '../logger';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var ContainerConfigParser = /** @class */ (function () {
    function ContainerConfigParser(config, container) {
        if (container === void 0) { container = new Container(); }
        this.config = config;
        this.container = container;
        this.logger = Logger.child('ContainerConfigParser');
    }
    ContainerConfigParser.prototype.parse = function () {
        var _this = this;
        Object.entries(this.config).forEach(function (_a) {
            var key = _a[0], binding = _a[1];
            if (_this.isClassBinding(binding)) {
                _this.bind(key, binding);
            }
            else if (_this.isSimpleConstantBinding(binding)) {
                _this.bindAsConstant(key, binding.toConstant);
            }
            else if (_this.isParentConstantBinding(binding)) {
                _this.bindAsConstantToParent(key, binding);
            }
            else if (_this.isAncestorConstantBinding(binding)) {
                _this.bindAsConstantToAncestor(key, binding);
            }
            else if (_this.isDynamicBinding(binding)) {
                _this.bindAsDynamic(key, binding);
            }
            else if (_this.isParentBinding(binding)) {
                _this.bindToParent(key, binding);
            }
            else if (_this.isAncestorBinding(binding)) {
                _this.bindToAncestor(key, binding);
            }
            else {
                _this.logger.warn('Unable to parse dependency with key "', key, '" and value: ', binding);
            }
        });
        return this.container;
    };
    ContainerConfigParser.prototype.bind = function (key, binding) {
        var _this = this;
        var bindings = Array.isArray(binding)
            ? binding.map(function (impl) { return _this.container.bind(key).to(impl); })
            : [this.container.bind(key).to(binding)];
        return {
            whenInjectedInto: function (parent) { bindings.forEach(function (partialBinding) { return partialBinding.whenInjectedInto(parent); }); },
            whenAnyAncestorIs: function (ancestor) { bindings.forEach(function (partialBinding) { return partialBinding.whenAnyAncestorIs(ancestor); }); }
        };
    };
    ContainerConfigParser.prototype.bindAsConstant = function (key, constant) {
        return this.container.bind(key).toConstantValue(constant);
    };
    ContainerConfigParser.prototype.bindAsConstantToParent = function (key, binding) {
        var _this = this;
        Object.entries(binding.toConstantWhenInjectedInto).forEach(function (_a) {
            var parent = _a[0], constant = _a[1];
            return _this.bindAsConstant(key, constant).whenInjectedInto(parent);
        });
    };
    ContainerConfigParser.prototype.bindAsConstantToAncestor = function (key, binding) {
        var _this = this;
        Object.entries(binding.toConstantWhenAnyAncestorIs).forEach(function (_a) {
            var parent = _a[0], constant = _a[1];
            return _this.bindAsConstant(key, constant).whenAnyAncestorIs(parent);
        });
    };
    ContainerConfigParser.prototype.bindAsDynamic = function (key, binding) {
        this.container.bind(key).toDynamicValue(binding.toDynamic);
    };
    ContainerConfigParser.prototype.bindToAncestor = function (key, binding) {
        var _this = this;
        if (binding.default) {
            this.bind(key, binding.default);
        }
        Object.entries(binding.whenAnyAncestorIs)
            .forEach(function (_a) {
            var parent = _a[0], contextualBinding = _a[1];
            return _this.bind(key, contextualBinding).whenAnyAncestorIs(parent);
        });
    };
    ContainerConfigParser.prototype.bindToParent = function (key, binding) {
        var _this = this;
        if (binding.default) {
            this.bind(key, binding.default);
        }
        Object.entries(binding.whenInjectedInto)
            .forEach(function (_a) {
            var parent = _a[0], contextualBinding = _a[1];
            return _this.bind(key, contextualBinding).whenInjectedInto(parent);
        });
    };
    ContainerConfigParser.prototype.isAncestorBinding = function (binding) {
        return binding.whenAnyAncestorIs !== undefined;
    };
    ContainerConfigParser.prototype.isClassBinding = function (binding) {
        return Array.isArray(binding) || typeof binding === 'function';
    };
    ContainerConfigParser.prototype.isSimpleConstantBinding = function (binding) {
        return binding.toConstant !== undefined;
    };
    ContainerConfigParser.prototype.isParentConstantBinding = function (binding) {
        return binding.toConstantWhenInjectedInto !== undefined;
    };
    ContainerConfigParser.prototype.isAncestorConstantBinding = function (binding) {
        return binding.toConstantWhenAnyAncestorIs !== undefined;
    };
    ContainerConfigParser.prototype.isDynamicBinding = function (binding) {
        return binding.toDynamic !== undefined;
    };
    ContainerConfigParser.prototype.isParentBinding = function (binding) {
        return binding.whenInjectedInto !== undefined;
    };
    return ContainerConfigParser;
}());
export { ContainerConfigParser };
