// @empathyco/x-components/js/x-modules/facets/store/module.js
import Vue from 'vue';
import { facets } from './getters/facets.getter.js';
import { selectedFiltersByFacet } from './getters/selected-filters-by-facet.getter.js';
import { selectedFilters } from './getters/selected-filters.getter.js';

/**
 * {@link XStoreModule} For the facets module.
 *
 * @internal
 */
const facetsXStoreModule = {
    state: () => ({
        filters: {},
        groups: {},
        facets: {},
        preselectedFilters: []
    }),
    getters: {
        selectedFilters,
        selectedFiltersByFacet,
        facets
    },
    mutations: {
        setFilter(state, filter) {
            Vue.set(state.filters, filter.id, Object.freeze(filter));
        },
        setFilters(state, filters) {
            filters.forEach(filter => Vue.set(state.filters, filter.id, Object.freeze(filter)));
        },
        setPreselectedFilters(state, filters) {
            state.preselectedFilters = filters;
        },
        removeFilter(state, { id }) {
            Vue.delete(state.filters, id);
        },
        removeFilters(state, filters) {
            filters.forEach(({ id }) => Vue.delete(state.filters, id));
        },
        setFacetGroup(state, { facetId, groupId }) {
            Vue.set(state.groups, facetId, groupId);
        },
        removeFacet(state, { id }) {
            Vue.delete(state.facets, id);
        },
        setFacet(state, facet) {
            Vue.set(state.facets, facet.id, facet);
        }
    },
    actions: {}
};

export { facetsXStoreModule };
