// @empathyco/x-components/js/x-modules/search/store/actions/fetch-and-save-search-response.action.js
import { createFetchAndSaveActions } from '../../../../store/utils/fetch-and-save-action.utils.js';

const { fetchAndSave, cancelPrevious } = createFetchAndSaveActions({
    fetch({ dispatch, state }, request) {
        return dispatch('fetchSearchResponse', request ? enrichRequest(request, state) : null);
    },
    onSuccess({ dispatch }, response) {
        dispatch('saveSearchResponse', response);
    }
});
/**
 * Enriches the {@link SearchRequest} object with the origin and page properties taken from the
 * {@link SearchState | search state}.
 *
 * @param request - The {@link InternalSearchRequest}.
 * @param state - {@link SearchState}.
 *
 * @returns The search request.
 * @internal
 */
function enrichRequest(request, state) {
    const { page, ...restRequest } = request;
    const { config: { pageSize }, origin, results } = state;
    const start = page === 1 ? 0 : results.length;
    return {
        ...restRequest,
        // eslint-disable-next-line @typescript-eslint/no-extra-parens
        ...(origin && { origin }),
        start,
        rows: pageSize * page - start
    };
}
/**
 * Default implementation for {@link SearchActions.fetchAndSaveSearchResponse} action.
 *
 * @public
 */
const fetchAndSaveSearchResponse = fetchAndSave;
/**
 * Default implementation for {@link SearchActions.cancelFetchAndSaveSearchResponse} action.
 *
 * @public
 */
const cancelFetchAndSaveSearchResponse = cancelPrevious;

export { cancelFetchAndSaveSearchResponse, fetchAndSaveSearchResponse };
