// @empathyco/x-components/js/x-modules/facets/components/filters/editable-number-range-filter.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import { Prop, Watch, Component, Vue } from 'vue-property-decorator';
import { facetsXModule } from '../../x-module.js';
import { xComponentMixin } from '../../../../components/x-component.mixin.js';

/**
 * Renders an editable number range filter. It has two input fields to handle min and max values,
 * emitting the needed events when clicked.
 *
 * It provides a default slot, with some utils bind, to customize the whole component; and two
 * named slots `apply-content` and `clear-content` to override each button content.
 *
 * If `instant` prop is true, the needed events are emitted immediately; else, apply button is
 * rendered to confirm to do it. False by default.
 *
 * If `clear` prop is true, clear button, which sets to null component min and max values, is
 * rendered. True by default.
 *
 * @public
 */
let EditableNumberRangeFilter = class EditableNumberRangeFilter extends Vue {
    constructor() {
        super(...arguments);
        /**
         * Component min value.
         *
         * @internal
         */
        this.min = null;
        /**
         * Component max value.
         *
         * @internal
         */
        this.max = null;
        this.rangeFilterMin = 'minimum amount';
        this.rangeFilterMax = 'maximum amount';
    }
    /**
     * It watches the filter range values passed as property and updates component range values if
     * they change.
     *
     * @param newRange - New range value.
     *
     * @internal
     */
    onFilterChanged(newRange) {
        this.min = newRange.min;
        this.max = newRange.max;
    }
    /**
     * It watches range values in order to emit the event with the change if `isInstant`
     * property is true.
     *
     * @internal
     */
    instantEmitUserModifiedFilter() {
        if (this.isInstant) {
            this.emitUserModifiedFilter();
        }
    }
    /**
     * Dynamic CSS classes.
     *
     * @returns Object which contains dynamic CSS classes.
     *
     * @internal
     */
    get cssClasses() {
        return { 'x-editable-number-range-filter--error': this.hasError };
    }
    /**
     * Returns {@link @empathyco/x-types#RangeValue | RangeValue} with component min and max
     * values.
     *
     * @returns Range value object with component values.
     *
     * @internal
     */
    get range() {
        return { min: this.min, max: this.max };
    }
    /**
     * It returns true if the property `hasClearButton` is true and there are values to clear.
     *
     * @returns True if the clear button has to be rendered.
     *
     * @internal
     */
    get renderClearButton() {
        return this.hasClearButton && (this.min !== null || this.max !== null);
    }
    /**
     * It checks if component min and max values are valid.
     *
     * @returns True if there is any error in the component min and max values.
     *
     * @internal
     */
    get hasError() {
        return this.min !== null && this.max !== null && this.min > this.max;
    }
    /**
     * It checks if component min and max values are different than the ones within the filter
     * provided as property.
     *
     * @returns True if they are different.
     *
     * @internal
     */
    get areValuesDifferent() {
        return this.min !== this.filter.range.min || this.max !== this.filter.range.max;
    }
    /**
     * It returns the number if possible or null otherwise.
     *
     * @param value - Value.
     * @returns The element value as a number if possible or null.
     *
     * @internal
     */
    parseRangeValue(value) {
        return isNaN(value) ? null : value;
    }
    /**
     * `min` setter.
     *
     * @param value - The component `min` value to be set.
     *
     * @internal
     */
    setMin(value) {
        this.min = this.parseRangeValue(value);
    }
    /**
     * `max` setter.
     *
     * @param value - The component `max` value to be set.
     *
     * @internal
     */
    setMax(value) {
        this.max = this.parseRangeValue(value);
    }
    /**
     * It sets component `min` and `max` values to null and it emits the change if component is
     * working in instant mode.
     *
     * @internal
     */
    clearValues() {
        this.min = null;
        this.max = null;
    }
    /**
     * It emits {@link FacetsXEvents.UserModifiedEditableNumberRangeFilter} event if there are no
     * errors and component `min` and `max` values are different than `filter.range` ones.
     *
     * @internal
     */
    emitUserModifiedFilter() {
        if (!this.hasError && this.areValuesDifferent) {
            this.$x.emit('UserModifiedEditableNumberRangeFilter', {
                ...this.filter,
                range: this.range
            });
        }
    }
};
__decorate([
    Prop({ required: true })
], EditableNumberRangeFilter.prototype, "filter", void 0);
__decorate([
    Prop({ default: false })
], EditableNumberRangeFilter.prototype, "isInstant", void 0);
__decorate([
    Prop({ default: true })
], EditableNumberRangeFilter.prototype, "hasClearButton", void 0);
__decorate([
    Watch('filter.range', { immediate: true, deep: true })
], EditableNumberRangeFilter.prototype, "onFilterChanged", null);
__decorate([
    Watch('range', { deep: true })
], EditableNumberRangeFilter.prototype, "instantEmitUserModifiedFilter", null);
EditableNumberRangeFilter = __decorate([
    Component({
        mixins: [xComponentMixin(facetsXModule)]
    })
], EditableNumberRangeFilter);
var script = EditableNumberRangeFilter;

export { script as default };
