// @empathyco/x-components/js/x-modules/facets/components/filters/editable-number-range-filter.vue.js
import script from './editable-number-range-filter.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(
    "div",
    {
      staticClass: "x-editable-number-range-filter",
      class: _vm.cssClasses,
      attrs: { "data-test": "editable-number-range-filter" },
    },
    [
      _vm._t(
        "default",
        function () {
          return [
            _c("input", {
              staticClass:
                "\n        x-input\n        x-editable-number-range-filter__input x-editable-number-range-filter__input--min\n      ",
              attrs: {
                name: "min",
                type: "number",
                "data-test": "range-min",
                "aria-label": _vm.rangeFilterMin,
              },
              domProps: { value: _vm.min },
              on: {
                change: function ($event) {
                  return _vm.setMin($event.target.valueAsNumber)
                },
              },
            }),
            _vm._v(" "),
            _c("input", {
              staticClass:
                "\n        x-input\n        x-editable-number-range-filter__input x-editable-number-range-filter__input--max\n      ",
              attrs: {
                name: "max",
                type: "number",
                "data-test": "range-max",
                "aria-label": _vm.rangeFilterMax,
              },
              domProps: { value: _vm.max },
              on: {
                change: function ($event) {
                  return _vm.setMax($event.target.valueAsNumber)
                },
              },
            }),
            _vm._v(" "),
            !_vm.isInstant
              ? _c(
                  "button",
                  {
                    staticClass:
                      "x-button x-editable-number-range-filter__apply",
                    attrs: {
                      disabled: _vm.hasError,
                      "data-test": "range-apply",
                    },
                    on: { click: _vm.emitUserModifiedFilter },
                  },
                  [
                    _vm._t("apply-content", function () {
                      return [_vm._v("✓")]
                    }),
                  ],
                  2
                )
              : _vm._e(),
            _vm._v(" "),
            _vm.renderClearButton
              ? _c(
                  "button",
                  {
                    staticClass:
                      "x-button x-editable-number-range-filter__clear",
                    attrs: { "data-test": "range-clear" },
                    on: { click: _vm.clearValues },
                  },
                  [
                    _vm._t("clear-content", function () {
                      return [_vm._v("𐄂")]
                    }),
                  ],
                  2
                )
              : _vm._e(),
          ]
        },
        null,
        {
          min: _vm.min,
          max: _vm.max,
          setMin: _vm.setMin,
          setMax: _vm.setMax,
          emitUserModifiedFilter: _vm.emitUserModifiedFilter,
          clearValues: _vm.clearValues,
          hasError: _vm.hasError,
        }
      ),
    ],
    2
  )
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-9aa789ba_0", { source: ".x-editable-number-range-filter--error .x-editable-number-range-filter__input {\n  border-color: red;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = undefined;
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
