// @empathyco/x-components/js/x-modules/facets/components/filters/simple-filter.vue.js
import script from './simple-filter.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c("RenderlessFilter", {
    staticClass: "x-simple-filter",
    class: _vm.cssClasses,
    attrs: { clickEvents: _vm._clickEvents, filter: _vm.filter },
    scopedSlots: _vm._u(
      [
        {
          key: "default",
          fn: function (ref) {
            var filter = ref.filter;
            var clickFilter = ref.clickFilter;
            var cssClasses = ref.cssClasses;
            var isDisabled = ref.isDisabled;
            return [
              _vm._t(
                "default",
                function () {
                  return [
                    _c(
                      "button",
                      {
                        class: cssClasses,
                        attrs: {
                          "aria-checked": filter.selected.toString(),
                          disabled: isDisabled,
                          "data-test": "filter",
                          role: "checkbox",
                        },
                        on: { click: clickFilter },
                      },
                      [
                        _vm._t(
                          "label",
                          function () {
                            return [_vm._v(_vm._s(filter.label))]
                          },
                          { filter: filter }
                        ),
                      ],
                      2
                    ),
                  ]
                },
                null,
                {
                  filter: filter,
                  clickFilter: clickFilter,
                  cssClasses: cssClasses,
                  isDisabled: isDisabled,
                }
              ),
            ]
          },
        },
      ],
      null,
      true
    ),
  })
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = undefined;
  /* scoped */
  const __vue_scope_id__ = undefined;
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    undefined,
    undefined,
    undefined
  );

export { __vue_component__ as default };
