// @empathyco/x-components/js/components/column-picker/base-column-picker-dropdown.vue.js
import script from './base-column-picker-dropdown.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c("BaseDropdown", {
    attrs: {
      value: _vm.selectedColumns,
      items: _vm.columns,
      animation: _vm.animation,
    },
    on: { change: _vm.emitEvent },
    scopedSlots: _vm._u(
      [
        _vm.$scopedSlots.toggle
          ? {
              key: "toggle",
              fn: function (ref) {
                var item = ref.item;
                var isOpen = ref.isOpen;
                return [
                  _vm._t("toggle", null, null, { item: item, isOpen: isOpen }),
                ]
              },
            }
          : null,
        {
          key: "item",
          fn: function (ref) {
            var item = ref.item;
            var isSelected = ref.isSelected;
            var isHighlighted = ref.isHighlighted;
            return [
              _vm._t("item", null, null, {
                item: item,
                isSelected: isSelected,
                isHighlighted: isHighlighted,
              }),
            ]
          },
        },
      ],
      null,
      true
    ),
  })
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = undefined;
  /* scoped */
  const __vue_scope_id__ = undefined;
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    undefined,
    undefined,
    undefined
  );

export { __vue_component__ as default };
