// @empathyco/x-components/js/components/column-picker/base-column-picker-list.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import { mixins } from 'vue-class-component';
import { Component } from 'vue-property-decorator';
import __vue_component__ from '../base-event-button.vue.js';
import ColumnPickerMixin from './column-picker.mixin.js';

/**
 * Column picker list component renders a list of buttons to choose the columns number.
 *
 * @remarks It extends {@link ColumnPickerMixin}.
 *
 * @public
 */
let BaseColumnPickerList = class BaseColumnPickerList extends mixins(ColumnPickerMixin) {
    /**
     * Maps the column to an object containing: the `column` and `CSS classes`.
     *
     * @returns An array of objects containing the column number and CSS classes.
     *
     * @internal
     */
    get columnsWithCssClasses() {
        return this.columns.map(column => ({
            column,
            cssClasses: [
                `x-column-picker-list__item--${column}-cols`,
                {
                    'x-column-picker-list__item--is-selected': this.selectedColumns === column,
                    'x-option-list__item--is-selected': this.selectedColumns === column
                }
            ],
            isSelected: this.selectedColumns === column,
            events: {
                UserClickedColumnPicker: column,
                ColumnsNumberProvided: column
            }
        }));
    }
};
BaseColumnPickerList = __decorate([
    Component({
        components: { BaseEventButton: __vue_component__ }
    })
], BaseColumnPickerList);
var script = BaseColumnPickerList;

export { script as default };
