// @empathyco/x-components/js/components/column-picker/column-picker.mixin.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Watch } from 'vue-property-decorator';
import Component from 'vue-class-component';
import { XEmit, XOn } from '../decorators/bus.decorators.js';

/**
 * Mixin to share Column Pickers logic.
 *
 * @public
 */
let ColumnPickerMixin = class ColumnPickerMixin extends Vue {
    constructor() {
        super(...arguments);
        /**
         * Selected column, `value` prop by default or the first `columns` item.
         *
         * @internal
         */
        this.selectedColumns = this.providedSelectedColumns;
    }
    /**
     * Retrieves the provided selected column.
     *
     * @returns The provided `value`, or the first value of the list of possible columns.
     * @internal
     */
    get providedSelectedColumns() {
        return this.value ?? this.columns[0];
    }
    /**
     * Synchronizes the selected column with the one provided by argument.
     *
     * @param columns - The column number.
     *
     * @internal
     */
    setSelectedColumns(columns) {
        this.selectedColumns = columns;
    }
    /**
     * Synchronizes the number of selected columns with the provided selected column value.
     *
     * @param column - The new number of columns.
     *
     * @internal
     */
    emitChange(column) {
        if (this.value !== column) {
            this.$emit('change', column);
        }
    }
    /**
     * Synchronizes the columns number before mounting the component. If the real number of selected
     * columns equals the provided columns, it emits the event to sync it with every other component.
     * If it is not equal it means that the user has already selected a number of columns, so we emit
     * a `change` event so developers can sync the provided value.
     *
     * @internal
     */
    beforeMount() {
        if (this.selectedColumns === this.providedSelectedColumns) {
            this.$x.emit('ColumnsNumberProvided', this.selectedColumns);
        }
        else {
            this.emitChange(this.selectedColumns);
        }
    }
};
__decorate([
    Prop({ required: false })
], ColumnPickerMixin.prototype, "value", void 0);
__decorate([
    Prop({ required: true })
], ColumnPickerMixin.prototype, "columns", void 0);
__decorate([
    XEmit('ColumnsNumberProvided', { immediate: false })
], ColumnPickerMixin.prototype, "selectedColumns", void 0);
__decorate([
    Watch('providedSelectedColumns'),
    XOn('ColumnsNumberProvided')
], ColumnPickerMixin.prototype, "setSelectedColumns", null);
__decorate([
    Watch('selectedColumns')
], ColumnPickerMixin.prototype, "emitChange", null);
ColumnPickerMixin = __decorate([
    Component({
        model: {
            event: 'change'
        }
    })
], ColumnPickerMixin);
var ColumnPickerMixin$1 = ColumnPickerMixin;

export { ColumnPickerMixin$1 as default };
