// @empathyco/x-components/js/components/layouts/fixed-header-and-asides-layout.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import { Component } from 'vue-property-decorator';
import { mixins } from 'vue-class-component';
import __vue_component__$1 from '../../x-modules/scroll/components/main-scroll.vue.js';
import { animateTranslate } from '../animations/animate-translate/animate-translate.factory.js';
import __vue_component__ from '../modals/base-id-modal.vue.js';
import __vue_component__$2 from '../../x-modules/scroll/components/scroll.vue.js';
import LayoutsMixin from './layouts.mixin.js';

/**
 * Component for use as Layout to be filled with the rest of the components.
 *
 * @public
 */
let FixedHeaderAndAsidesLayout = class FixedHeaderAndAsidesLayout extends mixins(LayoutsMixin) {
    constructor() {
        super(...arguments);
        this.scrollPosition = 0;
        this.rightAsideAnimation = animateTranslate('right');
        this.leftAsideAnimation = animateTranslate('left');
    }
    setPosition(position) {
        this.scrollPosition = position;
    }
    get isBackdropVisible() {
        return this.scrollPosition > 0;
    }
};
FixedHeaderAndAsidesLayout = __decorate([
    Component({
        components: {
            BaseIdModal: __vue_component__,
            MainScroll: __vue_component__$1,
            Scroll: __vue_component__$2
        }
    })
], FixedHeaderAndAsidesLayout);
var script = FixedHeaderAndAsidesLayout;

export { script as default };
