// @empathyco/x-components/js/components/layouts/fixed-header-and-asides-layout.vue.js
import script from './fixed-header-and-asides-layout.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(
    "MainScroll",
    [
      _c(
        "Scroll",
        {
          staticClass: "x-layout x-layout--fixed-header-and-asides",
          class: { "dev-mode": _vm.devMode },
          attrs: { id: "main-scroll" },
          on: { scroll: _vm.setPosition },
        },
        [
          _c("div", {
            key: "header-backdrop",
            staticClass: "x-layout__header-backdrop",
            class: {
              "x-layout__header-backdrop--is-visible": _vm.isBackdropVisible,
            },
          }),
          _vm._v(" "),
          _vm.hasContent("header")
            ? _c(
                "header",
                {
                  key: "header",
                  staticClass: "x-layout__header x-list x-list--horizontal",
                },
                [
                  _vm._t("header", function () {
                    return [
                      _vm.devMode
                        ? _c("span", { staticClass: "slot-helper" }, [
                            _vm._v("HEADER"),
                          ])
                        : _vm._e(),
                    ]
                  }),
                ],
                2
              )
            : _vm._e(),
          _vm._v(" "),
          _vm.hasContent("sub-header")
            ? _c(
                "div",
                { key: "sub-header", staticClass: "x-layout__sub-header" },
                [
                  _c(
                    "div",
                    { staticClass: "x-layout__sub-header-content" },
                    [
                      _vm._t("sub-header", function () {
                        return [
                          _vm.devMode
                            ? _c("span", { staticClass: "slot-helper" }, [
                                _vm._v("SUB HEADER"),
                              ])
                            : _vm._e(),
                        ]
                      }),
                    ],
                    2
                  ),
                ]
              )
            : _vm._e(),
          _vm._v(" "),
          _vm.hasContent("toolbar")
            ? _c(
                "section",
                { key: "toolbar", staticClass: "x-layout__toolbar" },
                [
                  _vm._t("toolbar", function () {
                    return [
                      _vm.devMode
                        ? _c("span", { staticClass: "slot-helper" }, [
                            _vm._v("TOOLBAR"),
                          ])
                        : _vm._e(),
                    ]
                  }),
                ],
                2
              )
            : _vm._e(),
          _vm._v(" "),
          _vm.hasContent("main")
            ? _c(
                "main",
                {
                  key: "main",
                  staticClass: "x-layout__main x-list x-list--vertical",
                },
                [
                  _vm._t("main", function () {
                    return [
                      _vm.devMode
                        ? _c("span", { staticClass: "slot-helper" }, [
                            _vm._v("MAIN"),
                          ])
                        : _vm._e(),
                    ]
                  }),
                ],
                2
              )
            : _vm._e(),
          _vm._v(" "),
          _vm.hasContent("left-aside")
            ? _c(
                "BaseIdModal",
                {
                  key: "left-aside",
                  staticClass: "x-layout__aside x-layout__aside--left",
                  attrs: {
                    animation: _vm.leftAsideAnimation,
                    modalId: "left-aside",
                  },
                },
                [
                  _vm._t("left-aside", function () {
                    return [
                      _vm.devMode
                        ? _c("span", { staticClass: "slot-helper" }, [
                            _vm._v("LEFT ASIDE"),
                          ])
                        : _vm._e(),
                    ]
                  }),
                ],
                2
              )
            : _vm._e(),
          _vm._v(" "),
          _vm.hasContent("right-aside")
            ? _c(
                "BaseIdModal",
                {
                  key: "right-aside",
                  staticClass: "x-layout__aside x-layout__aside--right",
                  attrs: {
                    animation: _vm.rightAsideAnimation,
                    modalId: "right-aside",
                  },
                },
                [
                  _vm._t("right-aside", function () {
                    return [
                      _vm.devMode
                        ? _c("span", { staticClass: "slot-helper" }, [
                            _vm._v("RIGHT ASIDE"),
                          ])
                        : _vm._e(),
                    ]
                  }),
                ],
                2
              )
            : _vm._e(),
          _vm._v(" "),
          _vm.hasContent("scroll-to-top")
            ? _c(
                "div",
                {
                  key: "scroll-to-top",
                  staticClass: "x-layout__scroll-to-top",
                },
                [
                  _vm._t("scroll-to-top", function () {
                    return [
                      _vm.devMode
                        ? _c("span", { staticClass: "slot-helper" }, [
                            _vm._v("SCROLL TO TOP"),
                          ])
                        : _vm._e(),
                    ]
                  }),
                ],
                2
              )
            : _vm._e(),
        ],
        1
      ),
    ],
    1
  )
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-e252e204_0", { source: ".dev-mode .slot-helper[data-v-e252e204] {\n  font-family: inherit;\n  color: grey;\n  box-sizing: border-box;\n  display: flex;\n  height: 100%;\n  width: 100%;\n  justify-content: center;\n  align-items: center;\n  border: dashed 1px grey;\n  border-radius: 10px;\n}\n.x-layout[data-v-e252e204] {\n  display: grid;\n  align-content: stretch;\n  min-height: 100%;\n  max-height: 100%;\n  --x-size-margin-max-width: calc((100vw - var(--x-size-max-width-layout, 1440px)) / 2);\n  --x-size-margin-layout: max(\n    var(--x-size-min-margin-layout, 20px),\n    var(--x-size-margin-max-width)\n  );\n  grid-template-rows: [page-start header-start] auto [header-end sub-header-start] auto [sub-header-end toolbar-start] auto [toolbar-end main-start] 1fr [main-end page-end];\n  grid-template-columns: [page-start] var(--x-size-margin-layout) [max-width-start] 1fr [max-width-end] var(--x-size-margin-layout) [page-end];\n}\n.x-layout__header[data-v-e252e204] {\n  position: sticky;\n  top: -0.5px;\n  z-index: 2;\n  grid-row: header;\n  grid-column: page;\n  max-height: var(--x-size-max-height-layout-header, auto);\n  padding: 0 var(--x-size-margin-layout);\n  background: var(--x-color-background-layout-header, transparent);\n  border-color: var(--x-size-border-color-layout-header, transparent);\n  border-width: var(--x-size-border-width-layout-header, 0);\n  border-style: solid;\n}\n.x-layout__sub-header[data-v-e252e204] {\n  grid-row: sub-header;\n  grid-column: page;\n  padding: 0 var(--x-size-margin-layout);\n  background: var(--x-color-background-layout-sub-header, transparent);\n  border-color: var(--x-size-border-color-layout-sub-header, transparent);\n  border-width: var(--x-size-border-width-layout-sub-header, 0);\n  border-style: solid;\n}\n.x-layout__toolbar[data-v-e252e204] {\n  grid-row: toolbar;\n  grid-column: max-width;\n}\n.x-layout__main[data-v-e252e204] {\n  grid-row: main;\n  grid-column: max-width;\n}\n.x-layout__aside[data-v-e252e204] {\n  pointer-events: none;\n}\n.x-layout__aside.x-modal[data-v-e252e204] {\n  z-index: 3;\n  flex-flow: row nowrap;\n}\n.x-layout__aside.x-modal[data-v-e252e204]  .x-modal__content {\n  background-color: transparent;\n  height: 100%;\n  width: var(--x-size-width-layout-aside, 300px);\n}\n.x-layout__aside[data-v-e252e204]  > *:not(.slot-helper) {\n  pointer-events: all;\n}\n.x-layout__aside--right.x-modal[data-v-e252e204] {\n  justify-content: flex-end;\n}\n.x-layout__scroll-to-top[data-v-e252e204] {\n  position: fixed;\n  z-index: 1;\n  bottom: var(--x-size-margin-bottom-layout-scroll-to-top, 16px);\n  right: var(--x-size-margin-right-layout-scroll-to-top, 16px);\n}\n.x-layout__header-backdrop[data-v-e252e204] {\n  grid-row: page;\n  grid-column: page;\n  position: sticky;\n  top: -0.5px;\n  z-index: 1;\n  height: var(--x-size-height-layout-backdrop, 40vh);\n  width: 100%;\n  pointer-events: none;\n  background-color: var(--x-color-background-layout-header-backdrop, var(--x-color-base-neutral-100, white));\n  -webkit-mask: linear-gradient(to bottom, rgba(0, 0, 0, 0.8), rgba(0, 0, 0, 0));\n          mask: linear-gradient(to bottom, rgba(0, 0, 0, 0.8), rgba(0, 0, 0, 0));\n  opacity: 0;\n  transition: opacity 0.2s ease-out;\n}\n.x-layout__header-backdrop--is-visible[data-v-e252e204] {\n  opacity: 1;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-e252e204";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
