// @empathyco/x-components/js/components/layouts/multi-column-max-width-layout.vue.js
import script from './multi-column-max-width-layout.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(
    "div",
    {
      staticClass: "x-layout x-layout--columns",
      class: { "dev-mode": _vm.devMode },
      style: { height: _vm.hasContent("main-body") ? "100%" : "auto" },
    },
    [
      _c("header", { staticClass: "x-layout__header" }, [
        _vm.hasContent("header-start")
          ? _c(
              "div",
              { staticClass: "x-list x-layout__header-start" },
              [
                _vm._t("header-start", function () {
                  return [
                    _vm.devMode
                      ? _c("span", { staticClass: "slot-helper" }, [
                          _vm._v("HEADER START"),
                        ])
                      : _vm._e(),
                  ]
                }),
              ],
              2
            )
          : _vm._e(),
        _vm._v(" "),
        _vm.hasContent("header-middle")
          ? _c(
              "div",
              { staticClass: "x-list x-layout__header-middle" },
              [
                _vm._t("header-middle", function () {
                  return [
                    _vm.devMode
                      ? _c("span", { staticClass: "slot-helper" }, [
                          _vm._v("HEADER MIDDLE"),
                        ])
                      : _vm._e(),
                  ]
                }),
              ],
              2
            )
          : _vm._e(),
        _vm._v(" "),
        _vm.hasContent("header-end")
          ? _c(
              "div",
              { staticClass: "x-list x-layout__header-end" },
              [
                _vm._t("header-end", function () {
                  return [
                    _vm.devMode
                      ? _c("span", { staticClass: "slot-helper" }, [
                          _vm._v("HEADER END"),
                        ])
                      : _vm._e(),
                  ]
                }),
              ],
              2
            )
          : _vm._e(),
      ]),
      _vm._v(" "),
      _vm.hasContent("sub-header")
        ? _c("div", { staticClass: "x-layout__sub-header" }, [
            _c(
              "div",
              { staticClass: "x-layout__sub-header-content" },
              [
                _vm._t("sub-header", function () {
                  return [
                    _vm.devMode
                      ? _c("span", { staticClass: "slot-helper" }, [
                          _vm._v("SUB HEADER"),
                        ])
                      : _vm._e(),
                  ]
                }),
              ],
              2
            ),
          ])
        : _vm._e(),
      _vm._v(" "),
      _vm.hasContent("toolbar-aside", "toolbar-body")
        ? _c("section", { staticClass: "x-layout__toolbar" }, [
            _c(
              "aside",
              { staticClass: "x-list x-layout__toolbar-aside" },
              [
                _vm._t("toolbar-aside", function () {
                  return [
                    _vm.devMode
                      ? _c("span", { staticClass: "slot-helper" }, [
                          _vm._v("TOOLBAR ASIDE"),
                        ])
                      : _vm._e(),
                  ]
                }),
              ],
              2
            ),
            _vm._v(" "),
            _c(
              "div",
              { staticClass: "x-list x-layout__toolbar-body" },
              [
                _vm._t("toolbar-body", function () {
                  return [
                    _vm.devMode
                      ? _c("span", { staticClass: "slot-helper" }, [
                          _vm._v("TOOLBAR BODY"),
                        ])
                      : _vm._e(),
                  ]
                }),
              ],
              2
            ),
          ])
        : _vm._e(),
      _vm._v(" "),
      _c(
        "main",
        { staticClass: "x-layout__main" },
        [
          _vm.hasContent("main-aside")
            ? _c(
                "BaseIdTogglePanel",
                {
                  staticClass: "x-layout__collapse-aside",
                  attrs: {
                    panelId: "aside-panel",
                    animation: _vm.asideAnimation,
                  },
                },
                [
                  _c(
                    "Scroll",
                    {
                      staticClass: "x-layout__aside-scroll",
                      attrs: { id: "aside-scroll" },
                    },
                    [
                      _c(
                        "div",
                        {
                          staticClass:
                            "x-layout__main-aside x-list x-list--vertical",
                        },
                        [
                          _vm._t("main-aside", function () {
                            return [
                              _vm.devMode
                                ? _c(
                                    "span",
                                    {
                                      staticClass: "slot-helper",
                                      staticStyle: { height: "110vh" },
                                    },
                                    [_vm._v("MAIN ASIDE")]
                                  )
                                : _vm._e(),
                            ]
                          }),
                        ],
                        2
                      ),
                    ]
                  ),
                ],
                1
              )
            : _vm._e(),
          _vm._v(" "),
          _c(
            "MainScroll",
            [
              _vm.hasContent("main-body")
                ? _c(
                    "Scroll",
                    {
                      staticClass: "x-layout__body-scroll",
                      attrs: { id: "main-scroll" },
                    },
                    [
                      _c(
                        "section",
                        {
                          staticClass:
                            "x-layout__main-body x-list x-list--vertical",
                        },
                        [
                          _vm._t("main-body", function () {
                            return [
                              _vm.devMode
                                ? _c(
                                    "span",
                                    {
                                      staticClass: "slot-helper",
                                      staticStyle: { height: "110vh" },
                                    },
                                    [_vm._v("MAIN BODY")]
                                  )
                                : _vm._e(),
                            ]
                          }),
                        ],
                        2
                      ),
                    ]
                  )
                : _vm._e(),
            ],
            1
          ),
        ],
        1
      ),
      _vm._v(" "),
      _vm.hasContent("scroll-to-top")
        ? _c("div", { staticClass: "x-layout__scroll-to-top" }, [
            _c(
              "div",
              { staticClass: "x-layout__scroll-to-top-content" },
              [
                _vm._t("scroll-to-top", function () {
                  return [
                    _vm.devMode
                      ? _c(
                          "span",
                          {
                            staticClass: "slot-helper",
                            staticStyle: { height: "50px" },
                          },
                          [_vm._v("SCROLL TO TOP")]
                        )
                      : _vm._e(),
                  ]
                }),
              ],
              2
            ),
          ])
        : _vm._e(),
    ]
  )
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-24f2b448_0", { source: ":root {\n  --x-size-min-margin-layout-columns: var(--x-size-base-06);\n  --x-size-column-gap-layout-columns: var(--x-size-base-06);\n  --x-size-padding-top-layout-columns-header: var(--x-size-base-07);\n  --x-size-padding-bottom-layout-columns-header: var(--x-size-base-05);\n  --x-size-padding-top-layout-columns-toolbar: var(--x-size-base-05);\n  --x-size-padding-bottom-layout-columns-toolbar: var(--x-size-base-03);\n  --x-size-margin-bottom-layout-columns-scroll-to-top: var(--x-size-base-03);\n  --x-size-margin-left-layout-columns-scroll-to-top: var(--x-size-base-03);\n  --x-color-background-layout-columns: var(--x-color-base-neutral-100);\n  --x-color-border-layout-columns: var(--x-color-base-neutral-70);\n  --x-color-border-layout-columns-header: var(--x-color-border-layout-columns);\n  --x-color-border-layout-columns-sub-header: var(--x-color-border-layout-columns);\n  --x-color-border-layout-columns-toolbar: var(--x-color-border-layout-columns);\n  --x-color-border-layout-columns-main: var(--x-color-border-layout-columns);\n  --x-color-border-layout-columns-main-aside: var(--x-color-border-layout-columns);\n  --x-color-border-layout-columns-main-body: var(--x-color-border-layout-columns);\n  --x-size-border-width-layout-columns-header: 0;\n  --x-size-border-width-layout-columns-toolbar: 1px 0 0;\n}", map: undefined, media: undefined })
,inject("data-v-24f2b448_1", { source: ".dev-mode .slot-helper[data-v-24f2b448] {\n  font-family: inherit;\n  color: grey;\n  box-sizing: border-box;\n  display: flex;\n  height: 100%;\n  width: 100%;\n  justify-content: center;\n  align-items: center;\n  border: dashed 1px grey;\n  border-radius: 10px;\n}\n.x-layout[data-v-24f2b448] {\n  --x-size-gap-default: var(--x-size-column-gap-layout-columns, 20px);\n  --x-size-margin-max-width: calc((100vw - var(--x-size-max-width-layout-columns, 1440px)) / 2);\n  --x-size-margin-layout-columns: max(\n    var(--x-size-min-margin-layout-columns, 20px),\n    var(--x-size-margin-max-width)\n  );\n  --x-number-columns-header-middle: calc(var(--x-number-columns-layout-columns, 5) - 2);\n  --x-number-columns-sub-header-content: var(--x-number-columns-header-middle);\n  --x-number-columns-toolbar-body: calc(var(--x-number-columns-layout-columns, 5) - 1);\n  display: grid;\n  align-content: stretch;\n  max-height: 100%;\n  grid-template-columns: [page-start] 1fr [scroll-to-top-start] calc(var(--x-size-margin-layout-columns) + var(--x-size-gap-default)) [scroll-to-top-end page-end];\n  grid-template-rows: [page-start header-start] auto [header-end sub-header-start] auto [sub-header-end toolbar-start] auto [toolbar-end main-start] 1fr [scroll-to-top-start] auto [main-end scroll-to-top-end page-end];\n}\n.x-layout__header[data-v-24f2b448] {\n  grid-column: page;\n  grid-row: header;\n  display: grid;\n  column-gap: var(--x-size-gap-default);\n  grid-template-columns: [header-start] var(--x-size-margin-layout-columns) [start-content-start] 1fr [start-content-end middle-content-start] repeat(var(--x-number-columns-header-middle), 1fr) [middle-content-end end-content-start] 1fr [end-content-end] var(--x-size-margin-layout-columns) [header-end];\n  padding-top: var(--x-size-padding-top-layout-columns-header, 0);\n  padding-bottom: var(--x-size-padding-bottom-layout-columns-header, 0);\n  margin-top: var(--x-size-margin-top-layout-columns-header, 0);\n  margin-bottom: var(--x-size-margin-bottom-layout-columns-header, 0);\n  background-color: var(--x-color-background-layout-columns-header, transparent);\n  border-color: var(--x-color-border-layout-columns-header, transparent);\n  border-style: solid;\n  border-width: var(--x-size-border-width-layout-columns-header, 0);\n}\n.x-layout__header-start[data-v-24f2b448] {\n  grid-column: start-content;\n}\n.x-layout__header-start.x-list[data-v-24f2b448] {\n  flex-flow: var(--x-flow-layout-columns-header-start, row nowrap);\n  justify-content: var(--x-size-justify-layout-columns-header-start, flex start);\n  align-items: var(--x-size-align-layout-columns-header-start, flex start);\n}\n.x-layout__header-middle[data-v-24f2b448] {\n  grid-column: middle-content;\n}\n.x-layout__header-middle.x-list[data-v-24f2b448] {\n  flex-flow: var(--x-flow-layout-columns-header-middle, row nowrap);\n  justify-content: var(--x-size-justify-layout-columns-header-middle, center);\n  align-items: var(--x-size-align-layout-columns-header-middle, flex start);\n}\n.x-layout__header-end[data-v-24f2b448] {\n  grid-column: end-content;\n}\n.x-layout__header-end.x-list[data-v-24f2b448] {\n  flex-flow: var(--x-flow-layout-columns-header-end, row nowrap);\n  justify-content: var(--x-size-justify-layout-columns-header-end, flex-end);\n  align-items: var(--x-size-align-layout-columns-header-end, flex-start);\n}\n.x-layout__sub-header[data-v-24f2b448] {\n  grid-column: page;\n  grid-row: sub-header;\n  display: grid;\n  column-gap: var(--x-size-gap-default);\n  grid-template-columns: [sub-header-start] var(--x-size-margin-layout-columns) 1fr [sub-header-content-start] repeat(var(--x-number-columns-sub-header-content), 1fr) [sub-header-content-end] 1fr var(--x-size-margin-layout-columns) [sub-header-end];\n  padding-top: var(--x-size-padding-top-layout-columns-sub-header, 0);\n  padding-bottom: var(--x-size-padding-bottom-layout-columns-sub-header), 0;\n  margin-top: var(--x-size-margin-top-layout-columns-sub-header, 0);\n  margin-bottom: var(--x-size-margin-bottom-layout-columns-sub-header, 0);\n  background-color: var(--x-color-background-layout-columns-sub-header, transparent);\n  border-color: var(--x-color-border-layout-columns-sub-header, transparent);\n  border-style: solid;\n  border-width: var(--x-size-border-width-layout-columns-sub-header, 0);\n}\n.x-layout__sub-header-content[data-v-24f2b448] {\n  grid-column: sub-header-content;\n}\n.x-layout__sub-header-content.x-list[data-v-24f2b448] {\n  flex-flow: var(--x-flow-layout-columns-sub-header, row nowrap);\n  justify-content: var(--x-size-justify-layout-columns-sub-header, flex-start);\n  align-items: var(--x-size-align-layout-columns-sub-header, flex-start);\n}\n.x-layout__toolbar[data-v-24f2b448] {\n  grid-column: page;\n  grid-row: toolbar;\n  display: grid;\n  column-gap: var(--x-size-gap-default);\n  grid-template-columns: [toolbar-start] var(--x-size-margin-layout-columns) [toolbar-aside-start] 1fr [toolbar-aside-end toolbar-body-start] repeat(var(--x-number-columns-toolbar-body), 1fr) [toolbar-body-end] var(--x-size-margin-layout-columns) [toolbar-end];\n  padding-top: var(--x-size-padding-top-layout-columns-toolbar, 0);\n  padding-bottom: var(--x-size-padding-bottom-layout-columns-toolbar, 0);\n  margin-top: var(--x-size-margin-top-layout-columns-toolbar, 0);\n  margin-bottom: var(--x-size-margin-bottom-layout-columns-toolbar, 0);\n  background-color: var(--x-color-background-layout-columns-toolbar, transparent);\n  border-color: var(--x-color-border-layout-columns-toolbar, transparent);\n  border-style: solid;\n  border-width: var(--x-size-border-width-layout-columns-toolbar, 0);\n}\n.x-layout__toolbar-aside[data-v-24f2b448] {\n  grid-column: toolbar-aside;\n}\n.x-layout__toolbar-aside.x-list[data-v-24f2b448] {\n  flex-flow: var(--x-flow-layout-columns-toolbar-aside, row nowrap);\n  justify-content: var(--x-size-justify-layout-columns-toolbar-aside, flex-start);\n  align-items: var(--x-size-align-layout-columns-toolbar-aside, center);\n}\n.x-layout__toolbar-body[data-v-24f2b448] {\n  grid-column: toolbar-body;\n}\n.x-layout__toolbar-body.x-list[data-v-24f2b448] {\n  flex-flow: var(--x-flow-layout-columns-toolbar-body, row nowrap);\n  justify-content: var(--x-size-justify-layout-columns-toolbar-body, flex-end);\n  align-items: var(--x-size-align-layout-columns-toolbar-body, center);\n}\n.x-layout__main[data-v-24f2b448] {\n  grid-column: page;\n  grid-row: main;\n  min-height: 0;\n  display: flex;\n  flex-flow: row nowrap;\n  justify-content: stretch;\n  padding-left: calc(var(--x-size-margin-layout-columns) + var(--x-size-gap-default));\n  padding-top: var(--x-size-padding-top-layout-columns-main, 0);\n  padding-bottom: var(--x-size-padding-bottom-layout-columns-main, 0);\n  margin-top: var(--x-size-margin-top-layout-columns-main, 0);\n  margin-bottom: var(--x-size-margin-bottom-layout-columns-main, 0);\n  background-color: var(--x-color-background-layout-columns-main, transparent);\n  border-color: var(--x-color-border-layout-columns-main, transparent);\n  border-style: solid;\n  border-width: var(--x-size-border-width-layout-columns-main, 0);\n}\n.x-layout__collapse-aside[data-v-24f2b448] {\n  width: calc((100% - var(--x-size-gap-default) * (var(--x-number-columns-layout-columns, 5)) - var(--x-size-margin-layout-columns)) / var(--x-number-columns-layout-columns, 5) + var(--x-size-gap-default));\n}\n.x-layout__aside-scroll[data-v-24f2b448] {\n  height: 100%;\n  margin-right: var(--x-size-gap-default);\n}\n.x-layout__body-scroll[data-v-24f2b448] {\n  flex: 1 1 auto;\n  width: 0;\n  padding-right: calc(var(--x-size-margin-layout-columns) + var(--x-size-gap-default) - 16px);\n}\n.x-layout__body-scroll.x-scroll[data-v-24f2b448] {\n  --x-string-overflow-scroll: scroll;\n}\n.x-layout__scroll-to-top[data-v-24f2b448] {\n  grid-area: scroll-to-top;\n  justify-content: flex-end;\n  width: 100%;\n  display: flex;\n  z-index: 1;\n  pointer-events: none;\n}\n[dir=\"ltr\"] .x-layout__scroll-to-top-content[data-v-24f2b448] {\n  padding-left: var(--x-size-margin-left-layout-columns-scroll-to-top, 10px);\n}\n[dir=\"rtl\"] .x-layout__scroll-to-top-content[data-v-24f2b448] {\n  padding-right: var(--x-size-margin-left-layout-columns-scroll-to-top, 10px);\n}\n.x-layout__scroll-to-top-content[data-v-24f2b448] {\n  flex: 1 0 auto;\n  padding-bottom: var(--x-size-margin-bottom-layout-columns-scroll-to-top, 10px);\n}\n.x-layout__scroll-to-top-content[data-v-24f2b448]  > * {\n  pointer-events: all;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-24f2b448";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
