// @empathyco/x-components/js/components/result/base-result-image.vue.js
import script from './base-result-image.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(
    "picture",
    {
      ref: "image",
      staticClass: "x-picture x-result-picture",
      attrs: { "data-test": "result-picture" },
    },
    [
      !_vm.hasImageLoaded && !_vm.hasAllImagesFailed
        ? _c(
            _vm.animation,
            {
              tag: "component",
              staticClass: "x-picture__image x-picture__image--placeholder",
              attrs: { "data-test": "result-picture-placeholder" },
            },
            [_vm._t("placeholder")],
            2
          )
        : _vm._e(),
      _vm._v(" "),
      _c(
        _vm.animation,
        { tag: "component" },
        [
          _vm.imageSrc
            ? _c("img", {
                directives: [
                  {
                    name: "show",
                    rawName: "v-show",
                    value: _vm.hasImageLoaded,
                    expression: "hasImageLoaded",
                  },
                ],
                staticClass: "x-picture__image x-result-picture-image",
                attrs: {
                  alt: _vm.result.name,
                  src: _vm.imageSrc,
                  "data-test": "result-picture-image",
                },
                on: { error: _vm.flagImageAsFailed, load: _vm.flagImageLoaded },
              })
            : _vm.hasAllImagesFailed
            ? _c(
                "NoElement",
                { staticClass: "x-picture__image x-picture__image--fallback" },
                [_vm._t("fallback")],
                2
              )
            : _vm._e(),
        ],
        1
      ),
    ],
    1
  )
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-0b620073_0", { source: ".x-result-picture[data-v-0b620073] {\n  min-width: 1px;\n  min-height: 1px;\n  position: relative;\n}\n.x-result-picture__image[data-v-0b620073] {\n  max-width: 100%;\n  max-height: 100%;\n  object-fit: contain;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-0b620073";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
