// @empathyco/x-components/js/components/result/base-result-link.vue.js
import script from './base-result-link.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(
    "a",
    {
      staticClass: "x-result-link",
      attrs: { href: _vm.result.url, "data-test": "result-link" },
      on: {
        click: _vm.emitUserClickedAResult,
        contextmenu: function ($event) {
          return _vm.emitUserClickedAResult.apply(null, arguments)
        },
        mouseup: function ($event) {
          if ("button" in $event && $event.button !== 1) {
            return null
          }
          return _vm.emitUserClickedAResult.apply(null, arguments)
        },
      },
    },
    [_vm._t("default", null, { result: _vm.result })],
    2
  )
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-2fa908e2_0", { source: ".x-result-link[data-v-2fa908e2] {\n  text-decoration: none;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-2fa908e2";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
