// @empathyco/x-components/js/components/result/base-result-image.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { NoElement } from '../no-element.js';

/**
 * Component to be reused that renders an `<img>`.
 *
 * @public
 */
let BaseResultImage = class BaseResultImage extends Vue {
    constructor() {
        super(...arguments);
        /**
         * The image has entered in the port view.
         *
         * @public
         */
        this.hasEnteredView = false;
        /**
         * An array of images that failed to load.
         *
         * @public
         */
        this.failedImages = [];
        /**
         * Indicates if the result image is loaded.
         *
         * @public
         */
        this.hasImageLoaded = false;
    }
    /**
     * Checks if intersection observer is available in window object.
     *
     * @returns Boolean.
     *
     * @internal
     */
    get isIntersectionObserverAvailable() {
        return 'IntersectionObserver' in window;
    }
    mounted() {
        this.hasEnteredView = !this.isIntersectionObserverAvailable;
        if (this.isIntersectionObserverAvailable) {
            this.createObserver();
        }
    }
    /**
     * Gets the src from the result image.
     *
     * @returns The result image src.
     *
     * @internal
     */
    get imageSrc() {
        if (this.hasEnteredView && this.result.images.length > 0) {
            const image = this.result.images.find(image => !this.failedImages.includes(image));
            return image ?? '';
        }
        return '';
    }
    /**
     * Creates an intersection observer in the image element.
     *
     * @internal
     */
    createObserver() {
        const image = this.$refs.image;
        const observer = new IntersectionObserver(this.observerHandler.bind(this));
        observer.observe(image);
    }
    /**
     * Observe all the observables items and detects when a element is intersected.
     *
     * @param entries - The observed items.
     * @param observer - The intersection observer object.
     *
     * @internal
     */
    observerHandler(entries, observer) {
        entries.forEach(entry => {
            if (entry.isIntersecting) {
                this.hasEnteredView = true;
                observer.disconnect();
            }
        });
    }
    /**
     * Sets an image as failed.
     *
     * @internal
     */
    flagImageAsFailed() {
        this.failedImages.push(this.imageSrc);
    }
    /**
     * Checks if all the images failed.
     *
     * @returns Boolean.
     *
     * @internal
     */
    get hasAllImagesFailed() {
        return this.failedImages.length === this.result.images.length;
    }
    /**
     * Marks an image as loaded.
     *
     * @internal
     */
    flagImageLoaded() {
        this.hasImageLoaded = true;
    }
};
__decorate([
    Prop({ default: () => NoElement })
], BaseResultImage.prototype, "animation", void 0);
__decorate([
    Prop({ required: true })
], BaseResultImage.prototype, "result", void 0);
BaseResultImage = __decorate([
    Component({
        components: {
            NoElement
        }
    })
], BaseResultImage);
var script = BaseResultImage;

export { script as default };
