// @empathyco/x-components/js/components/result/base-result-rating.vue.js
import script from './base-result-rating.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _vm.result.rating && _vm.result.rating.value
    ? _c(
        _vm.link ? "a" : "div",
        {
          tag: "component",
          staticClass: "x-result-rating",
          attrs: { href: _vm.link, "data-test": "result-rating" },
          on: { click: _vm.emitClickedEvent },
        },
        [
          _vm._t(
            "default",
            function () {
              return [
                _c(
                  "BaseRating",
                  _vm._b(
                    {
                      attrs: { value: _vm.result.rating.value },
                      scopedSlots: _vm._u(
                        [
                          {
                            key: "empty-icon",
                            fn: function () {
                              return [_vm._t("empty-icon")]
                            },
                            proxy: true,
                          },
                          {
                            key: "filled-icon",
                            fn: function () {
                              return [_vm._t("filled-icon")]
                            },
                            proxy: true,
                          },
                        ],
                        null,
                        true
                      ),
                    },
                    "BaseRating",
                    _vm.$attrs,
                    false
                  )
                ),
              ]
            },
            { rating: _vm.result.rating.value, result: _vm.result }
          ),
        ],
        2
      )
    : _vm._e()
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-ae12fb70_0", { source: ".x-result-rating[data-v-ae12fb70] {\n  color: inherit;\n  text-decoration: none;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-ae12fb70";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
