// @empathyco/x-adapter/dist/esm/empathy/mappers/response/empathy-tagging.mapper.js
import { __decorate } from "tslib";
import { injectable } from 'inversify';
import { Logger } from '../../logger';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyTaggingMapper = /** @class */ (function () {
    function EmpathyTaggingMapper() {
        this.logger = Logger.child('EmpathyTaggingMapper');
    }
    EmpathyTaggingMapper.prototype.map = function (taggingUrl, tagging) {
        try {
            var _a = this.extractUrlParameters(taggingUrl), url = _a.url, params = _a.params;
            Object.assign(tagging, { url: url, params: params });
        }
        catch (e) {
            this.logger.warn('[EmpathyBrokerTaggingMapper]', 'Invalid tagging url: ', taggingUrl);
        }
        return tagging;
    };
    EmpathyTaggingMapper.prototype.extractUrlParameters = function (taggingUrl) {
        var params = {};
        var url = new URL(taggingUrl);
        url.searchParams.forEach(function (value, key) {
            if (Array.isArray(params[key])) {
                params[key].push(value);
            }
            else if (params[key]) {
                params[key] = [params[key], value];
            }
            else {
                params[key] = value;
            }
        });
        params.follow = false;
        return {
            url: "".concat(url.origin).concat(url.pathname),
            params: params
        };
    };
    EmpathyTaggingMapper = __decorate([
        injectable()
    ], EmpathyTaggingMapper);
    return EmpathyTaggingMapper;
}());
export { EmpathyTaggingMapper };
