// @empathyco/x-adapter/dist/esm/empathy/mappers/response/empathy-banner.mapper.js
import { __decorate, __metadata, __param } from "tslib";
import { injectable, multiInject } from 'inversify';
import { DEPENDENCIES } from '../../container/container.const';
import { pipeMappers } from '../pipe-mappers';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyBannerMapper = /** @class */ (function () {
    function EmpathyBannerMapper(taggingMappers) {
        this.mapTagging = pipeMappers.apply(void 0, taggingMappers);
    }
    EmpathyBannerMapper.prototype.map = function (rawBanner, banner, context) {
        return Object.assign(banner, {
            modelName: 'Banner',
            id: rawBanner.id,
            title: rawBanner.title,
            url: rawBanner.url,
            image: rawBanner.imagename,
            tagging: {
                click: this.mapTagging(rawBanner.trackable_url, {}, context)
            }
        });
    };
    EmpathyBannerMapper = __decorate([
        injectable(),
        __param(0, multiInject(DEPENDENCIES.ResponseMappers.queryTagging)),
        __metadata("design:paramtypes", [Array])
    ], EmpathyBannerMapper);
    return EmpathyBannerMapper;
}());
export { EmpathyBannerMapper };
