// @empathyco/x-adapter/dist/esm/empathy/mappers/response/empathy-promoted.mapper.js
import { __decorate, __metadata, __param } from "tslib";
import { injectable, multiInject } from 'inversify';
import { DEPENDENCIES } from '../../container/container.const';
import { pipeMappers } from '../pipe-mappers';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyPromotedMapper = /** @class */ (function () {
    function EmpathyPromotedMapper(taggingMappers) {
        this.mapTagging = pipeMappers.apply(void 0, taggingMappers);
    }
    EmpathyPromotedMapper.prototype.map = function (rawPromoted, promoted, context) {
        return Object.assign(promoted, {
            modelName: 'Promoted',
            id: rawPromoted.id,
            title: rawPromoted.title,
            url: rawPromoted.url,
            image: rawPromoted.imagename,
            tagging: {
                click: this.mapTagging(rawPromoted.trackable_url, {}, context)
            }
        });
    };
    EmpathyPromotedMapper = __decorate([
        injectable(),
        __param(0, multiInject(DEPENDENCIES.ResponseMappers.queryTagging)),
        __metadata("design:paramtypes", [Array])
    ], EmpathyPromotedMapper);
    return EmpathyPromotedMapper;
}());
export { EmpathyPromotedMapper };
