// @empathyco/x-adapter/dist/esm/empathy/mappers/response/empathy-redirection.mapper.js
import { __decorate, __metadata, __param } from "tslib";
import { injectable, multiInject } from 'inversify';
import { DEPENDENCIES } from '../../container/container.const';
import { pipeMappers } from '../pipe-mappers';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyRedirectionMapper = /** @class */ (function () {
    function EmpathyRedirectionMapper(taggingMappers) {
        this.mapTagging = pipeMappers.apply(void 0, taggingMappers);
    }
    EmpathyRedirectionMapper.prototype.map = function (rawDirect, redirection, context) {
        return Object.assign(redirection, {
            modelName: 'Redirection',
            id: rawDirect.id,
            url: rawDirect.url,
            tagging: {
                click: this.mapTagging(rawDirect.trackable_url, {}, context)
            }
        });
    };
    EmpathyRedirectionMapper = __decorate([
        injectable(),
        __param(0, multiInject(DEPENDENCIES.ResponseMappers.queryTagging)),
        __metadata("design:paramtypes", [Array])
    ], EmpathyRedirectionMapper);
    return EmpathyRedirectionMapper;
}());
export { EmpathyRedirectionMapper };
