// @empathyco/x-adapter/dist/esm/empathy/mappers/response/empathy-partial-result.mapper.js
import { __decorate, __metadata, __param } from "tslib";
import { injectable, multiInject } from 'inversify';
import { DEPENDENCIES } from '../../container/container.const';
import { pipeMappers } from '../pipe-mappers';
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var EmpathyPartialResultMapper = /** @class */ (function () {
    function EmpathyPartialResultMapper(resultMappers) {
        this.mapResult = pipeMappers.apply(void 0, resultMappers);
    }
    EmpathyPartialResultMapper.prototype.map = function (rawPartialResult, partialResult, context) {
        var _this = this;
        return Object.assign(partialResult, {
            query: rawPartialResult.suggestion,
            totalResults: rawPartialResult.numFound,
            results: rawPartialResult.docs.map(function (rawResult) { return _this.mapResult(rawResult, {}, context); })
        });
    };
    EmpathyPartialResultMapper = __decorate([
        injectable(),
        __param(0, multiInject(DEPENDENCIES.ResponseMappers.results)),
        __metadata("design:paramtypes", [Array])
    ], EmpathyPartialResultMapper);
    return EmpathyPartialResultMapper;
}());
export { EmpathyPartialResultMapper };
