// @empathyco/x-components/js/x-modules/facets/entities/filter-entity.factory.js
import { isFacetFilter } from '@empathyco/x-types';
import { EditableNumberRangeFilterEntity } from './editable-number-range-filter.entity.js';
import { HierarchicalFilterEntity } from './hierarchical-filter.entity.js';
import { NumberRangeFilterEntity } from './number-range-filter.entity.js';
import { RawFilterEntity } from './raw-filter.entity.js';
import { SimpleFilterEntity } from './simple-filter.entity.js';

/**
 * Creates {@link FilterEntity | FilterEntities} based on the provided configs.
 *
 * @internal
 */
class FilterEntityFactory {
    constructor() {
        /**
         * The registered entities by default to be used by the Factory.
         */
        this.entities = [
            SimpleFilterEntity,
            HierarchicalFilterEntity,
            NumberRangeFilterEntity,
            EditableNumberRangeFilterEntity,
            RawFilterEntity
        ];
        /**
         * The registered modifiers grouped by the facetId to be applied.
         */
        this.modifiersByFacetId = {};
        /**
         * The registered modifiers grouped by the filter `ModelName` to be applied.
         */
        this.modifiersByFilterModelName = {};
        /**
         * Contains the instantiated entities for each facet.
         *
         * @internal
         */
        this.cache = {};
    }
    /**
     * Creates a new FilterEntity from a filter.
     *
     * @param store - The {@link https://vuex.vuejs.org/guide/ | Vuex Store} that the entity will
     * use.
     * @param filter - The {@link @empathyco/x-types#Filter | Filter} necessary to know what
     * {@link FilterEntity} to create.
     * @returns The {@link FilterEntity} created by the factory.
     */
    getFilterEntity(store, filter) {
        const cacheKey = isFacetFilter(filter) ? filter.facetId : '__unknown-facet__';
        return this.cache[cacheKey] ?? (this.cache[cacheKey] = this.createFilterEntity(store, filter));
    }
    /**
     * Creates a brand new {@link FilterEntity} for the given
     * {@link @empathyco/x-types#Filter | Filter}.
     *
     * @param store - The store which should be mutated through the entity.
     * @param filter - The filter to create the entity for.
     * @returns A new {@link FilterEntity} for the given {@link @empathyco/x-types#Filter | Filter}.
     * @internal
     */
    createFilterEntity(store, filter) {
        const filterEntityConstructor = this.entities.find(entity => entity.accepts(filter));
        if (!filterEntityConstructor) {
            throw new Error(`Entity configuration for ${filter.modelName} not registered.`);
        }
        const entity = new filterEntityConstructor(store);
        const modifiers = isFacetFilter(filter)
            ? this.modifiersByFacetId[filter.facetId] ??
                this.modifiersByFilterModelName[filter.modelName] ??
                []
            : [];
        return modifiers.reduce((modifiedEntity, modifier) => new modifier(store, modifiedEntity), entity);
    }
    /**
     * Registers a new {@link FilterEntity} to be used by the factory.
     *
     * @param entity - The new {@link FilterEntity} to be registered in the factory.
     */
    registerFilterEntity(entity) {
        if (!this.entities.includes(entity)) {
            this.entities.push(entity);
        }
    }
    /**
     * Registers a list of modifiers to be used with the {@link FilterEntity | FilterEntities} of a
     * particular facet.
     *
     * @param facetId - The facet id whose Entities will be modified.
     * @param modifiers - The list of modifiers to be registered.
     */
    registerModifierByFacetId(facetId, ...modifiers) {
        this.updateModifiers(this.modifiersByFacetId, facetId, modifiers);
    }
    /**
     * Registers a list of modifiers to be used with the {@link FilterEntity | FilterEntities} of a
     * particular facet.
     *
     * @param filterModelName - The facet ModelName whose Entities will be modified.
     * @param modifiers - The list of modifiers to be registered.
     */
    registerModifierByFilterModelName(filterModelName, ...modifiers) {
        this.updateModifiers(this.modifiersByFilterModelName, filterModelName, modifiers);
    }
    /**
     * Updates the modifiers for a particular facet.
     *
     * @param modifiersRecord - The map of modifiers to be updated.
     * @param modifierKey - The id whose modifiers will be updated.
     * @param modifiers - The list of new modifiers to be registered.
     */
    updateModifiers(modifiersRecord, modifierKey, modifiers) {
        if (!modifiersRecord[modifierKey]) {
            modifiersRecord[modifierKey] = [];
        }
        const facetModifiers = modifiersRecord[modifierKey];
        const newModifiers = modifiers.filter(modifier => !facetModifiers.includes(modifier));
        facetModifiers.push(...newModifiers);
    }
}
/**
 * The public global instance of the Singleton.
 *
 * @remarks The constructor is not private, so it's possible to create more instances,
 * to simplify the testing.
 */
FilterEntityFactory.instance = new FilterEntityFactory();

export { FilterEntityFactory };
