// @empathyco/x-components/js/x-modules/facets/entities/hierarchical-filter.entity.js
import { isHierarchicalFilter } from '@empathyco/x-types';
import { addFacetIfNotPresent } from './add-facet-if-not-present.js';

/**
 * Allows selecting and deselecting a filter of {@link @empathyco/x-types#HierarchicalFilter
 * | HierarchicalFilter}.
 *
 * @internal
 */
class HierarchicalFilterEntity {
    constructor(store) {
        this.store = store;
    }
    /**
     * Deselects the hierarchical filter and all of its descendants.
     *
     * @param filter - The filter to deselect.
     */
    deselect(filter) {
        this.saveFilter({ ...filter, selected: false });
        this.deselectDescendants(filter);
        addFacetIfNotPresent(this.store, filter.facetId, 'HierarchicalFacet');
    }
    /**
     * Selects the hierarchical filter and its ancestors.
     *
     * @param filter - The filter to select.
     */
    select(filter) {
        this.saveFilter({ ...filter, selected: true });
        this.selectAncestors(filter);
        addFacetIfNotPresent(this.store, filter.facetId, 'HierarchicalFacet');
    }
    /**
     * Deselects all the descendants of the given filter, saving them to the store.
     *
     * @remarks The descendants hierarchy is retrieved from the store.
     * @param filter - The filter to deselect its descendants.
     * @internal
     */
    deselectDescendants(filter) {
        if (filter.children) {
            filter.children.forEach(childId => {
                const child = this.getFilterById(childId);
                if (child) {
                    this.saveFilter({ ...child, selected: false });
                    this.deselectDescendants(child);
                }
            });
        }
    }
    /**
     * Selects all the ancestors of the given filter, saving them to the store.
     *
     * @remarks The ancestors hierarchy is retrieved from the store.
     * @param filter - The filter to select its ancestors.
     * @internal
     */
    selectAncestors(filter) {
        if (filter.parentId) {
            const parent = this.getFilterById(filter.parentId);
            if (parent) {
                this.saveFilter({ ...parent, selected: true });
                this.selectAncestors(parent);
            }
        }
    }
    /**
     * Retrieves a hierarchical filter from the store by its id.
     *
     * @param id - The id of the filter to retrieve.
     * @returns The hierarchical filter retrieved from the store.
     * @internal
     */
    getFilterById(id) {
        return this.store.state.x.facets.filters[id];
    }
    /**
     * Saves the given filter to the store.
     *
     * @param filter - The filter to save to the store.
     * @internal
     */
    saveFilter(filter) {
        this.store.commit('x/facets/setFilter', filter);
    }
}
HierarchicalFilterEntity.accepts = isHierarchicalFilter;

export { HierarchicalFilterEntity };
