// @empathyco/x-components/js/x-modules/next-queries/store/module.js
import { setQuery } from '../../../store/utils/query.utils.js';
import { setStatus } from '../../../store/utils/status-store.utils.js';
import { cancelFetchAndSaveNextQueries, fetchAndSaveNextQueries } from './actions/fetch-and-save-next-queries.action.js';
import { fetchNextQueries } from './actions/fetch-next-queries.action.js';
import { setQueryFromLastHistoryQuery } from './actions/set-query-from-last-history-query.action.js';
import { setUrlParams } from './actions/set-url-params.action.js';
import { nextQueries } from './getters/next-queries.getter.js';
import { request } from './getters/request.getter.js';

/**
 * {@link XStoreModule} For the next-queries module.
 *
 * @internal
 */
const nextQueriesXStoreModule = {
    state: () => ({
        query: '',
        nextQueries: [],
        searchedQueries: [],
        status: 'initial',
        config: {
            maxItemsToRequest: 20,
            hideSessionQueries: true,
            loadOnInit: true
        },
        params: {}
    }),
    getters: {
        request,
        nextQueries
    },
    mutations: {
        setQuery,
        setNextQueries(state, nextQueries) {
            state.nextQueries = nextQueries;
        },
        setSearchedQueries(state, searchedQueries) {
            state.searchedQueries = searchedQueries;
        },
        setStatus,
        setParams(state, params) {
            state.params = params;
        }
    },
    actions: {
        cancelFetchAndSaveNextQueries,
        fetchAndSaveNextQueries,
        fetchNextQueries,
        setQueryFromLastHistoryQuery,
        setUrlParams
    }
};

export { nextQueriesXStoreModule };
