// @empathyco/x-components/js/x-modules/search/wiring.js
import { namespacedWireCommit, namespacedWireDispatch, namespacedWireDispatchWithoutPayload } from '../../wiring/namespaced-wires.factory.js';
import { createWiring } from '../../wiring/wiring.utils.js';

/**
 * `search` {@link XModuleName | XModule name}.
 *
 * @internal
 */
const moduleName = 'search';
/**
 * WireCommit for {@link SearchXModule}.
 *
 * @internal
 */
const wireCommit = namespacedWireCommit(moduleName);
/**
 * WireDispatch for {@link SearchXModule}.
 *
 * @internal
 */
const wireDispatch = namespacedWireDispatch(moduleName);
/**
 * WireDispatchWithoutPayload for {@link SearchXModule}.
 *
 * @internal
 */
const wireDispatchWithoutPayload = namespacedWireDispatchWithoutPayload(moduleName);
/**
 * Cancels the {@link SearchActions.fetchAndSaveSearchResponse} request promise.
 *
 * @public
 */
const cancelFetchAndSaveSearchResponseWire = wireDispatchWithoutPayload('cancelFetchAndSaveSearchResponse');
/**
 * Sets the search state `origin`.
 *
 * @public
 */
const saveOriginWire = wireDispatch('saveOrigin', ({ metadata }) => metadata);
/**
 * Requests and stores the search response.
 *
 * @public
 */
const fetchAndSaveSearchResponseWire = wireDispatch('fetchAndSaveSearchResponse');
/**
 * Resets the search state `spellcheckedQuery` to its initial value, an empty string.
 *
 * @public
 */
const resetSpellcheckQuery = wireCommit('setSpellcheck', '');
/**
 * Sets the search state `relatedTags`.
 *
 * @public
 */
const setRelatedTags = wireCommit('setRelatedTags');
/**
 * Sets the search state `query`.
 *
 * @public
 */
const setSearchQuery = wireCommit('setQuery');
/**
 * Sets the search state `selectedFilters`.
 *
 * @public
 */
const setSelectedFilters = wireCommit('setSelectedFilters');
/**
 * Sets the search state `sort`.
 *
 * @public
 */
const setSort = wireCommit('setSort');
/**
 * Sets the search state `query`.
 *
 * @public
 */
const setUrlParams = wireDispatch('setUrlParams');
/**
 * Sets the search state `page`.
 *
 * @public
 */
const setSearchPage = wireCommit('setPage');
/**
 * Sets the search state `params`.
 *
 * @public
 */
const setSearchExtraParams = wireCommit('setParams');
/**
 * Increases the current search state `page` by one.
 *
 * @public
 */
const increasePageAppendingResultsWire = wireDispatchWithoutPayload('increasePageAppendingResults');
/**
 * Resets the search state `isAppendingResults`.
 *
 * @public
 */
const resetAppending = wireCommit('setIsAppendResults', false);
/**
 * Batches state resets after {@link SearchGetters.request} parameters update.
 *
 * @public
 */
const resetStateWire = wireDispatch('resetState', ({ eventPayload: newRequest, metadata: { oldValue } }) => ({
    newRequest,
    oldRequest: oldValue
}));
/**
 * Search wiring.
 *
 * @internal
 */
const searchWiring = createWiring({
    ParamsLoadedFromUrl: {
        setUrlParams,
        saveOriginWire
    },
    UserAcceptedAQuery: {
        setSearchQuery,
        saveOriginWire
    },
    UserAcceptedSpellcheckQuery: {
        resetSpellcheckQuery
    },
    UserClearedQuery: {
        setSearchQuery,
        cancelFetchAndSaveSearchResponseWire
    },
    UserClickedASort: {
        setSort
    },
    UserPickedARelatedTag: {
        saveOriginWire
    },
    UserReachedResultsListEnd: {
        increasePageAppendingResultsWire
    },
    SearchRequestChanged: {
        fetchAndSaveSearchResponseWire
    },
    SearchRequestUpdated: {
        resetStateWire
    },
    SelectedRelatedTagsChanged: {
        setRelatedTags
    },
    SelectedFiltersChanged: {
        setSelectedFilters
    },
    ResultsChanged: {
        resetAppending
    },
    SelectedSortProvided: {
        setSort
    },
    ExtraParamsChanged: {
        setSearchExtraParams
    }
});

export { cancelFetchAndSaveSearchResponseWire, fetchAndSaveSearchResponseWire, increasePageAppendingResultsWire, resetAppending, resetSpellcheckQuery, resetStateWire, saveOriginWire, searchWiring, setRelatedTags, setSearchExtraParams, setSearchPage, setSearchQuery, setSelectedFilters, setSort, setUrlParams };
