// @empathyco/x-adapter/dist/cjs/empathy/requestors/beacon-tracking.requestor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BeaconTrackingRequestor = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var BeaconTrackingRequestor = /** @class */ (function () {
    function BeaconTrackingRequestor() {
        var _this = this;
        this.isAdblockPresent = false;
        this.detectAdblock()
            .then(function () {
            _this.isAdblockPresent = false;
        })
            .catch(function () {
            _this.isAdblockPresent = true;
        });
    }
    BeaconTrackingRequestor.prototype.request = function (trackingRequest) {
        var url = this.buildUrl(trackingRequest);
        if (!this.isAdblockPresent && navigator.sendBeacon(url)) {
            return Promise.resolve();
        }
        else {
            if (this.isAdblockPresent) {
                return fetch(url);
            }
            return Promise.reject('Beacon not queued');
        }
    };
    BeaconTrackingRequestor.prototype.buildUrl = function (trackingRequest) {
        var url = new URL(trackingRequest.url);
        Object.entries(trackingRequest.params).forEach(function (_a) {
            var key = _a[0], value = _a[1];
            return url.searchParams.append(key, value);
        });
        return url.href;
    };
    BeaconTrackingRequestor.prototype.detectAdblock = function () {
        var flaggedURL = 'https://google.com/pagead/js/adsbygoogle.js';
        if (!('fetch' in window)) {
            return Promise.resolve();
        }
        return fetch(new Request(flaggedURL, {
            method: 'HEAD',
            mode: 'no-cors'
        }));
    };
    BeaconTrackingRequestor = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__metadata("design:paramtypes", [])
    ], BeaconTrackingRequestor);
    return BeaconTrackingRequestor;
}());
exports.BeaconTrackingRequestor = BeaconTrackingRequestor;
