// @empathyco/x-adapter/dist/cjs/empathy/requestors/feature.requestor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FeatureRequestor = void 0;
var tslib_1 = require("tslib");
var x_get_safe_property_chain_1 = require("@empathyco/x-get-safe-property-chain");
var inversify_1 = require("inversify");
var container_const_1 = require("../container/container.const");
var pipe_mappers_1 = require("../mappers/pipe-mappers");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var FeatureRequestor = /** @class */ (function () {
    function FeatureRequestor(config, feature, httpClient, endpointsService, requestMappers, mappers, beforeRequest, beforeResponseTransformed, responseTransformed) {
        if (beforeRequest === void 0) { beforeRequest = []; }
        if (beforeResponseTransformed === void 0) { beforeResponseTransformed = []; }
        if (responseTransformed === void 0) { responseTransformed = []; }
        var _this = this;
        this.config = config;
        this.feature = feature;
        this.httpClient = httpClient;
        this.endpointsService = endpointsService;
        this.beforeRequest = beforeRequest;
        this.beforeResponseTransformed = beforeResponseTransformed;
        this.responseTransformed = responseTransformed;
        this.mapRequest = pipe_mappers_1.pipeMappers.apply(void 0, requestMappers);
        this.featureConfig = config.features[feature];
        this.responseMappers = {};
        Object.keys(this.featureConfig.responsePaths).forEach(function (entity) {
            _this.responseMappers[entity] = pipe_mappers_1.pipeMappers.apply(void 0, mappers[entity]);
        }, {});
    }
    FeatureRequestor.prototype.request = function (rawRequest, requestOptions) {
        var _this = this;
        if (requestOptions === void 0) { requestOptions = {}; }
        var feature = this.feature;
        var url = this.endpointsService.buildUrl(this.featureConfig.endpoint);
        if (requestOptions.ttlInMinutes === undefined) {
            requestOptions.ttlInMinutes = this.featureConfig.cacheTTLInMinutes;
        }
        var requestContext = { requestOptions: requestOptions, feature: feature, url: url };
        var request = this.mapRequest(rawRequest, {}, requestContext);
        var beforeRequestContext = tslib_1.__assign(tslib_1.__assign({}, requestContext), { rawRequest: rawRequest, request: request });
        this.runHooks(this.beforeRequest, beforeRequestContext);
        return this.httpClient.get(url, request, requestOptions)
            .then(function (rawResponse) {
            var responseContext = tslib_1.__assign(tslib_1.__assign({}, beforeRequestContext), { rawResponse: rawResponse });
            _this.runHooks(_this.beforeResponseTransformed, responseContext);
            var response = _this.transformResponse(rawResponse, responseContext);
            var responseTransformedContext = tslib_1.__assign(tslib_1.__assign({}, responseContext), { response: response });
            _this.runHooks(_this.responseTransformed, responseTransformedContext);
            return response;
        });
    };
    FeatureRequestor.prototype.runHooks = function (hooks, context) {
        hooks.forEach(function (hook) { return hook(context); });
    };
    FeatureRequestor.prototype.transformResponse = function (rawResponse, context) {
        var _this = this;
        return Object.entries(this.featureConfig.responsePaths)
            .reduce(function (transformedResponse, _a) {
            var entityName = _a[0], rawResponsePath = _a[1];
            var rawResponseSelection = (0, x_get_safe_property_chain_1.getSafePropertyChain)(rawResponse, rawResponsePath);
            if (rawResponseSelection !== undefined) {
                // @ts-ignore The error does not make sense: Record<string, any> cannot be indexed with type string???
                transformedResponse[entityName] = _this.transformResponseSelection(rawResponseSelection, entityName, context);
            }
            return transformedResponse;
        }, {});
    };
    FeatureRequestor.prototype.transformResponseSelection = function (rawResponseSelection, entityName, context) {
        var mapFn = this.responseMappers[entityName];
        return Array.isArray(rawResponseSelection)
            ? rawResponseSelection.map(function (rawEntity) { return mapFn(rawEntity, {}, context); })
            : mapFn(rawResponseSelection, {}, context);
    };
    FeatureRequestor = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.config)),
        tslib_1.__param(1, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.featureName)),
        tslib_1.__param(2, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.httpClient)),
        tslib_1.__param(3, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.endpointsService)),
        tslib_1.__param(4, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.requestMappers)),
        tslib_1.__param(5, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.entityMappers)),
        tslib_1.__param(6, (0, inversify_1.optional)()),
        tslib_1.__param(6, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.Hooks.beforeRequest)),
        tslib_1.__param(7, (0, inversify_1.optional)()),
        tslib_1.__param(7, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.Hooks.beforeResponseTransformed)),
        tslib_1.__param(8, (0, inversify_1.optional)()),
        tslib_1.__param(8, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.Hooks.responseTransformed)),
        tslib_1.__metadata("design:paramtypes", [Object, String, Object, Object, Array, Object, Array, Array, Array])
    ], FeatureRequestor);
    return FeatureRequestor;
}());
exports.FeatureRequestor = FeatureRequestor;
