// @empathyco/x-adapter/dist/cjs/empathy/requestors/requestors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Requestors = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
var container_const_1 = require("../container/container.const");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var Requestors = /** @class */ (function () {
    function Requestors(nextQueries, topRecommendations, sectionRecommendations, clicksRecommendations, queriesRecommendations, userRecommendations, relatedTags, searchById, search, suggestions, track) {
        this.nextQueries = nextQueries;
        this.topRecommendations = topRecommendations;
        this.sectionRecommendations = sectionRecommendations;
        this.clicksRecommendations = clicksRecommendations;
        this.queriesRecommendations = queriesRecommendations;
        this.userRecommendations = userRecommendations;
        this.relatedTags = relatedTags;
        this.searchById = searchById;
        this.search = search;
        this.suggestions = suggestions;
        this.track = track;
    }
    Requestors = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.Requestors.nextQueries)),
        tslib_1.__param(1, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.Requestors.topRecommendations)),
        tslib_1.__param(2, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.Requestors.sectionRecommendations)),
        tslib_1.__param(3, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.Requestors.clicksRecommendations)),
        tslib_1.__param(4, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.Requestors.queriesRecommendations)),
        tslib_1.__param(5, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.Requestors.userRecommendations)),
        tslib_1.__param(6, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.Requestors.relatedTags)),
        tslib_1.__param(7, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.Requestors.searchById)),
        tslib_1.__param(8, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.Requestors.search)),
        tslib_1.__param(9, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.Requestors.suggestions)),
        tslib_1.__param(10, (0, inversify_1.inject)(container_const_1.DEPENDENCIES.Requestors.track)),
        tslib_1.__metadata("design:paramtypes", [Object, Object, Object, Object, Object, Object, Object, Object, Object, Object, Object])
    ], Requestors);
    return Requestors;
}());
exports.Requestors = Requestors;
