// @empathyco/x-components/js/components/modals/base-events-modal.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { XOn } from '../decorators/bus.decorators.js';
import { isElementEqualOrContained } from '../../utils/html.js';
import __vue_component__ from './base-modal.vue.js';

/**
 * Component containing a modal that emits a {@link XEventsTypes.UserClickedCloseEventsModal} when
 * clicking outside the content rendered in the default slot and can receive, through the
 * eventsToCloseModal prop, a list of {@link XEvent | xEvents} to listen to in order to close
 * also the modal, eventsToOpenModal prop,  another list of {@link XEvent | xEvents} to customize
 * the events to listen to open the modal and a prop, displayOverlay, to display an
 * overlay over the rest of the html. The default slot offers the possibility to customize the
 * modal content.
 *
 * @public
 */
let BaseEventsModal = class BaseEventsModal extends Vue {
    constructor() {
        super(...arguments);
        /**
         * Whether the modal is open or not.
         */
        this.isOpen = false;
    }
    /**
     * Opens the modal.
     *
     * @param _payload - The payload of the event that opened the modal.
     * @param metadata - The metadata of the event that opened the modal.
     *
     * @internal
     */
    openModal(_payload, metadata) {
        if (!this.isOpen) {
            this.openerElement = metadata.target;
            this.isOpen = true;
        }
    }
    /**
     * Closes the modal.
     *
     * @internal
     */
    closeModal() {
        if (this.isOpen) {
            this.isOpen = false;
        }
    }
    /**
     * Emits the event defined in the {@link BaseEventsModal.bodyClickEvent} event unless the passed
     * event target is the button that opened the modal.
     *
     * @param event - The event that triggered the close attempt.
     * @public
     */
    emitBodyClickEvent(event) {
        // Prevents clicking the open button when the panel is already open to close the panel.
        if (!this.openerElement ||
            !isElementEqualOrContained(this.openerElement, event.target)) {
            this.$x.emit(this.bodyClickEvent, undefined, { target: this.$el });
        }
    }
};
__decorate([
    Prop()
], BaseEventsModal.prototype, "animation", void 0);
__decorate([
    Prop({ default: () => ['UserClickedOpenEventsModal'] })
], BaseEventsModal.prototype, "eventsToOpenModal", void 0);
__decorate([
    Prop({
        default: () => ['UserClickedCloseEventsModal', 'UserClickedOutOfEventsModal']
    })
], BaseEventsModal.prototype, "eventsToCloseModal", void 0);
__decorate([
    Prop({ default: 'UserClickedOutOfEventsModal' })
], BaseEventsModal.prototype, "bodyClickEvent", void 0);
__decorate([
    XOn(component => component.eventsToOpenModal)
], BaseEventsModal.prototype, "openModal", null);
__decorate([
    XOn(component => component.eventsToCloseModal)
], BaseEventsModal.prototype, "closeModal", null);
BaseEventsModal = __decorate([
    Component({
        components: { BaseModal: __vue_component__ }
    })
], BaseEventsModal);
var script = BaseEventsModal;

export { script as default };
