// @empathyco/x-components/js/components/modals/base-modal.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { NoElement } from '../no-element.js';

/**
 * Base component with no XPlugin dependencies that serves as a utility for constructing more
 * complex modals.
 *
 * @public
 */
let BaseModal = class BaseModal extends Vue {
    constructor() {
        super(...arguments);
        /** The previous value of the body overflow style. */
        this.previousBodyOverflow = '';
        /** The previous value of the HTML element overflow style. */
        this.previousHTMLOverflow = '';
        /** To animate the overlay opacity after and before the animation. */
        this.showOverlay = true;
    }
    mounted() {
        /* Watcher added after mount to prevent SSR from breaking */
        // eslint-disable-next-line @typescript-eslint/unbound-method
        this.$watch('open', this.syncBody, { immediate: true });
    }
    /**
     * Syncs the body to the open state of the modal, adding or removing styles and listeners.
     *
     * @param isOpen - True when the modal is opened.
     * @internal
     */
    syncBody(isOpen) {
        if (isOpen) {
            this.disableScroll();
            this.addBodyListeners();
            /* eslint-disable @typescript-eslint/unbound-method */
            this.$on('hook:beforeDestroy', this.removeBodyListeners);
            this.$on('hook:beforeDestroy', this.enableScroll);
            /* eslint-enable @typescript-eslint/unbound-method */
        }
        else {
            this.enableScroll();
            this.removeBodyListeners();
            /* eslint-disable @typescript-eslint/unbound-method */
            this.$off('hook:beforeDestroy', this.removeBodyListeners);
            this.$off('hook:beforeDestroy', this.enableScroll);
            /* eslint-enable @typescript-eslint/unbound-method */
        }
    }
    /**
     * Disables the scroll of both the body and the window.
     *
     * @internal
     */
    disableScroll() {
        this.previousBodyOverflow = document.body.style.overflow;
        this.previousHTMLOverflow = document.documentElement.style.overflow;
        document.body.style.overflow = document.documentElement.style.overflow = 'hidden';
    }
    /**
     * Restores the scroll of both the body and the window.
     *
     * @internal
     */
    enableScroll() {
        document.body.style.overflow = this.previousBodyOverflow;
        document.documentElement.style.overflow = this.previousHTMLOverflow;
        document.body.style.overflow = document.documentElement.style.overflow = '';
    }
    /**
     * Adds listeners to the body element ot detect if the modal should be closed.
     *
     * @internal
     */
    addBodyListeners() {
        /* eslint-disable @typescript-eslint/unbound-method */
        document.body.addEventListener('focusin', this.emitFocusInBody);
        /* eslint-enable @typescript-eslint/unbound-method */
    }
    /**
     * Removes the body listeners.
     *
     * @internal
     */
    removeBodyListeners() {
        /* eslint-disable @typescript-eslint/unbound-method */
        document.body.removeEventListener('focusin', this.emitFocusInBody);
        /* eslint-enable @typescript-eslint/unbound-method */
    }
    /**
     * Emits the `click:overlay` event if the click has been triggered in the overlay layer.
     *
     * @param event - The click event.
     * @internal
     */
    emitOverlayClicked(event) {
        this.$emit('click:overlay', event);
    }
    /**
     * Emits the `focusin:body` event if a focus event has been triggered outside the modal.
     *
     * @param event - The focusin event.
     * @internal
     */
    emitFocusInBody(event) {
        if (!this.$refs.modal.contains(event.target)) {
            this.$emit('focusin:body', event);
        }
    }
};
__decorate([
    Prop({ default: () => NoElement })
], BaseModal.prototype, "animation", void 0);
__decorate([
    Prop({ required: true })
], BaseModal.prototype, "open", void 0);
BaseModal = __decorate([
    Component
], BaseModal);
var script = BaseModal;

export { script as default };
