// @empathyco/x-components/js/components/modals/base-modal.vue.js
import script from './base-modal.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(
    _vm.animation,
    {
      tag: "component",
      on: {
        beforeEnter: function ($event) {
          _vm.showOverlay = false;
        },
        afterEnter: function ($event) {
          _vm.showOverlay = true;
        },
      },
    },
    [
      _vm.open
        ? _c(
            "div",
            { staticClass: "x-modal", attrs: { "data-test": "modal" } },
            [
              _c(
                "div",
                {
                  ref: "modal",
                  staticClass: "x-modal__content x-list",
                  attrs: { "data-test": "modal-content", role: "dialog" },
                },
                [_vm._t("default")],
                2
              ),
              _vm._v(" "),
              _c("div", {
                staticClass: "x-modal__overlay",
                class: { "x-modal__overlay--is-visible": _vm.showOverlay },
                attrs: { "data-test": "modal-overlay" },
                on: {
                  click: _vm.emitOverlayClicked,
                  keydown: _vm.emitOverlayClicked,
                },
              }),
            ]
          )
        : _vm._e(),
    ]
  )
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-389f85a8_0", { source: ".x-modal[data-v-389f85a8] {\n  position: fixed;\n  top: 0;\n  left: 0;\n  display: flex;\n  align-items: flex-start;\n  justify-content: flex-start;\n  width: 100%;\n  height: 100%;\n  z-index: 1;\n}\n.x-modal__content[data-v-389f85a8] {\n  z-index: 1;\n}\n.x-modal__overlay[data-v-389f85a8] {\n  width: 100%;\n  height: 100%;\n  position: fixed;\n  background-color: rgba(0, 0, 0, 0.7);\n  opacity: 0;\n}\n.x-modal__overlay--is-visible[data-v-389f85a8] {\n  transition: opacity 0.3s ease-out;\n  opacity: 1;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-389f85a8";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
