// @empathyco/x-components/js/components/modals/base-id-modal.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import { XOn } from '../decorators/bus.decorators.js';
import { isElementEqualOrContained } from '../../utils/html.js';
import __vue_component__ from './base-modal.vue.js';

/**
 * Component containing a modal expecting a required prop, named `modalId`. It reacts to
 * `UserClickedOpenModal`, `UserClickedCloseModal` and `UserClickedOutOfModal` events, when their
 * payload matches the component's 'modalId' prop, to handle its open/close state. The default
 * slot offers the possibility to customise the modal content.
 *
 * @public
 */
let BaseIdModal = class BaseIdModal extends Vue {
    constructor() {
        super(...arguments);
        /** Whether the modal is open or not. */
        this.isOpen = false;
    }
    /**
     * Opens the modal.
     *
     * @param modalId - The payload of the {@link XEventsTypes.UserClickedOpenModal} event.
     * @param metadata - The metadata of the emitted event.
     * @public
     */
    openModal(modalId, metadata) {
        if (!this.isOpen && this.modalId === modalId) {
            this.openerElement = metadata.target;
            this.isOpen = true;
        }
    }
    /**
     * Closes the modal.
     *
     * @param payload - The payload of the closing events:
     * {@link XEventsTypes.UserClickedCloseModal} or {@link XEventsTypes.UserClickedOutOfModal}.
     *
     * @public
     */
    closeModal(payload) {
        if (this.isOpen && this.modalId === payload) {
            this.isOpen = false;
        }
    }
    /**
     * Emits a {@link XEventsTypes.UserClickedOutOfModal} event unless the passed event target
     * is the button that opened the modal.
     *
     * @param event - The event that triggered the close attempt.
     * @public
     */
    emitClickOutOfModal(event) {
        // Prevents clicking the open button when the panel is already open to close the panel.
        if (!this.openerElement ||
            !isElementEqualOrContained(this.openerElement, event.target)) {
            this.$x.emit('UserClickedOutOfModal', this.modalId, { target: this.$el });
        }
    }
};
__decorate([
    Prop()
], BaseIdModal.prototype, "animation", void 0);
__decorate([
    Prop({ required: true })
], BaseIdModal.prototype, "modalId", void 0);
__decorate([
    XOn('UserClickedOpenModal')
], BaseIdModal.prototype, "openModal", null);
__decorate([
    XOn(['UserClickedCloseModal', 'UserClickedOutOfModal'])
], BaseIdModal.prototype, "closeModal", null);
BaseIdModal = __decorate([
    Component({
        components: { BaseModal: __vue_component__ }
    })
], BaseIdModal);
var script = BaseIdModal;

export { script as default };
