// @empathyco/x-components/js/x-modules/next-queries/components/next-queries-list.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import { mixins } from 'vue-class-component';
import { Prop, Component } from 'vue-property-decorator';
import { Getter } from '../../../components/decorators/store.decorators.js';
import { NoElement } from '../../../components/no-element.js';
import { ItemsListInjectionMixin } from '../../../components/items-list-injection.mixin.js';
import __vue_component__$1 from '../../../components/items-list.vue.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { groupItemsBy } from '../../../utils/array.js';
import __vue_component__ from '../../search/components/results-list.vue.js';
import { nextQueriesXModule } from '../x-module.js';

/**
 * Component that inserts groups of next queries in different positions of the injected search
 * items list, based on the provided configuration.
 *
 * @public
 */
let NextQueriesList = class NextQueriesList extends mixins(ItemsListInjectionMixin) {
    /**
     * The grouped next queries based on the given config.
     *
     * @returns A list of next queries groups.
     * @internal
     */
    get nextQueriesGroups() {
        return Object.values(groupItemsBy(this.nextQueries, (_, index) => Math.floor(index / this.maxNextQueriesPerGroup)))
            .slice(0, this.maxGroups)
            .map(nextQueries => ({
            modelName: 'NextQueriesGroup',
            id: nextQueries.map(nextQuery => nextQuery.query).join(','),
            nextQueries
        }));
    }
    /**
     * New list of {@link ListItem}s to render.
     *
     * @returns The new list of {@link ListItem}s with the next queries groups inserted.
     * @internal
     */
    get items() {
        if (!this.injectedListItems) {
            return this.nextQueriesGroups;
        }
        return this.nextQueriesGroups.reduce((items, nextQueriesGroup, index) => {
            const targetIndex = this.offset + this.frequency * index;
            if (targetIndex <= items.length) {
                items.splice(targetIndex, 0, nextQueriesGroup);
            }
            return items;
        }, [...this.injectedListItems]);
    }
};
__decorate([
    Prop()
], NextQueriesList.prototype, "animation", void 0);
__decorate([
    Prop({ default: 24 })
], NextQueriesList.prototype, "offset", void 0);
__decorate([
    Prop({ default: 24 })
], NextQueriesList.prototype, "frequency", void 0);
__decorate([
    Prop({ default: 4 })
], NextQueriesList.prototype, "maxNextQueriesPerGroup", void 0);
__decorate([
    Prop()
], NextQueriesList.prototype, "maxGroups", void 0);
__decorate([
    Getter('nextQueries', 'nextQueries')
], NextQueriesList.prototype, "nextQueries", void 0);
NextQueriesList = __decorate([
    Component({
        components: {
            ResultsList: __vue_component__,
            NoElement,
            ItemsList: __vue_component__$1
        },
        mixins: [xComponentMixin(nextQueriesXModule)]
    })
], NextQueriesList);
var script = NextQueriesList;

export { script as default };
