// @empathyco/x-components/js/x-modules/next-queries/components/next-query.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Component } from 'vue-property-decorator';
import __vue_component__ from '../../../components/suggestions/base-suggestion.vue.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { nextQueriesXModule } from '../x-module.js';

/**
 * Renders a next query item which receives the suggestion that will be rendered as a prop. It
 * exposes a default slot to change the next query content. If the slot is not overridden,
 * it will render the suggestion query by default.
 *
 * @public
 */
let NextQuery = class NextQuery extends Vue {
    /**
     * Events list which are going to be emitted when a next query is selected.
     *
     * @returns The {@link XEvent | XEvents} to emit.
     * @public
     */
    get events() {
        return {
            UserSelectedANextQuery: this.suggestion
        };
    }
    /**
     * Checks if the next query is curated and if it should be highlighted.
     *
     * @returns True if the next query is curated and should be highlighted.
     *
     * @internal
     */
    get shouldHighlightCurated() {
        return this.highlightCurated && (this.suggestion.isCurated ?? false);
    }
};
__decorate([
    Prop({ required: true })
], NextQuery.prototype, "suggestion", void 0);
__decorate([
    Prop({ default: false, type: Boolean })
], NextQuery.prototype, "highlightCurated", void 0);
NextQuery = __decorate([
    Component({
        components: { BaseSuggestion: __vue_component__ },
        mixins: [xComponentMixin(nextQueriesXModule)]
    })
], NextQuery);
var script = NextQuery;

export { script as default };
