// @empathyco/x-components/js/x-modules/facets/components/lists/filters-search.vue.js
import script from './filters-search.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';
import __vue_create_injector__ from 'vue-runtime-helpers/dist/inject-style/browser.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(
    "div",
    {
      staticClass: "x-list x-filters-search",
      class: _vm.cssClasses,
      attrs: { "data-test": "filters-search" },
    },
    [
      _vm._t(
        "search",
        function () {
          return [
            _c("input", {
              staticClass: "x-input x-filters-search__input",
              attrs: {
                type: "search",
                "data-test": "filters-search-input",
                "aria-label": _vm.filtersSearchInputMessage,
              },
              domProps: { value: _vm.query },
              on: {
                input: function ($event) {
                  return _vm.setQuery($event.target.value)
                },
              },
            }),
          ]
        },
        null,
        { query: _vm.query, setQuery: _vm.setQuery, clearQuery: _vm.clearQuery }
      ),
      _vm._v(" "),
      _vm._t("default", null, { siftedFilters: _vm.siftedFilters }),
    ],
    2
  )
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = function (inject) {
    if (!inject) return
    inject("data-v-129f01b9_0", { source: ".x-filters-search__input[data-v-129f01b9]::-ms-clear {\n  display: none;\n  width: 0;\n  height: 0;\n}\n.x-filters-search__input[data-v-129f01b9]::-ms-reveal {\n  display: none;\n  width: 0;\n  height: 0;\n}\n.x-filters-search__input[data-v-129f01b9]::-webkit-search-decoration, .x-filters-search__input[data-v-129f01b9]::-webkit-search-cancel-button, .x-filters-search__input[data-v-129f01b9]::-webkit-search-results-button, .x-filters-search__input[data-v-129f01b9]::-webkit-search-results-decoration {\n  display: none;\n}", map: undefined, media: undefined });

  };
  /* scoped */
  const __vue_scope_id__ = "data-v-129f01b9";
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    __vue_create_injector__,
    undefined,
    undefined
  );

export { __vue_component__ as default };
