// @empathyco/x-components/js/x-modules/facets/components/lists/selected-filters-list.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import { Prop, Component } from 'vue-property-decorator';
import { isFacetFilter } from '@empathyco/x-types';
import { xComponentMixin } from '../../../../components/x-component.mixin.js';
import { toKebabCase } from '../../../../utils/string.js';
import FacetsMixin from '../facets.mixin.js';
import { facetsXModule } from '../../x-module.js';
import __vue_component__ from './selected-filters.vue.js';

/**
 * This component renders a list of selected filters from every facet, or from the facet
 * ids passed as property. It uses the SelectedFilters component (state).
 *
 * It provides two slots: a scoped one which name is the filter facet id; and a default one.
 * Both exposes the filter and renders the filter label by default.
 *
 * The property "alwaysVisible" handles if the component is rendered if no filters are selected.
 *
 * @remarks It extends {@link FacetsMixin}.
 *
 * @public
 */
let SelectedFiltersList = class SelectedFiltersList extends FacetsMixin {
    /**
     * Transforms a dictionary of Filters including the slot name.
     *
     * @param selectedFilters - A list of selected filters without slot name.
     *
     * @returns A dictionary of facets with the slot name.
     *
     * @internal
     */
    mapSlot(selectedFilters) {
        return selectedFilters.map(filter => ({
            slotName: isFacetFilter(filter) ? toKebabCase(filter.facetId) : 'default',
            selectedFilter: filter
        }));
    }
};
__decorate([
    Prop({ default: 'ul' })
], SelectedFiltersList.prototype, "animation", void 0);
SelectedFiltersList = __decorate([
    Component({
        components: { SelectedFilters: __vue_component__ },
        mixins: [xComponentMixin(facetsXModule)]
    })
], SelectedFiltersList);
var script = SelectedFiltersList;

export { script as default };
