// @empathyco/x-adapter/dist/cjs/empathy/config/empathy-adapter.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_EMPATHY_ADAPTER_CONFIG = void 0;
var cache_config_1 = require("./cache.config");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
exports.DEFAULT_EMPATHY_ADAPTER_CONFIG = {
    env: 'live',
    instance: 'demo',
    requestParams: {
        scope: 'desktop',
        lang: 'en'
    },
    features: {
        nextQueries: {
            endpoint: 'api{env}.empathybroker.com/search/v1/query/{instance}/nextqueries',
            responsePaths: {
                nextQueries: ''
            },
            cacheTTLInMinutes: cache_config_1.DEFAULT_CACHE_CONFIG.TTLInMinutes.Side
        },
        topRecommendations: {
            endpoint: 'api{env}.empathybroker.com/search/v1/query/{instance}/topclicked',
            responsePaths: {
                results: 'topclicked.docs'
            },
            cacheTTLInMinutes: cache_config_1.DEFAULT_CACHE_CONFIG.TTLInMinutes.Side
        },
        sectionRecommendations: {
            endpoint: 'api{env}.empathybroker.com/search/v1/query/{instance}/dw-topclicked',
            responsePaths: {
                results: 'docs',
                showTagging: 'ebTagging.query'
            },
            cacheTTLInMinutes: cache_config_1.DEFAULT_CACHE_CONFIG.TTLInMinutes.Side
        },
        clicksRecommendations: {
            endpoint: 'api{env}.empathybroker.com/search/v1/query/{instance}/dw-lastsearches',
            responsePaths: {
                results: 'docs',
                totalResults: 'numFound',
                showTagging: 'ebTagging.query'
            },
            cacheTTLInMinutes: cache_config_1.DEFAULT_CACHE_CONFIG.TTLInMinutes.Side
        },
        queriesRecommendations: {
            endpoint: 'api{env}.empathybroker.com/search/v1/query/{instance}/dw-productsbyqueries',
            responsePaths: {
                results: 'docs',
                totalResults: 'numFound',
                showTagging: 'ebTagging.query'
            },
            cacheTTLInMinutes: cache_config_1.DEFAULT_CACHE_CONFIG.TTLInMinutes.Side
        },
        userRecommendations: {
            endpoint: 'api{env}.empathybroker.com/search/v1/query/{instance}/dw-contextualizedproducts',
            responsePaths: {
                results: 'docs',
                totalResults: 'numFound',
                showTagging: 'ebTagging.query'
            },
            cacheTTLInMinutes: cache_config_1.DEFAULT_CACHE_CONFIG.TTLInMinutes.Side
        },
        relatedTags: {
            endpoint: 'api{env}.empathybroker.com/search/v1/query/{instance}/relatedtags',
            responsePaths: {
                relatedTags: ''
            },
            cacheTTLInMinutes: cache_config_1.DEFAULT_CACHE_CONFIG.TTLInMinutes.Side
        },
        search: {
            endpoint: 'api{env}.empathybroker.com/search/v1/query/{instance}/searchX',
            responsePaths: {
                banners: 'banner',
                facets: 'content.facets',
                partialResults: 'content.suggestions',
                promoteds: 'promoted',
                queryTagging: 'content.ebTagging.query',
                redirections: 'direct',
                results: 'content.docs',
                spellcheck: 'content.spellchecked',
                totalResults: 'content.numFound'
            },
            cacheTTLInMinutes: cache_config_1.DEFAULT_CACHE_CONFIG.TTLInMinutes.Catalog
        },
        searchById: {
            endpoint: 'api{env}.empathybroker.com/search/v1/query/{instance}/skusearch',
            responsePaths: {
                results: 'content.docs'
            },
            cacheTTLInMinutes: cache_config_1.DEFAULT_CACHE_CONFIG.TTLInMinutes.Catalog
        },
        suggestions: {
            endpoint: 'api{env}.empathybroker.com/search/v1/query/{instance}/empathize',
            responsePaths: {
                suggestions: 'topTrends'
            },
            cacheTTLInMinutes: cache_config_1.DEFAULT_CACHE_CONFIG.TTLInMinutes.Side
        },
        track: {
            endpoint: '',
            responsePaths: {}
        }
    },
    mappings: {
        query: {
            maxLength: 128,
            maxWords: 8
        },
        facets: {
            default: {
                modelName: 'SimpleFacet',
                isDynamic: false
            }, named: {}
        },
        tracking: {
            result: {
                add2cart: 'ebTagging.add2cart',
                click: 'ebTagging.click',
                checkout: 'ebTagging.checkout'
            }
        }
    }
};
