// @empathyco/x-adapter/dist/cjs/empathy/mappers/response.mappers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseMappers = void 0;
var tslib_1 = require("tslib");
var inversify_1 = require("inversify");
var container_const_1 = require("../container/container.const");
/**
 * TODO https://searchbroker.atlassian.net/browse/EX-2163
 *
 * @public
 */
var ResponseMappers = /** @class */ (function () {
    function ResponseMappers(banners, facets, nextQueries, partialResults, promoteds, queryTagging, redirections, relatedTags, results, showTagging, spellcheck, suggestions, totalResults) {
        this.banners = banners;
        this.facets = facets;
        this.nextQueries = nextQueries;
        this.partialResults = partialResults;
        this.promoteds = promoteds;
        this.queryTagging = queryTagging;
        this.redirections = redirections;
        this.relatedTags = relatedTags;
        this.results = results;
        this.showTagging = showTagging;
        this.spellcheck = spellcheck;
        this.suggestions = suggestions;
        this.totalResults = totalResults;
    }
    ResponseMappers = tslib_1.__decorate([
        (0, inversify_1.injectable)(),
        tslib_1.__param(0, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.banners)),
        tslib_1.__param(1, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.facets)),
        tslib_1.__param(2, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.nextQueries)),
        tslib_1.__param(3, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.partialResults)),
        tslib_1.__param(4, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.promoteds)),
        tslib_1.__param(5, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.queryTagging)),
        tslib_1.__param(6, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.redirections)),
        tslib_1.__param(7, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.relatedTags)),
        tslib_1.__param(8, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.results)),
        tslib_1.__param(9, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.showTagging)),
        tslib_1.__param(10, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.spellcheck)),
        tslib_1.__param(11, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.suggestions)),
        tslib_1.__param(12, (0, inversify_1.multiInject)(container_const_1.DEPENDENCIES.ResponseMappers.totalResults)),
        tslib_1.__metadata("design:paramtypes", [Array, Array, Array, Array, Array, Array, Array, Array, Array, Array, Array, Array, Array])
    ], ResponseMappers);
    return ResponseMappers;
}());
exports.ResponseMappers = ResponseMappers;
