// @empathyco/x-components/js/x-modules/history-queries/components/clear-history-queries.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Component } from 'vue-property-decorator';
import { State } from '../../../components/decorators/store.decorators.js';
import __vue_component__ from '../../../components/base-event-button.vue.js';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { historyQueriesXModule } from '../x-module.js';

/**
 * A button that when is pressed, emits the
 * {@link HistoryQueriesXEvents.UserPressedClearHistoryQueries} event, expressing the user
 * intention to clear the whole history of queries.
 *
 * @public
 */
let ClearHistoryQueries = class ClearHistoryQueries extends Vue {
    constructor() {
        super(...arguments);
        /**
         * The list of events that are going to be emitted when the button is pressed.
         *
         * @internal
         */
        this.clearHistoryQueriesEvents = {
            UserPressedClearHistoryQueries: undefined
        };
    }
    /**
     * Returns if the array of history queries is empty.
     *
     * @returns `true` if the {@link historyQueries} array is empty, `false` otherwise.
     * @internal
     */
    get isHistoryQueriesEmpty() {
        return this.historyQueries.length === 0;
    }
    /**
     * Dynamic CSS classes to add to the root element of this component.
     *
     * @returns A booleans dictionary where each key is the class name to add, and the boolean value
     * tells if it should be added or not.
     * @internal
     */
    get dynamicClasses() {
        return {
            'x-clear-history-queries--is-empty': this.isHistoryQueriesEmpty
        };
    }
};
__decorate([
    State('historyQueries', 'historyQueries')
], ClearHistoryQueries.prototype, "historyQueries", void 0);
ClearHistoryQueries = __decorate([
    Component({
        components: { BaseEventButton: __vue_component__ },
        mixins: [xComponentMixin(historyQueriesXModule)]
    })
], ClearHistoryQueries);
var script = ClearHistoryQueries;

export { script as default };
