// @empathyco/x-components/js/x-modules/history-queries/components/history-query.vue.js
import script from './history-query.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c(
    "div",
    { staticClass: "x-suggestion-group x-history-query" },
    [
      _c(
        "BaseSuggestion",
        _vm._b(
          {
            staticClass: "x-history-query__suggestion",
            attrs: { "data-test": "history-query", feature: "history_query" },
            scopedSlots: _vm._u(
              [
                {
                  key: "default",
                  fn: function (ref) {
                    var suggestion = ref.suggestion;
                    var queryHTML = ref.queryHTML;
                    return [
                      _vm._t("default", null, null, {
                        suggestion: suggestion,
                        queryHTML: queryHTML,
                      }),
                    ]
                  },
                },
              ],
              null,
              true
            ),
          },
          "BaseSuggestion",
          {
            suggestion: _vm.suggestion,
            suggestionSelectedEvents: _vm.suggestionSelectedEvents,
            query: _vm.query,
          },
          false
        )
      ),
      _vm._v(" "),
      _c(
        "RemoveHistoryQuery",
        {
          staticClass: "x-history-query__remove",
          attrs: {
            historyQuery: _vm.suggestion,
            "data-test": "remove-history-query",
          },
        },
        [
          _vm._t(
            "remove-button-content",
            function () {
              return [_vm._v("✕")]
            },
            null,
            { suggestion: _vm.suggestion }
          ),
        ],
        2
      ),
    ],
    1
  )
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = undefined;
  /* scoped */
  const __vue_scope_id__ = undefined;
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    undefined,
    undefined,
    undefined
  );

export { __vue_component__ as default };
