// @empathyco/x-components/js/x-modules/history-queries/components/my-history.vue.js
import script from './my-history.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _vm.hasHistoryQueries
    ? _c(
        _vm.animation,
        {
          tag: "component",
          staticClass: "x-my-history x-list",
          attrs: { tag: "ul" },
        },
        _vm._l(_vm.groupByDate, function (historyQueries, date) {
          return _c(
            "li",
            {
              key: date,
              staticClass: "x-my-history-item",
              attrs: { "data-test": "my-history-item" },
            },
            [
              _vm._t(
                "date",
                function () {
                  return [
                    _c(
                      "span",
                      {
                        staticClass: "x-my-history-item__date",
                        attrs: { "data-test": "my-history-date" },
                      },
                      [_vm._v(_vm._s(date))]
                    ),
                  ]
                },
                { date: date }
              ),
              _vm._v(" "),
              _c("BaseSuggestions", {
                staticClass: "x-my-history-queries",
                attrs: {
                  suggestions: historyQueries,
                  "data-test": "my-history-queries",
                  animation: _vm.animation,
                },
                scopedSlots: _vm._u(
                  [
                    {
                      key: "default",
                      fn: function (ref) {
                        var suggestion = ref.suggestion;
                        var index = ref.index;
                        return [
                          _vm._t(
                            "suggestion",
                            function () {
                              return [
                                _c("HistoryQuery", {
                                  staticClass: "x-history-queries__item",
                                  attrs: {
                                    suggestion: suggestion,
                                    "data-test": "history-query-item",
                                  },
                                  scopedSlots: _vm._u(
                                    [
                                      {
                                        key: "default",
                                        fn: function () {
                                          return [
                                            _vm._t(
                                              "suggestion-content",
                                              function () {
                                                return [
                                                  _c(
                                                    "div",
                                                    {
                                                      staticClass:
                                                        "x-list x-list--vertical",
                                                    },
                                                    [
                                                      _c("span", [
                                                        _vm._v(
                                                          _vm._s(
                                                            suggestion.query
                                                          )
                                                        ),
                                                      ]),
                                                      _vm._v(" "),
                                                      _c("span", [
                                                        _vm._v(
                                                          _vm._s(
                                                            _vm.formatTime(
                                                              suggestion.timestamp
                                                            )
                                                          )
                                                        ),
                                                      ]),
                                                    ]
                                                  ),
                                                ]
                                              },
                                              null,
                                              {
                                                suggestion: suggestion,
                                                index: index,
                                              }
                                            ),
                                          ]
                                        },
                                        proxy: true,
                                      },
                                      {
                                        key: "remove-button-content",
                                        fn: function () {
                                          return [
                                            _vm._t(
                                              "suggestion-remove-content",
                                              null,
                                              null,
                                              {
                                                suggestion: suggestion,
                                                index: index,
                                              }
                                            ),
                                          ]
                                        },
                                        proxy: true,
                                      },
                                    ],
                                    null,
                                    true
                                  ),
                                }),
                              ]
                            },
                            null,
                            { suggestion: suggestion, index: index }
                          ),
                        ]
                      },
                    },
                  ],
                  null,
                  true
                ),
              }),
            ],
            2
          )
        }),
        0
      )
    : _vm._e()
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = undefined;
  /* scoped */
  const __vue_scope_id__ = undefined;
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    undefined,
    undefined,
    undefined
  );

export { __vue_component__ as default };
