// @empathyco/x-components/js/x-modules/history-queries/components/my-history.vue_rollup-plugin-vue_script.vue.js
import { __decorate } from 'tslib';
import Vue from 'vue';
import { Prop, Inject, Component } from 'vue-property-decorator';
import { xComponentMixin } from '../../../components/x-component.mixin.js';
import { State } from '../../../components/decorators/store.decorators.js';
import __vue_component__$1 from '../../../components/suggestions/base-suggestions.vue.js';
import { groupItemsBy, isArrayEmpty } from '../../../utils/array.js';
import { historyQueriesXModule } from '../x-module.js';
import __vue_component__ from './history-query.vue.js';

/**
 * The component renders the full history of user searched queries grouped by the day
 * they were performed.
 *
 * @remarks
 *
 * Allows the user to select one of them, emitting the needed events.
 * A history query is just another type of suggestion that contains a query that the user has
 * made in the past.
 * @public
 */
let MyHistory = class MyHistory extends Vue {
    /**
     * Returns a record of history queries grouped by date.
     *
     * @example
     * ```typescript
     *  const historyQueriesGrouped = {
     *    'Monday, January 10th, 2022' : [{
     *      query: 'lego',
     *      modelName: 'HistoryQuery',
     *      timestamp: 121312312
     *    }],
     *    'Tuesday, January 11th, 2022' : [{
     *      query: 'barbie',
     *      modelName: 'HistoryQuery',
     *      timestamp: 15221212
     *    }]
     *  }
     * ```
     * @returns The history queries grouped by date.
     * @internal
     */
    get groupByDate() {
        return groupItemsBy(this.historyQueries, current => {
            return new Date(current.timestamp).toLocaleDateString(this.usedLocale, {
                day: 'numeric',
                weekday: 'long',
                month: 'long',
                year: 'numeric'
            });
        });
    }
    /**
     * Formats a timestamp into `hh:mm [PM/AM]` format.
     *
     * @example
     * ```typescript
     * // locale 'es'
     * console.log(formatTime(Date.now()) // '16:54'.
     *
     * // locale 'en'
     * console.log(formatTime(Date.now()) // '16:54 PM'.
     * ```
     * @param timestamp - The timestamp to format.
     * @returns The formatted time.
     * @internal
     */
    formatTime(timestamp) {
        return new Date(timestamp).toLocaleTimeString(this.usedLocale, {
            hour: '2-digit',
            minute: '2-digit'
        });
    }
    /**
     * The `hasHistoryQueries` computed property is a flag representing if there are history queries
     * stored.
     *
     * @returns True if there are history queries; false otherwise.
     * @internal
     */
    get hasHistoryQueries() {
        return !isArrayEmpty(this.historyQueries);
    }
    /**
     * The locale that it is going to be used. It can be the one send it by the snippet config or
     * the one pass it using the prop.
     *
     * @returns The locale to be used.
     * @internal
     */
    get usedLocale() {
        return this.snippetConfig?.lang ?? this.locale;
    }
};
__decorate([
    Prop({ default: 'ul' })
], MyHistory.prototype, "animation", void 0);
__decorate([
    Prop({ default: 'en' })
], MyHistory.prototype, "locale", void 0);
__decorate([
    State('historyQueries', 'historyQueries')
], MyHistory.prototype, "historyQueries", void 0);
__decorate([
    Inject('snippetConfig')
], MyHistory.prototype, "snippetConfig", void 0);
MyHistory = __decorate([
    Component({
        components: { HistoryQuery: __vue_component__, BaseSuggestions: __vue_component__$1 },
        mixins: [xComponentMixin(historyQueriesXModule)]
    })
], MyHistory);
var script = MyHistory;

export { script as default };
