// @empathyco/x-components/js/x-modules/next-queries/wiring.js
import { namespacedWireCommit, namespacedWireDispatch } from '../../wiring/namespaced-wires.factory.js';
import { createWiring } from '../../wiring/wiring.utils.js';

/**
 * `nextQueries` {@link XModuleName | XModule name}.
 *
 * @internal
 */
const moduleName = 'nextQueries';
/**
 * WireCommit for {@link NextQueriesXModule}.
 *
 * @internal
 */
const wireCommit = namespacedWireCommit(moduleName);
/**
 * WireDispatch for {@link NextQueriesXModule}.
 *
 * @internal
 */
const wireDispatch = namespacedWireDispatch(moduleName);
/**
 * Sets the next queries state `query`.
 *
 * @public
 */
const setNextQueriesQuery = wireCommit('setQuery');
/**
 * Sets the next queries state `query` from url.
 *
 * @public
 */
const setUrlParams = wireDispatch('setUrlParams');
/**
 * Sets the next queries state `params`.
 *
 * @public
 */
const setNextQueriesExtraParams = wireCommit('setParams');
/**
 * Requests and stores the next queries.
 *
 * @public
 */
const fetchAndSaveNextQueriesWire = wireDispatch('fetchAndSaveNextQueries');
/**
 * Sets the next queries state `query` with the last query in history queries.
 *
 * @public
 */
const setQueryFromLastHistoryQueryWire = wireDispatch('setQueryFromLastHistoryQuery');
/**
 * Sets the next queries state `searchedQueries` with the list of history queries.
 *
 * @public
 */
const setSearchedQueries = wireCommit('setSearchedQueries');
/**
 * Wiring configuration for the {@link NextQueriesXModule | next queries module}.
 *
 * @internal
 */
const nextQueriesWiring = createWiring({
    ParamsLoadedFromUrl: {
        setUrlParams
    },
    UserAcceptedAQuery: {
        setNextQueriesQuery
    },
    SessionHistoryQueriesChanged: {
        setSearchedQueries,
        // TODO setQueryFromLastHistoryQuery it has to be called only one time
        setQueryFromLastHistoryQueryWire
    },
    NextQueriesRequestChanged: {
        fetchAndSaveNextQueriesWire
    },
    ExtraParamsChanged: {
        setNextQueriesExtraParams
    }
});

export { fetchAndSaveNextQueriesWire, nextQueriesWiring, setNextQueriesExtraParams, setNextQueriesQuery, setQueryFromLastHistoryQueryWire, setSearchedQueries };
