// @empathyco/x-components/js/x-modules/popular-searches/components/popular-search.vue.js
import script from './popular-search.vue_rollup-plugin-vue_script.vue.js';
import __vue_normalize__ from 'vue-runtime-helpers/dist/normalize-component.mjs';

/* script */
const __vue_script__ = script;

/* template */
var __vue_render__ = function () {
  var _vm = this;
  var _h = _vm.$createElement;
  var _c = _vm._self._c || _h;
  return _c("BaseSuggestion", {
    staticClass: "x-popular-search",
    attrs: {
      suggestion: _vm.suggestion,
      suggestionSelectedEvents: _vm.events,
      "data-test": "popular-search",
      feature: "popular_search",
    },
    scopedSlots: _vm._u(
      [
        {
          key: "default",
          fn: function (ref) {
            var suggestion = ref.suggestion;
            return [
              _vm._t(
                "default",
                function () {
                  return [_vm._v(_vm._s(suggestion.query))]
                },
                { suggestion: suggestion }
              ),
            ]
          },
        },
      ],
      null,
      true
    ),
  })
};
var __vue_staticRenderFns__ = [];
__vue_render__._withStripped = true;

  /* style */
  const __vue_inject_styles__ = undefined;
  /* scoped */
  const __vue_scope_id__ = undefined;
  /* module identifier */
  const __vue_module_identifier__ = undefined;
  /* functional template */
  const __vue_is_functional_template__ = false;
  /* style inject */
  
  /* style inject SSR */
  
  /* style inject shadow dom */
  

  
  const __vue_component__ = /*#__PURE__*/__vue_normalize__(
    { render: __vue_render__, staticRenderFns: __vue_staticRenderFns__ },
    __vue_inject_styles__,
    __vue_script__,
    __vue_scope_id__,
    __vue_is_functional_template__,
    __vue_module_identifier__,
    false,
    undefined,
    undefined,
    undefined
  );

export { __vue_component__ as default };
